{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "omegafold.ipynb",
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU",
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/bforsbe/SK2534/blob/main/omegafold_znf.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Evolving protein sequences using OmegaFold**\n",
        "\n",
        "Omegafold (like ESMfold) is a protein langiage modelthat runs reasonably quick in a notebook like this. For more details see: [Github](https://github.com/HeliXonProtein/OmegaFold), [Preprint](https://www.biorxiv.org/content/10.1101/2022.07.21.500999v1)\n",
        "\n",
        "## Today\n",
        "\n",
        "Many sessions ago you mutated a DNA sequence to get a high mutation score, and tried to keep the fold intact. Today, we let the computers do this over and over again, and keep the mutations that score high without changing the fold.\n",
        "\n",
        "**Why?** Becasue it is a model of evolution, in the sense that there is a \"force\" keeping DNA and protein sequences functional. Today, that force is similarity to the original structure, whereas in nature it is probably keeping function. This is not quite the same, but since fold and function are so closely connected, it is a useful enough proxy.\n",
        "\n",
        "You will recognize many steps from before. Actually running a longer evolution is unfortunately too slow, so you will load and analyse some precomputed results when you've managed to get it running.\n"
      ],
      "metadata": {
        "id": "qEgffIPOyEgk"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "There are currently issues with ESMfold, so we will use OmegaFold, another protein languge model. It runs a bit slower, but is at least faster to install."
      ],
      "metadata": {
        "id": "jbRgM8wuQoVJ"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "oVkGfduwiAkw"
      },
      "outputs": [],
      "source": [
        "#@markdown ##Install OmegaFold\n",
        "import os,sys,re\n",
        "from IPython.utils import io\n",
        "if \"SETUP_DONE\" not in dir():\n",
        "  import torch\n",
        "  device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
        "  with io.capture_output() as captured:\n",
        "    if not os.path.isdir(\"OmegaFold\"):\n",
        "      %shell git clone --branch beta --quiet https://github.com/sokrypton/OmegaFold.git\n",
        "      # %shell cd OmegaFold; pip -q install -r requirements.txt\n",
        "      %shell pip -q install py3Dmol biopython==1.81\n",
        "      %shell apt-get install aria2 -qq > /dev/null\n",
        "      %shell aria2c -q -x 16 https://helixon.s3.amazonaws.com/release1.pt\n",
        "      %shell mkdir -p ~/.cache/omegafold_ckpt\n",
        "      %shell mv release1.pt ~/.cache/omegafold_ckpt/model.pt\n",
        "  SETUP_DONE = True"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We will work with another sequence today, which is a zinc-finger domain. This is a small protein daomin that is usually part of larger proteins, but which does fold on its own. Nature has reinvented this domain many times, which will make the sequence analysis more interesting. Fold it to have a look."
      ],
      "metadata": {
        "id": "MLxXBdTQRC0q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown ##Run **OmegaFold**\n",
        "from string import ascii_uppercase, ascii_lowercase\n",
        "import hashlib\n",
        "import torch\n",
        "from IPython.utils import io\n",
        "\n",
        "# Define device here to ensure it's available\n",
        "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
        "\n",
        "def get_hash(x): return hashlib.sha1(x.encode()).hexdigest()\n",
        "alphabet_list = list(ascii_uppercase+ascii_lowercase)\n",
        "\n",
        "jobname = \"1ZNF\" #@param {type:\"string\"}\n",
        "jobname = re.sub(r'\\W+', '', jobname)[:50]\n",
        "\n",
        "sequence = \"YKCGLCERSFVEKSALSRHQRVHKN\" #@param {type:\"string\"}\n",
        "sequence = re.sub(\"[^A-Z:]\", \"\", sequence.replace(\"/\",\":\").upper())\n",
        "sequence = re.sub(\":+\",\":\",sequence)\n",
        "sequence = re.sub(\"^[:]+\",\"\",sequence)\n",
        "sequence = re.sub(\"[:]+$\",\"\",sequence)\n",
        "\n",
        "copies = 1\n",
        "sequence = \":\".join([sequence] * copies)\n",
        "\n",
        "num_cycle = 2\n",
        "offset_rope = False\n",
        "\n",
        "ID = jobname+\"_\"+get_hash(sequence)[:5]\n",
        "seqs = sequence.split(\":\")\n",
        "lengths = [len(s) for s in seqs]\n",
        "\n",
        "subbatch_size = 1000\n",
        "def get_subbatch_size(L):\n",
        "  if L <  500: return 500\n",
        "  if L < 1000: return 200\n",
        "  return 150\n",
        "if subbatch_size < 0:\n",
        "    subbatch_size = get_subbatch_size(sum(lengths))\n",
        "\n",
        "u_seqs = list(set(seqs))\n",
        "\n",
        "if len(seqs) == 1: mode = \"mono\"\n",
        "elif len(u_seqs) == 1: mode = \"homo\"\n",
        "else: mode = \"hetero\"\n",
        "\n",
        "with open(f\"{ID}.fasta\",\"w\") as out:\n",
        "  out.write(f\">{ID}\\n{sequence}\\n\")\n",
        "\n",
        "# Construct the shell command, handling the boolean offset_rope\n",
        "shell_command = f\"python OmegaFold/main.py --device={device} --subbatch_size={subbatch_size} --num_cycle={num_cycle} {ID}.fasta .\"\n",
        "if offset_rope:\n",
        "    shell_command = f\"python OmegaFold/main.py --offset_rope --device={device} --subbatch_size={subbatch_size} --num_cycle={num_cycle} {ID}.fasta .\"\n",
        "\n",
        "%shell {shell_command}\n",
        "\n",
        "def renum_pdb_str(pdb_str, Ls=None, renum=True, offset=1):\n",
        "  if Ls is not None:\n",
        "    L_init = 0\n",
        "    new_chain = {}\n",
        "    for L,c in zip(Ls, alphabet_list):\n",
        "      new_chain.update({i:c for i in range(L_init,L_init+L)})\n",
        "      L_init += L\n",
        "\n",
        "  n,num,pdb_out = 0,offset,[]\n",
        "  resnum_ = None\n",
        "  chain_ = None\n",
        "  new_chain_ = new_chain[0]\n",
        "  for line in pdb_str.split(\"\\n\"):\n",
        "    if line[:4] == \"ATOM\":\n",
        "      chain = line[21:22]\n",
        "      resnum = int(line[22:22+5])\n",
        "      if resnum_ is None: resnum_ = resnum\n",
        "      if chain_ is None: chain_ = chain\n",
        "      if resnum != resnum_ or chain != chain_:\n",
        "        num += (resnum - resnum_)\n",
        "        n += 1\n",
        "        resnum_,chain_ = resnum,chain\n",
        "      if Ls is not None:\n",
        "        if new_chain[n] != new_chain_:\n",
        "          num = offset\n",
        "          new_chain_ = new_chain[n]\n",
        "      N = num if renum else resnum\n",
        "      if Ls is None: pdb_out.append(\"%s%4i%s\" % (line[:22],N,line[26:]))\n",
        "      else: pdb_out.append(\"%s%s%4i%s\" % (line[:21],new_chain[n],N,line[26:]))\n",
        "  return \"\\n\".join(pdb_out)\n",
        "\n",
        "pdb_str = renum_pdb_str(open(f\"{ID}.pdb\",'r').read(), Ls=lengths)\n",
        "o_pdb_file=f\"{ID}.pdb\"\n",
        "with open(o_pdb_file,\"w\") as out:\n",
        "  out.write(pdb_str)"
      ],
      "metadata": {
        "id": "CFCwEAa2oZEN",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown ##Display\n",
        "import py3Dmol\n",
        "\n",
        "\n",
        "pymol_color_list = [\"#33ff33\",\"#00ffff\",\"#ff33cc\",\"#ffff00\",\"#ff9999\",\"#e5e5e5\",\"#7f7fff\",\"#ff7f00\",\n",
        "                    \"#7fff7f\",\"#199999\",\"#ff007f\",\"#ffdd5e\",\"#8c3f99\",\"#b2b2b2\",\"#007fff\",\"#c4b200\",\n",
        "                    \"#8cb266\",\"#00bfbf\",\"#b27f7f\",\"#fcd1a5\",\"#ff7f7f\",\"#ffbfdd\",\"#7fffff\",\"#ffff7f\",\n",
        "                    \"#00ff7f\",\"#337fcc\",\"#d8337f\",\"#bfff3f\",\"#ff7fff\",\"#d8d8ff\",\"#3fffbf\",\"#b78c4c\",\n",
        "                    \"#339933\",\"#66b2b2\",\"#ba8c84\",\"#84bf00\",\"#b24c66\",\"#7f7f7f\",\"#3f3fa5\",\"#a5512b\"]\n",
        "\n",
        "def show_pdb(pdb_str, show_sidechains=False, show_mainchains=False,\n",
        "             color=\"pLDDT\", chains=None, vmin=50, vmax=90,\n",
        "             size=(800,480), hbondCutoff=4.0,\n",
        "             Ls=None,\n",
        "             animate=False):\n",
        "\n",
        "  if chains is None:\n",
        "    chains = 1 if Ls is None else len(Ls)\n",
        "  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js', width=size[0], height=size[1])\n",
        "  if animate:\n",
        "    view.addModelsAsFrames(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "  else:\n",
        "    view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "  if color == \"pLDDT\":\n",
        "    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':vmin,'max':vmax}}})\n",
        "  elif color == \"rainbow\":\n",
        "    view.setStyle({'cartoon': {'color':'spectrum'}})\n",
        "  elif color == \"chain\":\n",
        "    for n,chain,color in zip(range(chains),alphabet_list,pymol_color_list):\n",
        "       view.setStyle({'chain':chain},{'cartoon': {'color':color}})\n",
        "  if show_sidechains:\n",
        "    BB = ['C','O','N']\n",
        "    HP = [\"ALA\",\"GLY\",\"VAL\",\"ILE\",\"LEU\",\"PHE\",\"MET\",\"PRO\",\"TRP\",\"CYS\",\"TYR\"]\n",
        "    view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
        "                  {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "    view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]},\n",
        "                  {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "    view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n",
        "                  {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "  if show_mainchains:\n",
        "    BB = ['C','O','N','CA']\n",
        "    view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "  view.zoomTo()\n",
        "  if animate: view.animate()\n",
        "  return view\n",
        "\n",
        "color = \"confidence\" #@param [\"confidence\", \"rainbow\", \"chain\"]\n",
        "if color == \"confidence\": color = \"pLDDT\"\n",
        "show_sidechains = True #@param {type:\"boolean\"}\n",
        "show_mainchains = False #@param {type:\"boolean\"}\n",
        "show_pdb(pdb_str, color=color, show_sidechains=show_sidechains, show_mainchains=show_mainchains,\n",
        "         Ls=lengths).show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "dHl9aS-HHFNQ",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Today, we won't mutate manually, but we will keep this cell in case we want to verify anything later.\n",
        "\n",
        "#IMPORTANT\n",
        "You **actually need to run these cells as well**, because they have some functions that we need to automate the evolution can run it.\n",
        "The mutate&inspect, RMSD-score and comparison cells might also be useful later, if you want to check a mutation against the starting sequence."
      ],
      "metadata": {
        "id": "WAFzv_qhSAyp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Mutate DNA and inspect translation\n",
        "from IPython.display import display, HTML\n",
        "from textwrap import wrap\n",
        "import Bio\n",
        "from Bio.Seq import Seq\n",
        "\n",
        "# === Inputs ===\n",
        "original_seq = \"tacaaatgcgggttgtgcgagaggagctttgtggaaaagtctgccctttccagacaccagagggttcacaaaaat\" #@param {type:\"string\"}\n",
        "mutated_seq  = \"tacaaacgcgggttgtgcgagaggagctttgtggaaaagtctgccctttccagacaccagagggttcacaaaaat\" #@param {type:\"string\"}\n",
        "\n",
        "# Clean input\n",
        "original_seq = original_seq.replace(\" \", \"\").upper()\n",
        "mutated_seq  = mutated_seq.replace(\" \", \"\").upper()\n",
        "\n",
        "# === Checks ===\n",
        "if len(original_seq) != len(mutated_seq):\n",
        "    raise ValueError(f\"Sequences must have the same length (got {len(original_seq)} vs {len(mutated_seq)}).\")\n",
        "if len(original_seq) % 3 != 0:\n",
        "    raise ValueError(f\"Sequence length must be a multiple of 3 (got {len(original_seq)}).\")\n",
        "\n",
        "# === Genetic code (DNA codons) ===\n",
        "codon_table = {\n",
        "    \"TTT\":\"F\",\"TTC\":\"F\",\"TTA\":\"L\",\"TTG\":\"L\",\n",
        "    \"CTT\":\"L\",\"CTC\":\"L\",\"CTA\":\"L\",\"CTG\":\"L\",\n",
        "    \"ATT\":\"I\",\"ATC\":\"I\",\"ATA\":\"I\",\"ATG\":\"M\",\n",
        "    \"GTT\":\"V\",\"GTC\":\"V\",\"GTA\":\"V\",\"GTG\":\"V\",\n",
        "    \"TCT\":\"S\",\"TCC\":\"S\",\"TCA\":\"S\",\"TCG\":\"S\",\n",
        "    \"CCT\":\"P\",\"CCC\":\"P\",\"CCA\":\"P\",\"CCG\":\"P\",\n",
        "    \"ACT\":\"T\",\"ACC\":\"T\",\"ACA\":\"T\",\"ACG\":\"T\",\n",
        "    \"GCT\":\"A\",\"GCC\":\"A\",\"GCA\":\"A\",\"GCG\":\"A\",\n",
        "    \"TAT\":\"Y\",\"TAC\":\"Y\",\"TAA\":\"*\",\"TAG\":\"*\",\n",
        "    \"CAT\":\"H\",\"CAC\":\"H\",\"CAA\":\"Q\",\"CAG\":\"Q\",\n",
        "    \"AAT\":\"N\",\"AAC\":\"N\",\"AAA\":\"K\",\"AAG\":\"K\",\n",
        "    \"GAT\":\"D\",\"GAC\":\"D\",\"GAA\":\"E\",\"GAG\":\"E\",\n",
        "    \"TGT\":\"C\",\"TGC\":\"C\",\"TGA\":\"*\",\"TGG\":\"W\",\n",
        "    \"CGT\":\"R\",\"CGC\":\"R\",\"CGA\":\"R\",\"CGG\":\"R\",\n",
        "    \"AGT\":\"S\",\"AGC\":\"S\",\"AGA\":\"R\",\"AGG\":\"R\",\n",
        "    \"GGT\":\"G\",\"GGC\":\"G\",\"GGA\":\"G\",\"GGG\":\"G\"\n",
        "}\n",
        "\n",
        "def translate_dna(seq):\n",
        "    codons = wrap(seq, 3)\n",
        "    aa = [codon_table.get(c,\"?\") for c in codons]\n",
        "    return codons, aa\n",
        "\n",
        "from IPython.display import display, HTML\n",
        "\n",
        "def compare_sequences(seq1, seq2, space_every=10, spacer=\" \", no_number=False):\n",
        "    \"\"\"\n",
        "    Display two sequences in monospace, highlight differences in bold red,\n",
        "    insert spaces every n characters, and add numbering above each block.\n",
        "    \"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        raise ValueError(\"Sequences must be the same length\")\n",
        "\n",
        "    def highlight_block(s1, s2):\n",
        "        \"\"\"Highlight differences in one block of equal length\"\"\"\n",
        "        out1, out2 = [], []\n",
        "        for a, b in zip(s1, s2):\n",
        "            out1.append(a)\n",
        "            if a == b:\n",
        "                out2.append(b)\n",
        "            else:\n",
        "                out2.append(f\"<span style='color:red;font-weight:bold'>{b}</span>\")\n",
        "        return \"\".join(out1), \"\".join(out2)\n",
        "\n",
        "    # Split into blocks of size space_every\n",
        "    blocks1 = [seq1[i:i+space_every] for i in range(0, len(seq1), space_every)]\n",
        "    blocks2 = [seq2[i:i+space_every] for i in range(0, len(seq2), space_every)]\n",
        "\n",
        "    seq1_blocks, seq2_blocks, num_blocks = [], [], []\n",
        "    pos = 0\n",
        "    for b1, b2 in zip(blocks1, blocks2):\n",
        "        o1, o2 = highlight_block(b1, b2)\n",
        "        seq1_blocks.append(o1)\n",
        "        seq2_blocks.append(o2)\n",
        "        pos += len(b1)\n",
        "        num_blocks.append(str(pos).rjust(len(b1)))\n",
        "\n",
        "\n",
        "    number_line = spacer.join(num_blocks)\n",
        "    seq1_line   = spacer.join(seq1_blocks)\n",
        "    seq2_line   = spacer.join(seq2_blocks)\n",
        "\n",
        "    html = []\n",
        "    if no_number:\n",
        "        html = seq1_line + \"<br>\" + seq2_line\n",
        "    else:\n",
        "        html = number_line + \"<br>\" + seq1_line + \"<br>\" + seq2_line\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>{html}</pre>\"))\n",
        "\n",
        "def dna_identity(seq1: str, seq2: str) -> float:\n",
        "    \"\"\"Return % identity at DNA level\"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        raise ValueError(\"Sequences must have the same length\")\n",
        "    matches = sum(a == b for a, b in zip(seq1.upper(), seq2.upper()))\n",
        "    return 100 * matches / len(seq1)\n",
        "\n",
        "def protein_identity(seq1: str, seq2: str) -> float:\n",
        "    \"\"\"Translate DNA → protein and return % identity\"\"\"\n",
        "    if len(seq1) % 3 != 0 or len(seq2) % 3 != 0:\n",
        "        raise ValueError(\"DNA sequences must be divisible by 3\")\n",
        "    prot1 = str(Seq(seq1).translate(to_stop=False))\n",
        "    prot2 = str(Seq(seq2).translate(to_stop=False))\n",
        "    if len(prot1) != len(prot2):\n",
        "        raise ValueError(\"Translated proteins have different lengths\")\n",
        "    matches = sum(a == b for a, b in zip(prot1, prot2))\n",
        "    return 100 * matches / len(prot1)\n",
        "\n",
        "def mutation_score(seq1: str, seq2: str) -> float:\n",
        "    \"\"\"Return mutation score\"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        raise ValueError(\"Sequences must have the same length\")\n",
        "\n",
        "    protein_id = protein_identity(seq1, seq2)\n",
        "    dna_id = dna_identity(seq1, seq2)\n",
        "\n",
        "    return ((100.-protein_id)+(100.-dna_id))/2\n",
        "\n",
        "# === Translation ===\n",
        "codons_o, aa_o = translate_dna(original_seq)\n",
        "codons_m, aa_m = translate_dna(mutated_seq)\n",
        "\n",
        "# === HTML display ===\n",
        "\n",
        "compare_sequences(original_seq,mutated_seq,space_every=3)\n",
        "compare_sequences(aa_o,aa_m,space_every=1, no_number=True, spacer=\"   \")\n",
        "\n",
        "#display(HTML(f\"<pre style='font-family:monospace'>{\"   \".join(aa_o)}</pre>\"))\n",
        "#compare_sequences(\"\".join(codons_o),\"\".join(codons_m),space_every=30)\n",
        "\n",
        "print(\"\\n\\nMutaded sequence without spaces:\")\n",
        "display(HTML(f\"<pre style='font-family:monospace'>{\"\".join(aa_m)}</pre>\"))\n",
        "\n",
        "dna_id = dna_identity(original_seq, mutated_seq)\n",
        "#print(f\"\\nDNA % identity: {dna_id:.1f}%\")\n",
        "protein_id = protein_identity(original_seq, mutated_seq)\n",
        "#print(f\"Protein % identity: {protein_id:.1f}%\")\n",
        "print(\"\\n\")\n",
        "mut_score = mutation_score(original_seq,mutated_seq)\n",
        "print(f\"Mutation score is {mut_score:.2f}\")\n"
      ],
      "metadata": {
        "id": "z3HeD2V8epVW",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Calculate RMSD-based final score\n",
        "import os, re, requests\n",
        "import torch\n",
        "from jax.tree_util import tree_map\n",
        "from Bio.Seq import Seq\n",
        "from Bio.PDB import PDBParser, Superimposer, PDBIO\n",
        "import numpy as np\n",
        "import tempfile\n",
        "\n",
        "# --- Superpose prediction onto reference ---\n",
        "def align_pred_to_ref(pred_pdb_str, ref_pdb_file, out_file=\"pred_aligned.pdb\"):\n",
        "    parser = PDBParser(QUIET=True)\n",
        "\n",
        "    # Save predicted PDB string to temp file\n",
        "    with tempfile.NamedTemporaryFile(delete=False, suffix=\".pdb\") as tmp_pred:\n",
        "        tmp_pred.write(pred_pdb_str.encode())\n",
        "        pred_file = tmp_pred.name\n",
        "\n",
        "    pred = parser.get_structure(\"pred\", pred_file)\n",
        "    ref  = parser.get_structure(\"ref\", ref_pdb_file)\n",
        "\n",
        "    # Use CA atoms for alignment\n",
        "    pred_atoms = [a for a in pred.get_atoms() if a.get_id() == \"CA\"]\n",
        "    ref_atoms  = [a for a in ref.get_atoms()  if a.get_id() == \"CA\"]\n",
        "\n",
        "    n = min(len(pred_atoms), len(ref_atoms))\n",
        "    si = Superimposer()\n",
        "    si.set_atoms(ref_atoms[:n], pred_atoms[:n])  # align prediction to reference\n",
        "    si.apply(pred.get_atoms())                   # apply transform to prediction\n",
        "\n",
        "    # Write aligned prediction\n",
        "    io = PDBIO()\n",
        "    io.set_structure(pred)\n",
        "    io.save(out_file)\n",
        "\n",
        "    return out_file, si.rms\n",
        "\n",
        "def rmsd_to_similarity(rmsd, sigma=2.0):\n",
        "    \"\"\"\n",
        "    Convert RMSD (Å) to a similarity score in [0,1].\n",
        "    Smaller RMSD -> score close to 1, larger RMSD -> score close to 0.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    rmsd : float or np.array\n",
        "        Root-mean-square deviation in angstroms.\n",
        "    sigma : float\n",
        "        Scaling factor controlling decay. Typical RMSDs of 1-2 Å → high similarity.\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    similarity : float or np.array\n",
        "        Score in [0,1].\n",
        "    \"\"\"\n",
        "    return np.exp(-rmsd / sigma)\n",
        "\n",
        "# --- Download reference PDB ---\n",
        "ref_file=\"reference.pdb\"\n",
        "if not os.path.isfile(ref_file):\n",
        "  !wget -q -O reference.pdb https://raw.githubusercontent.com/bforsbe/SK2534/main/reference_omega_znf.pdb\n",
        "\n",
        "aligned_file, rmsd_val = align_pred_to_ref(pdb_str, ref_file)\n",
        "\n",
        "#for rmsd in [0.5, 1.0, 2.0, 5.0]:\n",
        "#    print(rmsd, rmsd_to_similarity(rmsd))\n",
        "pred_sim_score = rmsd_to_similarity(rmsd_val)\n",
        "mut_score=mutation_score(original_seq,mutated_seq)\n",
        "\n",
        "# --- Compare ---\n",
        "print(f\"Input file is {o_pdb_file}\")\n",
        "#print(f\"Protein sequence: {protein_seq}\")\n",
        "print(f\"mutation score: \\t{mut_score:.2f})\")\n",
        "print(f\"RMSD vs reference: \\t{rmsd_val:.3f} Å (score: {pred_sim_score:.2f})\")\n",
        "print(\"\\n\\n\")\n",
        "print(f\"Total score: \\t{mut_score*pred_sim_score:.3f}\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "c51BDdsXe68l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Display predicted (superposed) + reference {run: \"auto\"}\n",
        "import py3Dmol\n",
        "from Bio.PDB import PDBParser, Superimposer, PDBIO\n",
        "import tempfile, os\n",
        "from google.colab import files\n",
        "\n",
        "# --- Visualize both ---\n",
        "def show_pred_and_ref_aligned(pred_pdb_str, ref_pdb_file,\n",
        "                              color=\"pLDDT\", show_sidechains=False, show_mainchains=False,\n",
        "                              chains=None, vmin=50, vmax=90, size=(800,480), Ls=None, hide_pred=False,\n",
        "                              download_aligned_pdbs=False): # Added boolean option\n",
        "\n",
        "    aligned_file, rmsd = align_pred_to_ref(pred_pdb_str, ref_pdb_file)\n",
        "\n",
        "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',\n",
        "                        width=size[0], height=size[1])\n",
        "\n",
        "    # Reference (fixed, green)\n",
        "    with open(ref_pdb_file) as f:\n",
        "        ref_pdb_str = f.read()\n",
        "    view.addModel(ref_pdb_str,'pdb')\n",
        "    view.setStyle({'model':0},{'cartoon':{'color':'green','opacity':0.7}})\n",
        "\n",
        "    # Prediction (aligned, colored)\n",
        "    with open(aligned_file) as f:\n",
        "        pred_aligned_str = f.read()\n",
        "    if not hide_pred:\n",
        "        view.addModel(pred_aligned_str,'pdb')\n",
        "    if color == \"pLDDT\":\n",
        "        view.setStyle({'model':1,'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':vmin,'max':vmax}}})\n",
        "    elif color == \"rainbow\":\n",
        "        view.setStyle({'model':1,'cartoon': {'color':'spectrum'}})\n",
        "    elif color == \"chain\":\n",
        "        from string import ascii_uppercase\n",
        "        alphabet_list = list(ascii_uppercase)\n",
        "        for n,chain,c in zip(range(chains),alphabet_list,pymol_color_list):\n",
        "            view.setStyle({'model':1,'chain':chain},{'cartoon': {'color':c}})\n",
        "\n",
        "    if show_sidechains:\n",
        "        view.addStyle({'model':1,'stick':{}})\n",
        "\n",
        "    view.zoomTo()\n",
        "    print(f\"Structural alignment RMSD: {rmsd:.3f} Å\")\n",
        "\n",
        "    # Download aligned PDB files if the option is True\n",
        "    if download_aligned_pdbs:\n",
        "        # Download the aligned predicted PDB\n",
        "        files.download(aligned_file)\n",
        "        print(f\"Downloaded aligned predicted PDB: {aligned_file}\")\n",
        "\n",
        "        # Optionally, download the reference PDB as well for completeness\n",
        "        # files.download(ref_pdb_file)\n",
        "        # print(f\"Downloaded reference PDB: {ref_pdb_file}\")\n",
        "\n",
        "\n",
        "    return view\n",
        "\n",
        "\n",
        "# --- Params ---\n",
        "color = \"confidence\" #@param [\"confidence\", \"rainbow\", \"chain\"]\n",
        "if color == \"confidence\": color = \"pLDDT\"\n",
        "show_sidechains = True #@param {type:\"boolean\"}\n",
        "show_mainchains = False #@param {type:\"boolean\"}\n",
        "hide_predicted = False #@param {type:\"boolean\"}\n",
        "download_aligned_pdbs = False #@param {type:\"boolean\"}\n",
        "\n",
        "# --- Run ---\n",
        "\n",
        "show_pred_and_ref_aligned(pdb_str, \"reference.pdb\",\n",
        "                          color=color,\n",
        "                          show_sidechains=show_sidechains,\n",
        "                          show_mainchains=show_mainchains,\n",
        "                          Ls=lengths,\n",
        "                          hide_pred=hide_predicted,\n",
        "                          download_aligned_pdbs=download_aligned_pdbs).show() # Pass the new parameter"
      ],
      "metadata": {
        "cellView": "form",
        "id": "ssk3BxdUgdyS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Ready to evolve!\n",
        "\n",
        "So the procedure is to\n",
        "\n",
        "\n",
        "1. Suggest a random mutation to a random position in the sequence\n",
        "2. Use OmegaFold to predict the structure\n",
        "3. Compute a score based on sequence and structure similarity\n",
        "4. Accept if the score is better, otherwise reject\n",
        "\n",
        "This goes on for as long as you specify. For a trial run, we suggest maybe 25 iterations.  \n",
        "\n"
      ],
      "metadata": {
        "id": "wI1hVguJVG_2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# @title Iterative mutation evolution\n",
        "\n",
        "import numpy as np\n",
        "import numpy.random as random\n",
        "import os, re, requests # Import necessary modules for OmegaFold and file handling\n",
        "from string import ascii_uppercase, ascii_lowercase # Import for chain naming\n",
        "import hashlib # Import for hashing sequence\n",
        "import pickle # Import for saving/loading\n",
        "\n",
        "num_iterations = 25 # @param {type:\"integer\"}\n",
        "save_history = True # @param {type:\"boolean\"}\n",
        "history_filename = \"mutation_history.pkl\" # @param {type:\"string\"}\n",
        "load_history = False # @param {type:\"boolean\"}\n",
        "use_precomputed = False # @param {type:\"boolean\"}\n",
        "conservative_mutations = False # @param {type:\"boolean\"}\n",
        "\n",
        "conservative_substitutions = {\n",
        "    'A': ['G', 'S', 'T'], # Alanine (small, nonpolar)\n",
        "    'R': ['K', 'H'],     # Arginine (basic, positive)\n",
        "    'N': ['Q', 'S', 'T'], # Asparagine (polar, uncharged)\n",
        "    'D': ['E'],          # Aspartic Acid (acidic, negative)\n",
        "    'C': ['S'],          # Cysteine (sulfur-containing)\n",
        "    'Q': ['N', 'H', 'K', 'R', 'E', 'D'], # Glutamine (polar, uncharged) - larger group due to flexibility\n",
        "    'E': ['D', 'Q', 'N', 'K', 'R', 'H'], # Glutamic Acid (acidic, negative) - larger group due to flexibility\n",
        "    'G': ['A', 'S'],     # Glycine (small, nonpolar)\n",
        "    'H': ['R', 'K', 'N', 'Q'], # Histidine (basic, positive, aromatic)\n",
        "    'I': ['L', 'V', 'M'], # Isoleucine (hydrophobic)\n",
        "    'L': ['I', 'V', 'M'], # Leucine (hydrophobic)\n",
        "    'K': ['R', 'H', 'Q', 'N'], # Lysine (basic, positive)\n",
        "    'M': ['L', 'I', 'V'], # Methionine (hydrophobic, sulfur-containing)\n",
        "    'F': ['Y', 'W'],     # Phenylalanine (aromatic, hydrophobic)\n",
        "    'P': [],             # Proline (cyclic, often disrupts structure - few conservative substitutions)\n",
        "    'S': ['T', 'N', 'Q', 'A', 'G', 'C'], # Serine (polar, uncharged, hydroxyl)\n",
        "    'T': ['S', 'N', 'Q', 'A', 'G'], # Threonine (polar, uncharged, hydroxyl)\n",
        "    'W': ['Y', 'F'],     # Tryptophan (aromatic, hydrophobic, large)\n",
        "    'Y': ['F', 'W'],     # Tyrosine (aromatic, hydrophobic, hydroxyl)\n",
        "    'V': ['I', 'L', 'M']  # Valine (hydrophobic)\n",
        "}\n",
        "\n",
        "all_amino_acids = list(\"ACDEFGHIKLMNPQRSTVWY\")\n",
        "\n",
        "def generate_mutations(dna_sequence, codon_position, conservative_only=True):\n",
        "    \"\"\"\n",
        "    Generates a list of mutated DNA sequences with either conservative or any\n",
        "    amino acid substitutions at a given codon position.\n",
        "\n",
        "    Args:\n",
        "        dna_sequence: The original DNA sequence string.\n",
        "        codon_position: The 0-indexed position of the codon to mutate.\n",
        "        conservative_only: If True, only generates conservative substitutions.\n",
        "                           If False, generates all possible amino acid substitutions.\n",
        "\n",
        "    Returns:\n",
        "        A list of mutated DNA sequences.\n",
        "    \"\"\"\n",
        "    start_index = codon_position * 3\n",
        "    end_index = start_index + 3\n",
        "    original_codon = dna_sequence[start_index:end_index]\n",
        "\n",
        "    if original_codon not in codon_table:\n",
        "        print(f\"Warning: Original codon '{original_codon}' not found in codon table.\")\n",
        "        return []\n",
        "\n",
        "    original_amino_acid = codon_table[original_codon]\n",
        "\n",
        "    possible_sub_aas = []\n",
        "    if conservative_only:\n",
        "        if original_amino_acid in conservative_substitutions:\n",
        "            possible_sub_aas = conservative_substitutions[original_amino_acid]\n",
        "        # Note: If conservative_only is True and no conservative substitutions exist, possible_sub_aas remains empty.\n",
        "    else:\n",
        "        possible_sub_aas = [aa for aa in all_amino_acids if aa != original_amino_acid] # All other amino acids\n",
        "\n",
        "    mutated_sequences = []\n",
        "    for sub_aa in possible_sub_aas:\n",
        "        for codon, aa in codon_table.items():\n",
        "            if aa == sub_aa:\n",
        "                mutated_seq = list(dna_sequence)\n",
        "                mutated_seq[start_index:end_index] = list(codon)\n",
        "                mutated_sequences.append(\"\".join(mutated_seq))\n",
        "\n",
        "    return mutated_sequences\n",
        "\n",
        "# Example usage (assuming original_seq is defined from previous cells)\n",
        "# For testing, let's use the first codon (position 0)\n",
        "# mutated_dna_list = generate_mutations(original_seq, 0, conservative_only=True)\n",
        "# print(f\"Original DNA: {original_seq}\")\n",
        "# print(f\"Mutated DNA sequences (conservative substitutions at position 0):\")\n",
        "# for mut_seq in mutated_dna_list:\n",
        "#     print(mut_seq)\n",
        "\n",
        "# mutated_dna_list_any = generate_mutations(original_seq, 0, conservative_only=False)\n",
        "# print(f\"\\nMutated DNA sequences (any substitutions at position 0):\")\n",
        "# for mut_seq in mutated_dna_list_any:\n",
        "#     print(mut_seq)\n",
        "\n",
        "# Function to get hash for jobname\n",
        "def get_hash(x): return hashlib.sha1(x.encode()).hexdigest()\n",
        "alphabet_list = list(ascii_uppercase+ascii_lowercase)\n",
        "\n",
        "if use_precomputed:\n",
        "  history_filename=\"mutation_history_znf.pkl\"\n",
        "  load_history = True\n",
        "  #num_iterations=np.max([num_iterations,200])\n",
        "if not os.path.isfile(history_filename):\n",
        "  !wget -q -O mutation_history_znf.pkl https://raw.githubusercontent.com/bforsbe/SK2534/main/mutation_history_znf.pkl\n",
        "\n",
        "# Load history if requested\n",
        "if load_history and os.path.exists(history_filename):\n",
        "    print(f\"Loading history from {history_filename}\")\n",
        "    with open(history_filename, 'rb') as f:\n",
        "        history = pickle.load(f)\n",
        "    best_dna_sequence = history['best_dna_sequence']\n",
        "    best_protein_sequence = history['best_protein_sequence']\n",
        "    best_total_score = history['best_total_score']\n",
        "    accepted_mutation_scores = history['accepted_mutation_scores']\n",
        "    accepted_sim_scores = history['accepted_sim_scores']\n",
        "    accepted_total_scores = history['accepted_total_scores']\n",
        "    accepted_iterations = history['accepted_iterations']\n",
        "    all_mutation_scores = history['all_mutation_scores']\n",
        "    all_sim_scores = history['all_sim_scores']\n",
        "    all_total_scores = history['all_total_scores']\n",
        "    all_iterations = history['all_iterations']\n",
        "    all_tried_protein_sequences = history['all_tried_protein_sequences']\n",
        "    completed_iters = history['completed_iters']\n",
        "else:\n",
        "    best_dna_sequence = original_seq\n",
        "    best_protein_sequence = str(Bio.Seq.Seq(best_dna_sequence).translate(to_stop=False))\n",
        "    best_total_score = -1\n",
        "\n",
        "    accepted_mutation_scores = []\n",
        "    accepted_sim_scores = []\n",
        "    accepted_total_scores = []\n",
        "    accepted_iterations = []\n",
        "    all_mutation_scores = []\n",
        "    all_sim_scores = []\n",
        "    all_total_scores = []\n",
        "    all_iterations = []\n",
        "\n",
        "    all_tried_protein_sequences = []\n",
        "    all_tried_protein_sequences.append(best_protein_sequence)\n",
        "\n",
        "    completed_iters = 0\n",
        "\n",
        "num_codons = len(original_seq) // 3\n",
        "\n",
        "\n",
        "while completed_iters < num_iterations:\n",
        "    # Choose a random codon position to mutate\n",
        "    random_codon_pos = random.randint(0, num_codons - 1)\n",
        "\n",
        "    # Generate conservative mutations for this position\n",
        "    # This part seems a bit off, it should generate a list and then pick one\n",
        "    mutated_dna_list_single = generate_mutations(best_dna_sequence, random_codon_pos, conservative_only=conservative_mutations)\n",
        "\n",
        "\n",
        "    if not mutated_dna_list_single:\n",
        "        continue # Skip to the next iteration if no mutations are possible\n",
        "\n",
        "    # Select the first generated mutation (or choose randomly)\n",
        "    single_mutated_dna_seq = random.choice(mutated_dna_list_single) # Use random.choice to pick one mutation\n",
        "    single_mutated_protein_seq = str(Bio.Seq.Seq(single_mutated_dna_seq).translate(to_stop=False))\n",
        "\n",
        "    # Check if protein sequence has been tried before\n",
        "    if single_mutated_protein_seq in all_tried_protein_sequences:\n",
        "        continue\n",
        "\n",
        "    all_tried_protein_sequences.append(single_mutated_protein_seq)\n",
        "\n",
        "    # Predict structure using OmegaFold\n",
        "    # Adapt the OmegaFold execution code from cell CFCwEAa2oZEN\n",
        "    #try:\n",
        "    # Generate a unique jobname based on iteration and a hash of the sequence\n",
        "    jobname = f\"iter_{completed_iters}_{get_hash(single_mutated_protein_seq)[:5]}\"\n",
        "    ID = jobname\n",
        "    seqs = single_mutated_protein_seq.split(\":\")\n",
        "    lengths = [len(s) for s in seqs]\n",
        "\n",
        "    # Write fasta file\n",
        "    with open(f\"{ID}.fasta\",\"w\") as out:\n",
        "        out.write(f\">{ID}\\n{single_mutated_protein_seq}\\n\")\n",
        "\n",
        "    # Run OmegaFold command\n",
        "    # Use appropriate parameters, likely inherited from the successful OmegaFold cell\n",
        "    subbatch_size = 1000 # Assuming single chain, length < 500 from previous cell\n",
        "    num_cycle = 2 # Assuming from previous cell\n",
        "    offset_rope = False # Assuming from previous cell\n",
        "    device = \"cuda\" if torch.cuda.is_available() else \"cpu\" # Get device from setup cell\n",
        "\n",
        "    # Execute OmegaFold prediction\n",
        "    # Capture output to keep the notebook clean\n",
        "    with io.capture_output() as captured:\n",
        "        %shell python OmegaFold/main.py --offset_rope={offset_rope} --device={device} --subbatch_size={subbatch_size} --num_cycle={num_cycle} {ID}.fasta .\n",
        "\n",
        "    # Define the predicted PDB file name\n",
        "    single_mutated_pdb_file = f\"{ID}.pdb\"\n",
        "\n",
        "    # Read the predicted PDB file\n",
        "    with open(single_mutated_pdb_file, 'r') as f:\n",
        "        single_mutated_pdb_str = f.read()\n",
        "\n",
        "    # Renumber PDB if necessary (using the function from CFCwEAa2oZEN)\n",
        "    single_mutated_pdb_str = renum_pdb_str(single_mutated_pdb_str, Ls=lengths)\n",
        "    # Save the renumbered PDB string to a file for alignment\n",
        "    with open(single_mutated_pdb_file, \"w\") as out:\n",
        "          out.write(single_mutated_pdb_str)\n",
        "\n",
        "\n",
        "    # Align to reference and calculate RMSD (using the function from c51BDdsXe68l)\n",
        "    aligned_file_single, single_mutated_rmsd_val = align_pred_to_ref(single_mutated_pdb_str, \"reference.pdb\", out_file=f\"{ID}_aligned.pdb\")\n",
        "\n",
        "\n",
        "    # Convert RMSD to similarity score (using the function from c51BDdsXe68l)\n",
        "    single_mutated_pred_sim_score = rmsd_to_similarity(single_mutated_rmsd_val)\n",
        "\n",
        "    # Calculate mutation score (using the function from z3HeD2V8epVW)\n",
        "    single_current_mutation_score = mutation_score(original_seq, single_mutated_dna_seq)\n",
        "\n",
        "    # Calculate total score\n",
        "    single_current_total_score = single_current_mutation_score * single_mutated_pred_sim_score\n",
        "\n",
        "    # Store scores for ALL tried sequences\n",
        "    all_iterations.append(completed_iters + 1)\n",
        "    all_mutation_scores.append(single_current_mutation_score)\n",
        "    all_sim_scores.append(single_mutated_pred_sim_score)\n",
        "    all_total_scores.append(single_current_total_score)\n",
        "\n",
        "    # Print trial sequence and total score\n",
        "    print(f\"Iteration {completed_iters + 1}: Trial Sequence: {single_mutated_protein_seq}, Total Score: {single_current_total_score:.3f}\")\n",
        "\n",
        "    # Acceptance criteria: accept if total score improves or prediction is very similar\n",
        "    if single_current_total_score > best_total_score or single_mutated_pred_sim_score > 0.9:\n",
        "        best_total_score = single_current_total_score\n",
        "        best_dna_sequence = single_mutated_dna_seq\n",
        "        best_protein_sequence = single_mutated_protein_seq\n",
        "\n",
        "        # Store scores for ACCEPTED sequences\n",
        "        accepted_iterations.append(completed_iters + 1)\n",
        "        accepted_mutation_scores.append(single_current_mutation_score)\n",
        "        accepted_sim_scores.append(single_mutated_pred_sim_score)\n",
        "        accepted_total_scores.append(single_current_total_score)\n",
        "\n",
        "        #print(f\"  Accepted mutation. New best total score: {best_total_score:.3f}\")\n",
        "\n",
        "    completed_iters += 1\n",
        "\n",
        "    # Save history\n",
        "    if save_history:\n",
        "        history = {\n",
        "            'best_dna_sequence': best_dna_sequence,\n",
        "            'best_protein_sequence': best_protein_sequence,\n",
        "            'best_total_score': best_total_score,\n",
        "            'accepted_mutation_scores': accepted_mutation_scores,\n",
        "            'accepted_sim_scores': accepted_sim_scores,\n",
        "            'accepted_total_scores': accepted_total_scores,\n",
        "            'accepted_iterations': accepted_iterations,\n",
        "            'all_mutation_scores': all_mutation_scores,\n",
        "            'all_sim_scores': all_sim_scores,\n",
        "            'all_total_scores': all_total_scores,\n",
        "            'all_iterations': all_iterations,\n",
        "            'all_tried_protein_sequences': all_tried_protein_sequences,\n",
        "            'completed_iters': completed_iters\n",
        "        }\n",
        "        with open(history_filename, 'wb') as f:\n",
        "            pickle.dump(history, f)\n",
        "        # Also save as .npy for easier access (optional)\n",
        "        np.save(history_filename.replace('.pkl', '.npy'), history)\n",
        "\n",
        "\n",
        "#except Exception as e:\n",
        "#    print(f\"Iteration {completed_iters + 1}: Error processing mutation: {e}. Skipping.\")\n",
        "#    completed_iters += 1 # Increment even if error occurs to avoid infinite loops if errors persist\n",
        "#    continue # Continue to the next iteration if an error occurs\n",
        "\n",
        "print(\"\\nIterative mutation process complete.\")\n",
        "print(f\"Best Mutated DNA Sequence: {best_dna_sequence}\")\n",
        "print(f\"Best Mutated Protein Sequence: {best_protein_sequence}\")\n",
        "# Recalculate final scores for the best sequence\n",
        "best_mutation_score = mutation_score(original_seq, best_dna_sequence)\n",
        "# To get the final RMSD and prediction similarity score for the best sequence,\n",
        "# you would need to re-run the prediction and alignment for best_protein_sequence\n",
        "# or store these values when the best sequence is updated.\n",
        "# For simplicity, we'll just report the best total score found.\n",
        "print(f\"Best Total Score Found: {best_total_score:.3f}\")\n",
        "print(f\"Total number of tried protein sequences: {len(all_tried_protein_sequences)}\")\n",
        "print(f\"Total number of accepted mutations: {len(accepted_total_scores)}\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "kHO4EmWdi9CC",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now let's print the history with some clarifying colors and summary analysis."
      ],
      "metadata": {
        "id": "apfn0rfkXtmk"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "collapsed": true,
        "id": "1749ea61",
        "cellView": "form"
      },
      "source": [
        "#@title Print full mutation history\n",
        "\n",
        "from IPython.display import display, HTML\n",
        "import numpy as np\n",
        "\n",
        "# Assuming the history dictionary is loaded from the previous cell\n",
        "# history = np.load(history_filename.replace('.pkl', '.npy'), allow_pickle=True).item()\n",
        "\n",
        "all_iterations = history['all_iterations']\n",
        "all_total_scores = history['all_total_scores']\n",
        "all_tried_protein_sequences = history['all_tried_protein_sequences'] # These are already protein sequences\n",
        "\n",
        "accepted_iterations = history['accepted_iterations']\n",
        "accepted_total_scores = history['accepted_total_scores']\n",
        "\n",
        "\n",
        "# Function to format text with color and bold\n",
        "def colored_text(text, color):\n",
        "    return f\"<span style='color:{color};font-weight:bold'>{text}</span>\"\n",
        "\n",
        "def highlight_mutation(original_seq, mutated_seq, color):\n",
        "    \"\"\"Highlights the first different character in mutated_seq with the specified color and bold.\"\"\"\n",
        "    highlighted_seq = []\n",
        "    found_mutation = False\n",
        "    for i in range(len(original_seq)):\n",
        "        # Ensure we don't go out of bounds if sequences have different lengths\n",
        "        if i < len(mutated_seq):\n",
        "            # Check if the characters are different AND we haven't found the first mutation yet\n",
        "            if original_seq[i] != mutated_seq[i] and not found_mutation:\n",
        "                highlighted_seq.append(colored_text(mutated_seq[i], color))\n",
        "                found_mutation = True\n",
        "            else:\n",
        "                highlighted_seq.append(mutated_seq[i])\n",
        "        # If mutated_seq is shorter, just append the rest of the original (shouldn't happen with current logic)\n",
        "        # else:\n",
        "        #    highlighted_seq.append(original_seq[i])\n",
        "\n",
        "\n",
        "    return \"\".join(highlighted_seq)\n",
        "\n",
        "\n",
        "print(\"Mutation History:\")\n",
        "\n",
        "# Initialize best score from history or -1 if history is empty\n",
        "best_score_so_far = history['best_total_score'] if 'best_total_score' in history else -1\n",
        "\n",
        "# Keep track of the last accepted protein sequence\n",
        "last_accepted_protein_sequence = history['all_tried_protein_sequences'][0] # Start with the original sequence\n",
        "\n",
        "\n",
        "# Iterate through all tried sequences starting from the second one (index 1)\n",
        "for i in range(len(all_iterations)): # Iterate through all iterations (which correspond to trials after the first)\n",
        "    iteration = all_iterations[i]\n",
        "    current_protein_sequence = all_tried_protein_sequences[i+1] # Trial sequences are at index i+1\n",
        "\n",
        "    # The total score for the current sequence is in all_total_scores at index i\n",
        "    total_score = all_total_scores[i]\n",
        "\n",
        "\n",
        "    # Determine the color for highlighting the mutated amino acid\n",
        "    highlight_color = 'red'  # Default color for rejected\n",
        "    is_accepted = all_iterations[i] in accepted_iterations\n",
        "    if is_accepted:\n",
        "        highlight_color = 'blue'\n",
        "        # Note: The comparison for highlighting is still against the *last accepted* sequence\n",
        "        # even if the current one is now the new *last accepted*. This shows the change\n",
        "        # from the previous accepted state.\n",
        "\n",
        "\n",
        "    # Highlight the mutated amino acid by comparing to the LAST ACCEPTED sequence\n",
        "    highlighted_sequence = highlight_mutation(last_accepted_protein_sequence, current_protein_sequence, highlight_color)\n",
        "\n",
        "    # Update the last accepted sequence *after* highlighting if the current one was accepted\n",
        "    if is_accepted:\n",
        "         last_accepted_protein_sequence = current_protein_sequence\n",
        "\n",
        "\n",
        "    score_text = f\"Total Score: {total_score:.3f}\"\n",
        "    # Color the score text blue if accepted, green if a new best score, otherwise uncolored.\n",
        "    if is_accepted:\n",
        "        score_text = colored_text(score_text, 'blue')\n",
        "    # Check against the overall best score found so far, not just the last accepted score\n",
        "    if total_score > best_score_so_far:\n",
        "        score_text = colored_text(score_text, 'blue') # New best score takes precedence for coloring\n",
        "        best_score_so_far = total_score # Update best score for comparison\n",
        "\n",
        "\n",
        "    output_line = f\"Iteration {iteration}: Trial Sequence: {highlighted_sequence}, {score_text}\"\n",
        "\n",
        "    # Display the line without coloring the entire line\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>{output_line}</pre>\"))\n",
        "\n",
        "\n",
        "print(\"\\nHistory printing complete.\")\n",
        "\n",
        "print(\"\\nComparison: Original Sequence vs Best Found Sequence\")\n",
        "# Assuming original_seq (DNA) and best_protein_sequence (Protein) are available\n",
        "# Need to translate original_seq to protein for comparison\n",
        "original_protein_seq = str(Bio.Seq.Seq(original_seq).translate(to_stop=False))\n",
        "best_protein_sequence = history['best_protein_sequence']\n",
        "\n",
        "# Use the compare_sequences function from cell z3HeD2V8epVW\n",
        "# Need to ensure compare_sequences is defined in the environment or copied here\n",
        "# Assuming it's defined globally or in a preceding cell that has been run\n",
        "\n",
        "# To use compare_sequences with protein sequences, we need to adapt it slightly\n",
        "# or call it with appropriate parameters. The original compare_sequences is designed\n",
        "# for DNA with spaces every 10 or codons every 3. For protein, we might just want\n",
        "# to show the sequences and highlight differences.\n",
        "# Let's create a simple comparison display for protein sequences\n",
        "def codon_distance(codon1, codon2):\n",
        "    \"\"\"\n",
        "    Calculate the Hamming distance between two codons.\n",
        "    \"\"\"\n",
        "    return sum(c1 != c2 for c1, c2 in zip(codon1, codon2))\n",
        "\n",
        "def codon_optimize(ori_dna_seq, dna_seq):\n",
        "    num_codons = len(dna_seq) // 3\n",
        "    opti_mutated_dna_seq_list = list(dna_seq)\n",
        "    for pos in range(num_codons):\n",
        "            # Look for synonymous codons that score higher\n",
        "            mutated_codon = \"\".join(opti_mutated_dna_seq_list[pos*3:pos*3+3]) # Get codon from list\n",
        "            original_codon = ori_dna_seq[pos*3:pos*3+3] # Original codon (string)\n",
        "\n",
        "            # Handle potential issues if codon not in table (though unlikely with A,T,C,G input)\n",
        "            if mutated_codon not in codon_table:\n",
        "                continue\n",
        "\n",
        "            mutated_aa = codon_table[mutated_codon]\n",
        "            d = 0\n",
        "            best_codon = mutated_codon\n",
        "            for codon, aa in codon_table.items():\n",
        "                if aa == mutated_aa:\n",
        "                    d_c = codon_distance(codon, mutated_codon)\n",
        "                    if d_c > d:\n",
        "                        d = d_c\n",
        "                        best_codon = codon\n",
        "\n",
        "            # Apply the best codon (if a better one exists) to the optimized list\n",
        "            if best_codon != mutated_codon:\n",
        "                opti_mutated_dna_seq_list[pos*3:pos*3+3] = list(best_codon)\n",
        "\n",
        "    return \"\".join(opti_mutated_dna_seq_list) # Return string\n",
        "\n",
        "def compare_protein_sequences_display(seq1, seq2):\n",
        "    \"\"\"\n",
        "    Display two protein sequences, highlight differences in bold red.\n",
        "    \"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        print(\"Warning: Sequences have different lengths and cannot be compared directly.\")\n",
        "        print(f\"Original: {seq1}\")\n",
        "        print(f\"Best:     {seq2}\")\n",
        "        return\n",
        "\n",
        "    highlighted_seq1 = []\n",
        "    highlighted_seq2 = []\n",
        "    for i in range(len(seq1)):\n",
        "        if seq1[i] != seq2[i]:\n",
        "            highlighted_seq1.append(seq1[i])\n",
        "            highlighted_seq2.append(colored_text(seq2[i], 'red'))\n",
        "        else:\n",
        "            highlighted_seq1.append(seq1[i])\n",
        "            highlighted_seq2.append(seq2[i])\n",
        "\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>Original Protein: {\"\".join(highlighted_seq1)}</pre>\"))\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>Best Protein:     {\"\".join(highlighted_seq2)}</pre>\"))\n",
        "\n",
        "# Add a function to compare DNA sequences with codon highlighting\n",
        "def compare_dna_sequences_display(seq1, seq2, codon_table):\n",
        "    \"\"\"\n",
        "    Display two DNA sequences, highlight different codons in bold red.\n",
        "    Assumes sequences are divisible by 3.\n",
        "    \"\"\"\n",
        "    if len(seq1) != len(seq2) or len(seq1) % 3 != 0:\n",
        "        print(\"Warning: DNA sequences must be the same length and divisible by 3.\")\n",
        "        return\n",
        "\n",
        "    highlighted_seq1 = []\n",
        "    highlighted_seq2 = []\n",
        "    for i in range(0, len(seq1), 3):\n",
        "        codon1 = seq1[i:i+3]\n",
        "        codon2 = seq2[i:i+3]\n",
        "        if codon1 != codon2:\n",
        "            highlighted_seq1.append(codon1)\n",
        "            highlighted_seq2.append(colored_text(codon2, 'red'))\n",
        "        else:\n",
        "            highlighted_seq1.append(codon1)\n",
        "            highlighted_seq2.append(codon2)\n",
        "        if i + 3 < len(seq1):\n",
        "            highlighted_seq1.append(\" \")\n",
        "            highlighted_seq2.append(\" \")\n",
        "\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>Original DNA: {\"\".join(highlighted_seq1)}</pre>\"))\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>Best DNA:     {\"\".join(highlighted_seq2)}</pre>\"))\n",
        "\n",
        "\n",
        "compare_protein_sequences_display(original_protein_seq, best_protein_sequence)\n",
        "compare_dna_sequences_display(original_seq.replace(\" \", \"\"), best_dna_sequence, codon_table) # Pass codon_table\n",
        "\n",
        "opti_best_dna = codon_optimize(original_seq.replace(\" \", \"\"),best_dna_sequence)\n",
        "_,t_d = translate_dna(best_dna_sequence)\n",
        "_,t_o = translate_dna(opti_best_dna)\n",
        "if t_d != t_o:\n",
        "    print(\"bad optimization:\")\n",
        "    print(t_d)\n",
        "    print(t_o)\n",
        "else:\n",
        "    print(\"\\nOptimized Best DNA Sequence:\")\n",
        "    compare_dna_sequences_display(original_seq.replace(\" \", \"\"), opti_best_dna, codon_table)\n",
        "    pass\n",
        "\n",
        "print(f\"\\n(Protein identity is {protein_identity(original_seq.replace(' ', ''), best_dna_sequence):.1f}%)\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "But this  hardly tells the whole story. We can see from the history that many of the mutations result in poor scores - a result of a single mutation destroying the fold. So lets see this graphically.  "
      ],
      "metadata": {
        "id": "lqVO8Sk2X60O"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "b5aadf40"
      },
      "source": [
        "#@title Plot sequence history\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "# Assuming the history dictionary is loaded from the previous cell\n",
        "# history = np.load(history_filename.replace('.pkl', '.npy'), allow_pickle=True).item()\n",
        "\n",
        "all_iterations = history['all_iterations']\n",
        "all_mutation_scores = history['all_mutation_scores']\n",
        "all_sim_scores = history['all_sim_scores']\n",
        "all_total_scores = history['all_total_scores']\n",
        "\n",
        "accepted_iterations = history['accepted_iterations']\n",
        "accepted_mutation_scores = history['accepted_mutation_scores']\n",
        "accepted_sim_scores = history['accepted_sim_scores']\n",
        "accepted_total_scores = history['accepted_total_scores']\n",
        "\n",
        "# Normalize scores for the time series plots\n",
        "# Find the max values across both all and accepted scores to normalize consistently\n",
        "max_mutation_score = max(max(all_mutation_scores) if all_mutation_scores else 0,\n",
        "                         max(accepted_mutation_scores) if accepted_mutation_scores else 0)\n",
        "max_sim_score = max(max(all_sim_scores) if all_sim_scores else 0,\n",
        "                    max(accepted_sim_scores) if accepted_sim_scores else 0)\n",
        "max_total_score = max(max(all_total_scores) if all_total_scores else 0,\n",
        "                      max(accepted_total_scores) if accepted_total_scores else 0)\n",
        "\n",
        "\n",
        "normalized_all_mutation_scores = [score / max_mutation_score if max_mutation_score > 0 else 0 for score in all_mutation_scores]\n",
        "normalized_all_sim_scores = [score / max_sim_score if max_sim_score > 0 else 0 for score in all_sim_scores]\n",
        "normalized_all_total_scores = [score / max_total_score if max_total_score > 0 else 0 for score in all_total_scores]\n",
        "\n",
        "normalized_accepted_mutation_scores = [score / max_mutation_score if max_mutation_score > 0 else 0 for score in accepted_mutation_scores]\n",
        "normalized_accepted_sim_scores = [score / max_sim_score if max_sim_score > 0 else 0 for score in accepted_sim_scores]\n",
        "normalized_accepted_total_scores = [score / max_total_score if max_total_score > 0 else 0 for score in accepted_total_scores]\n",
        "\n",
        "\n",
        "fig, axes = plt.subplots(2, 2, figsize=(12, 10))\n",
        "fig.suptitle('Mutation Evolution Analysis', fontsize=16)\n",
        "\n",
        "# Subplot 1: All sequences - Mutation vs Similarity Scatter\n",
        "axes[0, 0].scatter(all_mutation_scores, all_sim_scores, alpha=0.5, label='All Mutations')\n",
        "axes[0, 0].plot(all_mutation_scores, all_sim_scores, alpha=0.2, linestyle='-', linewidth=1)\n",
        "axes[0, 0].set_xlabel('Mutation Score')\n",
        "axes[0, 0].set_ylabel('Similarity Score')\n",
        "axes[0, 0].set_title('All Sequences: Mutation vs Similarity')\n",
        "axes[0, 0].grid(True)\n",
        "\n",
        "# Subplot 2: Accepted sequences - Mutation vs Similarity Scatter\n",
        "axes[0, 1].scatter(accepted_mutation_scores, accepted_sim_scores, color='green', alpha=0.7, label='Accepted Mutations')\n",
        "axes[0, 1].plot(accepted_mutation_scores, accepted_sim_scores, color='green', alpha=0.4, linestyle='-', linewidth=1)\n",
        "axes[0, 1].set_xlabel('Mutation Score')\n",
        "axes[0, 1].set_ylabel('Similarity Score')\n",
        "axes[0, 1].set_title('Accepted Sequences: Mutation vs Similarity')\n",
        "axes[0, 1].grid(True)\n",
        "\n",
        "# Subplot 3: All sequences - Scores over Time (Normalized)\n",
        "axes[1, 0].plot(all_iterations, normalized_all_mutation_scores, label='Normalized Mutation Score')\n",
        "axes[1, 0].plot(all_iterations, normalized_all_sim_scores, label='Normalized Similarity Score')\n",
        "axes[1, 0].plot(all_iterations, normalized_all_total_scores, label='Normalized Total Score')\n",
        "axes[1, 0].set_xlabel('Iteration')\n",
        "axes[1, 0].set_ylabel('Normalized Score')\n",
        "axes[1, 0].set_title('All Sequences: Normalized Scores over Time')\n",
        "axes[1, 0].legend()\n",
        "axes[1, 0].grid(True)\n",
        "\n",
        "# Subplot 4: Accepted sequences - Scores over Time (Normalized)\n",
        "axes[1, 1].plot(accepted_iterations, normalized_accepted_mutation_scores, label='Normalized Mutation Score')\n",
        "axes[1, 1].plot(accepted_iterations, normalized_accepted_sim_scores, label='Normalized Similarity Score')\n",
        "axes[1, 1].plot(accepted_iterations, normalized_accepted_total_scores, label='Normalized Total Score')\n",
        "axes[1, 1].set_xlabel('Iteration')\n",
        "axes[1, 1].set_ylabel('Normalized Score')\n",
        "axes[1, 1].set_title('Accepted Sequences: Normalized Scores over Time')\n",
        "axes[1, 1].legend()\n",
        "axes[1, 1].grid(True)\n",
        "\n",
        "plt.tight_layout(rect=[0, 0.03, 1, 0.95]) # Adjust layout to prevent title overlap\n",
        "plt.show()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Left**: All steps, including rejected mutations\n",
        "\n",
        "**Right**: Only accepted steps\n",
        "\n",
        "Looking at the similarity score, many are very poor, indicative of a destroyed fold. And since we reject these, this procedure actually avoids ever going to far away from \"good sequences\".\n",
        "\n",
        "**Conclusion**:\n",
        "> For a random mutation, we can expect that the fold is quite likely destroyed."
      ],
      "metadata": {
        "id": "hrHYP8fUZMiH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## What mutations are ok then?\n",
        "\n",
        "Why don't we plot all good mutations and have a look? Try to see if there is reason in the mutations that didn't completely ruin the fold.\n",
        "\n",
        "**NOTE**: For your short run, statistics will be very poor, and the mutation score dominates, so this will likey not make a whole lot of sense until you use the longer precomputed evolution."
      ],
      "metadata": {
        "id": "jLBXwzV2ZcCk"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "67401602"
      },
      "source": [
        "#@title Plot all accepted mutations\n",
        "\n",
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "from collections import defaultdict\n",
        "\n",
        "sim_threshold = 0.1 #@param\n",
        "\n",
        "# Assuming the history dictionary is loaded from the previous cell\n",
        "# history = np.load(history_filename.replace('.pkl', '.npy'), allow_pickle=True).item()\n",
        "\n",
        "all_iterations = history['all_iterations']\n",
        "all_sim_scores = history['all_sim_scores']\n",
        "all_tried_protein_sequences = history['all_tried_protein_sequences'] # These are already protein sequences\n",
        "\n",
        "# Filter for mutations with similarity score >= 0.8\n",
        "high_similarity_mutations = []\n",
        "# Iterate through all tried sequences starting from the second one (index 1)\n",
        "for i in range(1, len(all_tried_protein_sequences)):\n",
        "    iteration = all_iterations[i-1] # Iteration numbers correspond to the mutation *leading to* the current sequence\n",
        "    previous_protein_sequence = all_tried_protein_sequences[i-1]\n",
        "    current_protein_sequence = all_tried_protein_sequences[i]\n",
        "    similarity_score = all_sim_scores[i-1] # Similarity score for the mutation at this iteration\n",
        "\n",
        "    if similarity_score >= sim_threshold:\n",
        "        # Find the mutated position and amino acids\n",
        "        mutated_pos = -1\n",
        "        original_aa = None\n",
        "        mutated_aa = None\n",
        "        for j in range(len(previous_protein_sequence)):\n",
        "            if previous_protein_sequence[j] != current_protein_sequence[j]:\n",
        "                mutated_pos = j\n",
        "                original_aa = previous_protein_sequence[j]\n",
        "                mutated_aa = current_protein_sequence[j]\n",
        "                break # Assuming only one mutation per step\n",
        "\n",
        "        if mutated_pos != -1:\n",
        "            high_similarity_mutations.append({'original_aa': original_aa,\n",
        "                                             'mutated_aa': mutated_aa,\n",
        "                                             'position': mutated_pos,\n",
        "                                             'similarity_score': similarity_score})\n",
        "\n",
        "# Count the frequency of each substitution\n",
        "substitution_counts = defaultdict(int)\n",
        "for mutation in high_similarity_mutations:\n",
        "    substitution_counts[(mutation['original_aa'], mutation['mutated_aa'])] += 1\n",
        "\n",
        "# Create a matrix (DataFrame) for visualization\n",
        "# Get all unique amino acids involved in these substitutions\n",
        "all_aas = sorted(list(set([m['original_aa'] for m in high_similarity_mutations] +\n",
        "                         [m['mutated_aa'] for m in high_similarity_mutations])))\n",
        "\n",
        "substitution_matrix = pd.DataFrame(0, index=all_aas, columns=all_aas)\n",
        "\n",
        "for (orig, mut), count in substitution_counts.items():\n",
        "    substitution_matrix.loc[orig, mut] = count\n",
        "\n",
        "# Visualize the matrix\n",
        "plt.figure(figsize=(10, 8))\n",
        "sns.heatmap(substitution_matrix, cmap=\"viridis\", annot=True, fmt=\"d\", linewidths=.5)\n",
        "plt.xlabel(\"Mutated Amino Acid\")\n",
        "plt.ylabel(\"Original Amino Acid\")\n",
        "plt.title(f\"Amino Acid Substitutions with Similarity Score >= {sim_threshold}\")\n",
        "plt.show()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "53d927e5"
      },
      "source": [
        "#@title Download sequences above a total score threshold\n",
        "\n",
        "from google.colab import files\n",
        "\n",
        "score_threshold = 40.0 #@param{type:\"number\"}\n",
        "\n",
        "# Filter sequences based on the score threshold\n",
        "high_scoring_sequences = [\n",
        "    seq for i, seq in enumerate(history['all_tried_protein_sequences'][1:]) # Start from index 1 to skip original\n",
        "    if history['all_total_scores'][i] >= score_threshold\n",
        "]\n",
        "\n",
        "# Add the original sequence if its score (or a derived score) meets the threshold\n",
        "# This assumes original_seq's score is available in history or can be calculated\n",
        "# If the original sequence's score is not explicitly in all_total_scores at index 0,\n",
        "# you might need to calculate it or handle it separately.\n",
        "# For simplicity, let's assume the first entry in all_total_scores corresponds to the first tried sequence (index 1 in all_tried_protein_sequences)\n",
        "# and the original sequence is implicitly handled if it leads to a high-scoring mutation.\n",
        "# A more explicit way if needed:\n",
        "# original_score = mutation_score(original_seq, original_seq) * rmsd_to_similarity(align_pred_to_ref(pdb_str, \"reference.pdb\")[1]) # Example calculation\n",
        "# if original_score >= score_threshold and history['all_tried_protein_sequences'][0] not in high_scoring_sequences:\n",
        "#    high_scoring_sequences.insert(0, history['all_tried_protein_sequences'][0])\n",
        "\n",
        "# Add the best protein sequence if it's not already included (it should be if its score > threshold)\n",
        "best_protein_sequence = history['best_protein_sequence']\n",
        "if best_protein_sequence not in high_scoring_sequences and history['best_total_score'] >= score_threshold:\n",
        "    high_scoring_sequences.append(best_protein_sequence)\n",
        "\n",
        "\n",
        "# Create the FASTA file content\n",
        "fasta_content = \"\"\n",
        "for i, seq in enumerate(high_scoring_sequences):\n",
        "    fasta_content += f\">Sequence_{i+1}\\n{seq}\\n\"\n",
        "\n",
        "# Define the filename\n",
        "fasta_filename = \"high_scoring_sequences.fasta\"\n",
        "\n",
        "# Write the FASTA content to a file\n",
        "with open(fasta_filename, \"w\") as f:\n",
        "    f.write(fasta_content)\n",
        "\n",
        "# Download the file\n",
        "files.download(fasta_filename)\n",
        "\n",
        "print(f\"Created and downloaded {fasta_filename} containing {len(high_scoring_sequences)} sequences with total score >= {score_threshold}.\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "359edac2"
      },
      "source": [
        "#@title Download history files\n",
        "\n",
        "from google.colab import files\n",
        "import os\n",
        "\n",
        "# Define the history filenames\n",
        "history_pkl_filename = \"mutation_history_5.pkl\"\n",
        "history_npy_filename = \"mutation_history_5.npy\" # Assuming you saved the .npy as well\n",
        "\n",
        "# Download the files if they exist\n",
        "if os.path.exists(history_pkl_filename):\n",
        "    files.download(history_pkl_filename)\n",
        "    print(f\"Downloaded {history_pkl_filename}\")\n",
        "else:\n",
        "    print(f\"{history_pkl_filename} not found.\")\n",
        "\n",
        "if os.path.exists(history_npy_filename):\n",
        "    files.download(history_npy_filename)\n",
        "    print(f\"Downloaded {history_npy_filename}\")\n",
        "else:\n",
        "    print(f\"{history_npy_filename} not found.\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4828a5d2"
      },
      "source": [
        "## Identify top sequences\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "0b70fb72"
      },
      "source": [
        "#@title Top scores\n",
        "all_total_scores = history['all_total_scores']\n",
        "all_tried_protein_sequences = history['all_tried_protein_sequences']\n",
        "all_iterations = history['all_iterations'] # Keep track of iterations to help potentially link back\n",
        "\n",
        "# Get indices of all scores sorted in descending order\n",
        "sorted_indices = np.argsort(all_total_scores)[::-1]\n",
        "\n",
        "# Select the top 20 indices\n",
        "top_20_indices = sorted_indices[:20]\n",
        "\n",
        "# Retrieve the top 20 protein sequences\n",
        "top_20_protein_sequences = [all_tried_protein_sequences[i] for i in top_20_indices]\n",
        "\n",
        "print(\"Top 20 Protein Sequences:\")\n",
        "for i, seq in enumerate(top_20_protein_sequences):\n",
        "    print(f\"{i+1}: {seq} (Score: {all_total_scores[top_20_indices[i]]:.3f})\")\n",
        "\n",
        "# Since the DNA sequences were not explicitly stored for every tried protein sequence in the history,\n",
        "# and reconstructing them based solely on the protein sequence and original_seq is not straightforward\n",
        "# due to silent mutations and the iterative nature, we cannot definitively link the top 20 protein\n",
        "# sequences back to their exact DNA sequences from the current history structure.\n",
        "# To achieve the next step of the main task (position-dependent amino acid distributions),\n",
        "# we need the protein sequences themselves, which we have.\n",
        "# For the subtask of finding the *DNA* sequences, we cannot complete it with the current history data structure.\n",
        "# We will proceed with the protein sequences for the main task and report failure for the subtask as defined."
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Lets see what maximum score we could expect for our protein under the current scoring, just to know how good we got."
      ],
      "metadata": {
        "id": "-BW6tVE4o_GP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Random mutation of original sequence\n",
        "import random\n",
        "import numpy as np # Import numpy\n",
        "\n",
        "def codon_distance(codon1, codon2):\n",
        "    \"\"\"\n",
        "    Calculate the Hamming distance between two codons.\n",
        "    \"\"\"\n",
        "    return sum(c1 != c2 for c1, c2 in zip(codon1, codon2))\n",
        "\n",
        "def introduce_random_mutations(original_dna_sequence: str, num_mutations: int) -> str:\n",
        "    \"\"\"\n",
        "    Introduces a specified number of random mutations into a DNA sequence.\n",
        "\n",
        "    Args:\n",
        "        original_dna_sequence: The original DNA sequence string.\n",
        "        num_mutations: The number of random mutations to introduce.\n",
        "\n",
        "    Returns:\n",
        "        A new DNA sequence string with the specified number of mutations.\n",
        "    \"\"\"\n",
        "    dna_list = list(original_dna_sequence)\n",
        "    bases = ['A', 'T', 'C', 'G']\n",
        "    seq_len = len(dna_list)\n",
        "\n",
        "    for _ in range(num_mutations):\n",
        "        # Randomly select an index\n",
        "        mutation_index = random.randint(0, seq_len - 1)\n",
        "\n",
        "        # Randomly select a new base different from the current one\n",
        "        current_base = dna_list[mutation_index]\n",
        "        new_base = random.choice(bases)\n",
        "        while new_base == current_base:\n",
        "            new_base = random.choice(bases)\n",
        "\n",
        "        # Apply the mutation to the list\n",
        "        dna_list[mutation_index] = new_base\n",
        "\n",
        "    # Convert the list back to a string and return\n",
        "    return \"\".join(dna_list)\n",
        "\n",
        "# 2. Initialize empty lists to store results\n",
        "protein_identities = []\n",
        "mutation_scores = []\n",
        "opti_mutation_scores = []\n",
        "\n",
        "repeats = 5 #@param{type:\"integer\"}\n",
        "\n",
        "for iter in range(repeats):\n",
        "\n",
        "    # 1. Define the original DNA sequence\n",
        "    original_dna_seq = original_seq\n",
        "\n",
        "    # Clean input\n",
        "    original_dna_seq = original_dna_seq.replace(\" \", \"\").upper()\n",
        "\n",
        "    # Define a range of mutation counts\n",
        "    # Let's try up to 100 mutations, as the sequence length is 96\n",
        "    max_mutations = 150 #@param{type:\"integer\"}\n",
        "\n",
        "    # 3. Create a loop to iterate through mutation counts\n",
        "    for num_mutations in range(max_mutations + 1):\n",
        "        # 4. Generate a mutated DNA sequence\n",
        "        mutated_dna_seq_list = list(introduce_random_mutations(original_dna_seq, num_mutations)) # Work with a list\n",
        "\n",
        "        # 5. Calculate the protein identity\n",
        "        # Need to convert the list back to a string for translation\n",
        "        mutated_dna_seq_str = \"\".join(mutated_dna_seq_list)\n",
        "        prot_id = protein_identity(original_dna_seq, mutated_dna_seq_str)\n",
        "        protein_identities.append(prot_id)\n",
        "\n",
        "        # 6. Calculate the mutation score\n",
        "        mut_score = mutation_score(original_dna_seq, mutated_dna_seq_str)\n",
        "\n",
        "        # Optimize mutation score by looking for synonymous codons\n",
        "        opti_mutated_dna_seq_list = list(mutated_dna_seq_list) # Create a copy to optimize\n",
        "        num_codons = len(original_dna_seq) // 3 # Ensure num_codons is defined here\n",
        "        for pos in range(num_codons):\n",
        "            # Look for synonymous codons that score higher\n",
        "            mutated_codon = \"\".join(opti_mutated_dna_seq_list[pos*3:pos*3+3]) # Get codon from list\n",
        "            original_codon = original_dna_seq[pos*3:pos*3+3] # Original codon (string)\n",
        "\n",
        "            # Handle potential issues if codon not in table (though unlikely with A,T,C,G input)\n",
        "            if mutated_codon not in codon_table:\n",
        "                continue\n",
        "\n",
        "            mutated_aa = codon_table[mutated_codon]\n",
        "            d = 0\n",
        "            best_codon = mutated_codon\n",
        "            for codon, aa in codon_table.items():\n",
        "                if aa == mutated_aa:\n",
        "                    d_c = codon_distance(codon, mutated_codon)\n",
        "                    if d_c > d:\n",
        "                        d = d_c\n",
        "                        best_codon = codon\n",
        "\n",
        "            # Apply the best codon (if a better one exists) to the optimized list\n",
        "            if best_codon != mutated_codon:\n",
        "                opti_mutated_dna_seq_list[pos*3:pos*3+3] = list(best_codon)\n",
        "\n",
        "\n",
        "        opti_mut_score = mutation_score(original_dna_seq, \"\".join(opti_mutated_dna_seq_list)) # Calculate score from optimized list\n",
        "\n",
        "\n",
        "        mutation_scores.append(mut_score)\n",
        "        opti_mutation_scores.append(opti_mut_score)\n",
        "\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Create the scatter plot\n",
        "plt.figure(figsize=(10, 6))\n",
        "plt.scatter(protein_identities, mutation_scores, label='Original Mutation Score', alpha=0.5)\n",
        "plt.scatter(protein_identities, opti_mutation_scores, label='Optimized Mutation Score (Synonymous Codons)', alpha=0.5)\n",
        "\n",
        "# Add title and labels\n",
        "plt.title(\"Mutation Score vs. Protein Identity\")\n",
        "plt.xlabel(\"Protein Identity (%)\")\n",
        "plt.ylabel(\"Mutation Score\")\n",
        "\n",
        "# Add grid and legend\n",
        "plt.grid(True)\n",
        "plt.legend()\n",
        "\n",
        "# Display the plot\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "pXzzKfJ7pHk8",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Now lets look at a longer evolution\n",
        "\n",
        "It's great that you can do your own evolution and get how its done - but it is quite slow. It will take you at least a couple of hours to get a substantially differnt sequence, so we have made such an evolution for you. Go back to the evolution iteration cell and tick the \"ues_precomputed\" option, and work your way through the analysis down to this point again. **After that**, you can continue with the sequence similarity analysis."
      ],
      "metadata": {
        "id": "_i1YOxCbc4nZ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Sequnece similarity analysis\n",
        "The longer evolution resulted in a sequence that may exist in nature - who knows, right? Well let's check. The followin analysis installs and runs analysis of your mutated sequence against some 3000 protein sequences that have the same fold, found across nature. They are all quite different, so maybe our mutates sequence or something close to it is in there?"
      ],
      "metadata": {
        "id": "eHKG9992do6d"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "D1CkYv7wE8lB",
        "collapsed": true
      },
      "source": [
        "%pip install biopython\n",
        "!apt-get update -y\n",
        "!apt-get install -y clustalw"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "A .fasta file is a typical format for storing many sequences (DNA or protein). Here we load 3000 sequences that we found using a tool called foldseek. It's a pretty cool tool, but we won't go into details on it here. You will simply use them to contextualize your original and mutates sequence."
      ],
      "metadata": {
        "id": "Ee6s6auBex5F"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "fe32dc38",
        "cellView": "form"
      },
      "source": [
        "#@title Load fasta from URL and save as template\n",
        "import requests\n",
        "import io\n",
        "from Bio import SeqIO\n",
        "\n",
        "# Define the URL of the FASTA file and the local filename\n",
        "url = \"https://raw.githubusercontent.com/bforsbe/SK2534/main/aligned_output.fasta\"\n",
        "local_fasta_file = \"template_aligned.fasta\"\n",
        "\n",
        "# Download the content of the file and save it locally if it doesn't exist\n",
        "try:\n",
        "    with open(local_fasta_file, \"r\") as f:\n",
        "        print(f\"Using existing template file: {local_fasta_file}\")\n",
        "except FileNotFoundError:\n",
        "    print(f\"Downloading template file from: {url}\")\n",
        "    response = requests.get(url)\n",
        "    response.raise_for_status() # Raise an exception for bad status codes\n",
        "    with open(local_fasta_file, \"w\") as f:\n",
        "        f.write(response.text)\n",
        "    print(f\"Template file downloaded and saved as: {local_fasta_file}\")\n",
        "\n",
        "# Read the sequences from the local FASTA file using SeqIO.parse\n",
        "sequences = list(SeqIO.parse(local_fasta_file, \"fasta\"))\n",
        "\n",
        "# Print the number of sequences loaded to verify\n",
        "print(f\"Number of sequences loaded from template: {len(sequences)}\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Ok, lets add the sequences we are interested in."
      ],
      "metadata": {
        "id": "rZ0NYybQgWXA"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "e1026ad4",
        "cellView": "form"
      },
      "source": [
        "#@title Add original and query sequence\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "\n",
        "# Define the original and query sequence strings\n",
        "original_sequence = \"YKCGLCERSFVEKSALSRHQRVHKN\" #@param {type:\"string\"}\n",
        "mutated_sequence = \"< your mutated sequence here >\" #@param {type:\"string\"}\n",
        "\n",
        "original_sequence_string = original_sequence\n",
        "query_sequence_string = mutated_sequence\n",
        "# Create SeqRecord objects for the original and query sequences with fixed IDs\n",
        "original_sequence = SeqRecord(Seq(original_sequence_string), id=\"Original_Sequence\", description=\"\")\n",
        "query_sequence = SeqRecord(Seq(query_sequence_string), id=\"Query_Sequence\", description=\"\")\n",
        "\n",
        "# Create a new list of sequences including the original sequences from the template, the original, and the query\n",
        "sequences_with_original_and_query = sequences + [original_sequence, query_sequence]\n",
        "\n",
        "# Print the updated number of sequences to verify\n",
        "print(f\"Updated number of sequences (including original and mutated): {len(sequences_with_original_and_query)}\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "44bfd6b2"
      },
      "source": [
        "#@title Prepare sequences for alignment\n",
        "# Modify sequence IDs to be unique and clear descriptions for the list including the query and original\n",
        "for i, seq in enumerate(sequences_with_original_and_query):\n",
        "    # Only modify IDs if they are not the original or query sequences\n",
        "    if seq.id != \"Original_Sequence\" and seq.id != \"Query_Sequence\":\n",
        "        seq.id = f\"{seq.id}_{i}\"\n",
        "    seq.description = \"\" # Clear description to avoid issues with some tools\n",
        "\n",
        "# Define a filename for the temporary FASTA file\n",
        "temp_in_file = \"temp_sequences_for_alignment.fasta\"\n",
        "\n",
        "# Write the modified sequences to the temporary FASTA file\n",
        "SeqIO.write(sequences_with_original_and_query, temp_in_file, \"fasta\")\n",
        "\n",
        "# Print confirmation message\n",
        "print(f\"Prepared sequences (including original and query) and saved to {temp_in_file}\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "52bfb8a7"
      },
      "source": [
        "## Perform multiple sequence alignment\n",
        "\n",
        "Use ClustalW to align the sequences from the temporary file and save the output to a new file.\n",
        "\n",
        "Use Biopython's ClustalwCommandline to perform multiple sequence alignment on the sequences in the temporary FASTA file and save the output to a new FASTA file.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "ea50dc4e"
      },
      "source": [
        "#@title Make MSA\n",
        "from Bio import AlignIO\n",
        "from Bio import Align\n",
        "from Bio.Align.Applications import ClustalwCommandline\n",
        "\n",
        "# Define the name for the output alignment file\n",
        "out_file_aln = \"aligned_with_query.aln\"\n",
        "out_file_fasta = \"aligned_with_query.fasta\"\n",
        "\n",
        "# Create a ClustalwCommandline object\n",
        "# Specify the path to the clustalw executable if it's not in your PATH\n",
        "clustalw_cline = ClustalwCommandline(\"clustalw\", infile=temp_in_file, outfile=out_file_aln)\n",
        "\n",
        "# Run the command line tool\n",
        "stdout, stderr = clustalw_cline()\n",
        "\n",
        "# Print confirmation message\n",
        "print(f\"Multiple sequence alignment performed and saved to {out_file_aln}\")\n",
        "\n",
        "# Convert aln -> fasta using AlignIO (still needed for writing)\n",
        "alignment_aln = AlignIO.read(out_file_aln, \"clustal\")\n",
        "AlignIO.write(alignment_aln, out_file_fasta, \"fasta\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f087f25f"
      },
      "source": [
        "## Read the alignment\n",
        "Read the aligned sequences from the output file using `AlignIO.read` (using AlignIO.read with the 'fasta' format and print the alignment object.)\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "0b805761"
      },
      "source": [
        "#@title Read alignment file\n",
        "from Bio import AlignIO\n",
        "\n",
        "# Define the name of the alignment file output by ClustalW\n",
        "alignment_file = out_file_aln\n",
        "\n",
        "# Read the alignment file using AlignIO.read with 'clustal' format\n",
        "try:\n",
        "    alignment = AlignIO.read(alignment_file, \"clustal\")\n",
        "\n",
        "    # Print the alignment object to verify\n",
        "    print(\"Alignment object:\")\n",
        "    print(alignment)\n",
        "\n",
        "except FileNotFoundError:\n",
        "    print(f\"Error: The file {alignment_file} was not found.\")\n",
        "except Exception as e:\n",
        "    print(f\"An error occurred while reading the file: {e}\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "04ed3f2c"
      },
      "source": [
        "## Analyze sequence identity\n",
        "\n",
        "Wow, that was a lot of prep work. Finally, we can e.g. calculate the maximum sequence identity of the query sequence against all other sequences in the alignment.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "23b59b4c"
      },
      "source": [
        "**Reasoning**:\n",
        "Calculate the maximum sequence identity by iterating through the alignment, comparing the query sequence to every other sequence, and keeping track of the highest identity score found.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "98c24d08"
      },
      "source": [
        "#@title Analyze sequence identity and display best match alignment\n",
        "# Find the query and original sequences in the alignment using their fixed IDs\n",
        "query_sequence_id = \"Query_Sequence\"\n",
        "original_sequence_id = \"Original_Sequence\"\n",
        "query_seq = None\n",
        "original_seq = None\n",
        "\n",
        "for record in alignment:\n",
        "    if record.id == query_sequence_id:\n",
        "        query_seq = record\n",
        "    elif record.id == original_sequence_id:\n",
        "        original_seq = record\n",
        "\n",
        "if query_seq is None:\n",
        "    print(f\"Error: Query sequence with ID '{query_sequence_id}' not found in the alignment.\")\n",
        "else:\n",
        "    max_identity = 0.0\n",
        "    best_match_id = None\n",
        "    best_match_seq = None\n",
        "\n",
        "    # Iterate through each sequence in the alignment\n",
        "    for record in alignment:\n",
        "        # Skip the query and original sequences themselves when finding best match\n",
        "        if record.id == query_sequence_id or record.id == original_sequence_id:\n",
        "            continue\n",
        "\n",
        "        # Calculate sequence identity\n",
        "        identical_residues = 0\n",
        "        aligned_length = 0\n",
        "        for i in range(alignment.get_alignment_length()):\n",
        "            query_residue = query_seq[i]\n",
        "            current_residue = record[i]\n",
        "\n",
        "            # Only consider positions where neither sequence has a gap\n",
        "            if query_residue != '-' and current_residue != '-':\n",
        "                aligned_length += 1\n",
        "                if query_residue == current_residue:\n",
        "                    identical_residues += 1\n",
        "\n",
        "        # Avoid division by zero if there are no aligned positions\n",
        "        if aligned_length > 0:\n",
        "            identity = (identical_residues / aligned_length) * 100\n",
        "            if identity > max_identity:\n",
        "                max_identity = identity\n",
        "                best_match_id = record.id\n",
        "                best_match_seq = record\n",
        "        else:\n",
        "            identity = 0.0 # Or handle as appropriate if no comparable positions\n",
        "\n",
        "    print(f\"Maximum sequence identity of the query sequence against others (excluding original): {max_identity:.2f}%\")\n",
        "    if best_match_id:\n",
        "        print(f\"Sequence with the highest identity: {best_match_id}\")\n",
        "\n",
        "        # Display the alignment of the query, original, and the best match\n",
        "        print(\"\\nAlignment of Query Sequence, Original Sequence, and Best Match:\")\n",
        "        # Determine the maximum length of the sequence IDs for alignment\n",
        "        max_id_length = max(len(query_seq.id), len(original_seq.id) if original_seq else 0, len(best_match_id))\n",
        "        if original_seq:\n",
        "             print(f\"{original_seq.seq} {original_seq.id.ljust(max_id_length)}\")\n",
        "        print(f\"{query_seq.seq} {query_seq.id.ljust(max_id_length)}\")\n",
        "        print(f\"{best_match_seq.seq} {best_match_id.ljust(max_id_length)}\")"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can search the internet for the identifier of the closest match to the mutated sequence. Sometimes this might be hard, in which case you can always use protein BLAST to search using the protein sequence itself. The original sequence was a human zinc-binding domain, where did you \"end up\"?"
      ],
      "metadata": {
        "id": "rniy3xCsheEN"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dd5b85e6"
      },
      "source": [
        "## Analyze conservation\n",
        "\n",
        "### Subtask:\n",
        "Analyze the conservation of residues in the query sequence compared to the original sequence, and compare this to the overall conservation at those positions in the multiple sequence alignment."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "df3954fd"
      },
      "source": [
        "**Reasoning**:\n",
        "Calculate the conservation score for each position in the overall alignment. Then, for each position in the query sequence, compare the residue to the corresponding residue in the original sequence and the conservation score at that position in the overall alignment. Print the results to show where the query mutation aligns with highly conserved or variable positions."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "cellView": "form",
        "id": "2d49c979"
      },
      "source": [
        "#@title Analyze conservation of query sequence mutations\n",
        "# Function to calculate conservation at a given column in the alignment\n",
        "def calculate_conservation(column):\n",
        "    \"\"\"Calculates the conservation score for a column in a multiple sequence alignment.\"\"\"\n",
        "    # Count the occurrences of each residue (excluding gaps)\n",
        "    residue_counts = {}\n",
        "    total_residues = 0\n",
        "    for residue in column:\n",
        "        if residue != '-':\n",
        "            total_residues += 1\n",
        "            residue_counts[residue] = residue_counts.get(residue, 0) + 1\n",
        "\n",
        "    # Calculate the conservation score (simple frequency of the most common residue)\n",
        "    if total_residues == 0:\n",
        "        return 0.0\n",
        "    max_count = max(residue_counts.values()) if residue_counts else 0\n",
        "    conservation = (max_count / total_residues) * 100\n",
        "    return conservation\n",
        "\n",
        "# Find the query and original sequences in the alignment\n",
        "query_sequence_id = \"Query_Sequence\"\n",
        "original_sequence_id = \"Original_Sequence\"\n",
        "query_seq = None\n",
        "original_seq = None\n",
        "\n",
        "for record in alignment:\n",
        "    if record.id == query_sequence_id:\n",
        "        query_seq = record\n",
        "    elif record.id == original_sequence_id:\n",
        "        original_seq = record\n",
        "\n",
        "if query_seq is None or original_seq is None:\n",
        "    print(\"Error: Query or Original sequence not found in the alignment.\")\n",
        "else:\n",
        "    print(\"Conservation Analysis:\")\n",
        "    print(\"-\" * 60)\n",
        "    print(f\"{'Position':<10} {'Original':<10} {'Query':<10} {'Overall Conservation (%)':<25} {'Observation':<15}\")\n",
        "    print(\"-\" * 60)\n",
        "\n",
        "    # Iterate through the alignment positions\n",
        "    for i in range(alignment.get_alignment_length()):\n",
        "        original_residue = original_seq[i]\n",
        "        query_residue = query_seq[i]\n",
        "\n",
        "        # Skip if both original and query have gaps\n",
        "        if original_residue == '-' and query_residue == '-':\n",
        "            continue\n",
        "\n",
        "        # Calculate conservation at this position in the overall alignment\n",
        "        column = alignment[:, i]\n",
        "        overall_conservation = calculate_conservation(column)\n",
        "\n",
        "        observation = \"\"\n",
        "        if original_residue != '-' and query_residue != '-':\n",
        "            if original_residue != query_residue:\n",
        "                observation = \"Mutation\"\n",
        "                if overall_conservation > 70: # Example threshold for high conservation\n",
        "                    observation += \" (Highly Conserved)\"\n",
        "                elif overall_conservation < 30: # Example threshold for low conservation\n",
        "                    observation += \" (Variable)\"\n",
        "            else:\n",
        "                 observation = \"No Mutation\"\n",
        "\n",
        "\n",
        "        print(f\"{i:<10} {original_residue:<10} {query_residue:<10} {overall_conservation:<25.2f} {observation:<15}\")\n",
        "\n",
        "    print(\"-\" * 60)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "1271d3f9"
      },
      "source": [
        "## Analyze Sequence Identity Distribution\n",
        "\n",
        "### Subtask:\n",
        "Calculate sequence identity against original and query sequences and visualize the distribution and relationship."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6e1fdef1"
      },
      "source": [
        "**Reasoning**:\n",
        "Calculate the sequence identity of each sequence in the alignment against the original and query sequences. Store these identities and then use matplotlib and seaborn to create a histogram of identities against the original and a scatterplot of identities against the original versus identities against the query."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "e1745cea",
        "cellView": "form"
      },
      "source": [
        "#@title Analyze and visualize sequence identity distribution\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import pandas as pd\n",
        "\n",
        "# Find the query and original sequences in the alignment\n",
        "query_sequence_id = \"Query_Sequence\"\n",
        "original_sequence_id = \"Original_Sequence\"\n",
        "query_seq = None\n",
        "original_seq = None\n",
        "\n",
        "for record in alignment:\n",
        "    if record.id == query_sequence_id:\n",
        "        query_seq = record\n",
        "    elif record.id == original_sequence_id:\n",
        "        original_seq = record\n",
        "\n",
        "if query_seq is None or original_seq is None:\n",
        "    print(\"Error: Query or Original sequence not found in the alignment.\")\n",
        "else:\n",
        "    identity_data = []\n",
        "    # Iterate through each sequence in the alignment\n",
        "    for record in alignment:\n",
        "        # Skip the query and original sequences themselves\n",
        "        if record.id == query_sequence_id or record.id == original_sequence_id:\n",
        "            continue\n",
        "\n",
        "        # Calculate sequence identity against original sequence\n",
        "        identical_residues_original = 0\n",
        "        aligned_length_original = 0\n",
        "        for i in range(alignment.get_alignment_length()):\n",
        "            original_residue = original_seq[i]\n",
        "            current_residue = record[i]\n",
        "\n",
        "            if original_residue != '-' and current_residue != '-':\n",
        "                aligned_length_original += 1\n",
        "                if original_residue == current_residue:\n",
        "                    identical_residues_original += 1\n",
        "\n",
        "        identity_original = (identical_residues_original / aligned_length_original) * 100 if aligned_length_original > 0 else 0.0\n",
        "\n",
        "        # Calculate sequence identity against query sequence\n",
        "        identical_residues_query = 0\n",
        "        aligned_length_query = 0\n",
        "        for i in range(alignment.get_alignment_length()):\n",
        "            query_residue = query_seq[i]\n",
        "            current_residue = record[i]\n",
        "\n",
        "            if query_residue != '-' and current_residue != '-':\n",
        "                aligned_length_query += 1\n",
        "                if query_residue == current_residue:\n",
        "                    identical_residues_query += 1\n",
        "\n",
        "        identity_query = (identical_residues_query / aligned_length_query) * 100 if aligned_length_query > 0 else 0.0\n",
        "\n",
        "        identity_data.append({'ID': record.id, 'Identity_vs_Original': identity_original, 'Identity_vs_Query': identity_query})\n",
        "\n",
        "    # Create a pandas DataFrame\n",
        "    identity_df = pd.DataFrame(identity_data)\n",
        "\n",
        "    # Plotting\n",
        "    plt.figure(figsize=(12, 6))\n",
        "\n",
        "    # Histogram of identity against original\n",
        "    plt.subplot(1, 2, 1)\n",
        "    sns.histplot(identity_df['Identity_vs_Original'], bins=20, kde=True)\n",
        "    plt.title('Distribution of Sequence Identity vs. Original Sequence')\n",
        "    plt.xlabel('Sequence Identity (%)')\n",
        "    plt.ylabel('Frequency')\n",
        "\n",
        "    # Scatterplot of identity vs original vs identity vs query\n",
        "    plt.subplot(1, 2, 2)\n",
        "    sns.scatterplot(data=identity_df, x='Identity_vs_Original', y='Identity_vs_Query')\n",
        "    plt.title('Sequence Identity vs. Original vs. Query')\n",
        "    plt.xlabel('Identity vs. Original (%)')\n",
        "    plt.ylabel('Identity vs. Query (%)')\n",
        "\n",
        "    plt.tight_layout()\n",
        "    plt.show()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install logomaker"
      ],
      "metadata": {
        "id": "gQf2MDTqQITC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Create sequence logo\n",
        "from Bio import AlignIO\n",
        "from collections import Counter\n",
        "import pandas as pd\n",
        "import logomaker\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "# Read the alignment (FASTA or ALN)\n",
        "alignment = AlignIO.read(\"aligned_with_query.aln\", \"clustal\")\n",
        "\n",
        "num_seqs = len(alignment)\n",
        "alignment_length = alignment.get_alignment_length()\n",
        "print(\"Number of sequences:\", num_seqs)\n",
        "print(\"Alignment length:\", alignment_length)\n",
        "\n",
        "# Build consensus sequence (preserving all columns)\n",
        "consensus = \"\"\n",
        "degenerate_consensus = \"\"\n",
        "counts_matrix = []  # for logo\n",
        "\n",
        "for col_idx in range(alignment_length):\n",
        "    column = alignment[:, col_idx]             # get column as string\n",
        "    counts = Counter(res for res in column if res != \"-\")  # ignore gaps\n",
        "    counts_matrix.append({aa: counts.get(aa, 0) for aa in \"ACDEFGHIKLMNPQRSTVW\"})\n",
        "\n",
        "    if counts:\n",
        "        # Majority vote for consensus\n",
        "        consensus += counts.most_common(1)[0][0]\n",
        "        # Degenerate: include all residues that appear in column\n",
        "        residues = sorted(counts.keys())\n",
        "        degenerate_consensus += \"[\" + \"\".join(residues) + \"]\" if len(residues) > 1 else residues[0]\n",
        "    else:\n",
        "        # Column is all gaps\n",
        "        consensus += \"-\"\n",
        "        degenerate_consensus += \"-\"\n",
        "\n",
        "print(\"Consensus:\", consensus)\n",
        "print(\"Degenerate consensus:\", degenerate_consensus)\n",
        "\n",
        "# Create a DataFrame for sequence logo\n",
        "df = pd.DataFrame(counts_matrix)\n",
        "df.index.name = \"Position\"\n",
        "\n",
        "# Plot logo\n",
        "plt.figure(figsize=(15,4))\n",
        "logo = logomaker.Logo(df, color_scheme='chemistry')\n",
        "plt.title(\"Sequence Logo\")\n",
        "\n",
        "ticks=True\n",
        "if ticks:\n",
        "    # 6. Add Query_Sequence as tick labels\n",
        "    query_sequence = None\n",
        "    # Access records using the .sequences attribute for modern Bio.Align objects\n",
        "    # Now reload using AlignIO (old API keeps dashes)\n",
        "    msa = AlignIO.read(\"aligned_with_query.fasta\", \"fasta\")\n",
        "\n",
        "    # Extract the gappy sequence\n",
        "    query_sequence = str([rec.seq for rec in msa if rec.id == \"Query_Sequence\"][0])\n",
        "    for record in msa:\n",
        "        if record.id == \"Query_Sequence\":\n",
        "            query_sequence = str(record.seq)\n",
        "            break\n",
        "    print(\"Query Sequence:\", query_sequence)\n",
        "\n",
        "    if query_sequence:\n",
        "        # Set tick positions to correspond to the positions in the sequence\n",
        "        tick_positions = range(len(query_sequence))\n",
        "        # Set tick labels to the characters of the query sequence\n",
        "        plt.gca().set_xticks(tick_positions)\n",
        "        plt.gca().set_xticklabels(list(query_sequence))\n",
        "        # Optional: Rotate labels if they overlap\n",
        "        plt.xticks(rotation=0)\n",
        "        plt.tick_params(axis='x', which='major', labelsize=6) # Adjust label size if needed\n",
        "\n",
        "plt.show()\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "XrhsS-RPQGs-"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}