{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "omegafold.ipynb",
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU",
    "gpuClass": "standard"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/bforsbe/SK2534/blob/main/omegafold.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**OmegaFold**\n",
        "for more details see: [Github](https://github.com/HeliXonProtein/OmegaFold), [Preprint](https://www.biorxiv.org/content/10.1101/2022.07.21.500999v1)\n",
        "\n",
        "#### **Tips and Instructions**\n",
        "- click the little ▶ play icon to the left of each cell below.\n",
        "- use \"/\" to specify chainbreaks, (eg. sequence=\"AAA/AAA\")\n",
        "- for homo-oligomeric predictions, set copies > 1\n"
      ],
      "metadata": {
        "id": "qEgffIPOyEgk"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "oVkGfduwiAkw"
      },
      "outputs": [],
      "source": [
        "#@markdown ##Install\n",
        "import os,sys,re\n",
        "from IPython.utils import io\n",
        "if \"SETUP_DONE\" not in dir():\n",
        "  import torch\n",
        "  device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
        "  with io.capture_output() as captured:\n",
        "    if not os.path.isdir(\"OmegaFold\"):\n",
        "      %shell git clone --branch beta --quiet https://github.com/sokrypton/OmegaFold.git\n",
        "      # %shell cd OmegaFold; pip -q install -r requirements.txt\n",
        "      %shell pip -q install py3Dmol biopython==1.81\n",
        "      %shell apt-get install aria2 -qq > /dev/null\n",
        "      %shell aria2c -q -x 16 https://helixon.s3.amazonaws.com/release1.pt\n",
        "      %shell mkdir -p ~/.cache/omegafold_ckpt\n",
        "      %shell mv release1.pt ~/.cache/omegafold_ckpt/model.pt\n",
        "  SETUP_DONE = True"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown ##Run **OmegaFold**\n",
        "from string import ascii_uppercase, ascii_lowercase\n",
        "import hashlib\n",
        "\n",
        "def get_hash(x): return hashlib.sha1(x.encode()).hexdigest()\n",
        "alphabet_list = list(ascii_uppercase+ascii_lowercase)\n",
        "\n",
        "jobname = \"test\" #@param {type:\"string\"}\n",
        "jobname = re.sub(r'\\W+', '', jobname)[:50]\n",
        "\n",
        "sequence = \"MKVIFLKKGEIKNVADGYANNFLFKQGLAIEA\" #@param {type:\"string\"}\n",
        "sequence = re.sub(\"[^A-Z:]\", \"\", sequence.replace(\"/\",\":\").upper())\n",
        "sequence = re.sub(\":+\",\":\",sequence)\n",
        "sequence = re.sub(\"^[:]+\",\"\",sequence)\n",
        "sequence = re.sub(\"[:]+$\",\"\",sequence)\n",
        "\n",
        "copies = 1 #@param {type:\"integer\"}\n",
        "sequence = \":\".join([sequence] * copies)\n",
        "#@markdown **Advanced Options**\n",
        "num_cycle = 4 #@param [\"1\", \"2\", \"4\", \"8\", \"16\", \"32\"] {type:\"raw\"}\n",
        "offset_rope = False #@param {type:\"boolean\"}\n",
        "\n",
        "ID = jobname+\"_\"+get_hash(sequence)[:5]\n",
        "seqs = sequence.split(\":\")\n",
        "lengths = [len(s) for s in seqs]\n",
        "\n",
        "def get_subbatch_size(L):\n",
        "  if L <  500: return 500\n",
        "  if L < 1000: return 200\n",
        "  return 150\n",
        "subbatch_size = get_subbatch_size(sum(lengths))\n",
        "\n",
        "u_seqs = list(set(seqs))\n",
        "\n",
        "if len(seqs) == 1: mode = \"mono\"\n",
        "elif len(u_seqs) == 1: mode = \"homo\"\n",
        "else: mode = \"hetero\"\n",
        "\n",
        "with open(f\"{ID}.fasta\",\"w\") as out:\n",
        "  out.write(f\">{ID}\\n{sequence}\\n\")\n",
        "\n",
        "%shell python OmegaFold/main.py --offset_rope={offset_rope} --device={device} --subbatch_size={subbatch_size} --num_cycle={num_cycle} {ID}.fasta .\n",
        "\n",
        "def renum_pdb_str(pdb_str, Ls=None, renum=True, offset=1):\n",
        "  if Ls is not None:\n",
        "    L_init = 0\n",
        "    new_chain = {}\n",
        "    for L,c in zip(Ls, alphabet_list):\n",
        "      new_chain.update({i:c for i in range(L_init,L_init+L)})\n",
        "      L_init += L\n",
        "\n",
        "  n,num,pdb_out = 0,offset,[]\n",
        "  resnum_ = None\n",
        "  chain_ = None\n",
        "  new_chain_ = new_chain[0]\n",
        "  for line in pdb_str.split(\"\\n\"):\n",
        "    if line[:4] == \"ATOM\":\n",
        "      chain = line[21:22]\n",
        "      resnum = int(line[22:22+5])\n",
        "      if resnum_ is None: resnum_ = resnum\n",
        "      if chain_ is None: chain_ = chain\n",
        "      if resnum != resnum_ or chain != chain_:\n",
        "        num += (resnum - resnum_)\n",
        "        n += 1\n",
        "        resnum_,chain_ = resnum,chain\n",
        "      if Ls is not None:\n",
        "        if new_chain[n] != new_chain_:\n",
        "          num = offset\n",
        "          new_chain_ = new_chain[n]\n",
        "      N = num if renum else resnum\n",
        "      if Ls is None: pdb_out.append(\"%s%4i%s\" % (line[:22],N,line[26:]))\n",
        "      else: pdb_out.append(\"%s%s%4i%s\" % (line[:21],new_chain[n],N,line[26:]))\n",
        "  return \"\\n\".join(pdb_out)\n",
        "\n",
        "pdb_str = renum_pdb_str(open(f\"{ID}.pdb\",'r').read(), Ls=lengths)\n",
        "o_pdb_file=f\"{ID}.pdb\"\n",
        "with open(o_pdb_file,\"w\") as out:\n",
        "  out.write(pdb_str)\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "CFCwEAa2oZEN",
        "outputId": "ab46fc05-88c9-4b15-fe3c-8f608d787767",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/usr/local/lib/python3.12/dist-packages/Bio/Data/SCOPData.py:18: BiopythonDeprecationWarning: The 'Bio.Data.SCOPData' module will be deprecated in a future release of Biopython in favor of 'Bio.Data.PDBData.\n",
            "  warnings.warn(\n",
            "/content/OmegaFold/omegafold/utils/protein_utils/residue_constants.py:483: UserWarning: Using torch.cross without specifying the dim arg is deprecated.\n",
            "Please either pass the dim explicitly or simply use torch.linalg.cross.\n",
            "The default value of dim will change to agree with that of linalg.cross in a future release. (Triggered internally at /pytorch/aten/src/ATen/native/Cross.cpp:63.)\n",
            "  eznorm = torch.cross(ex_normalized, ey_normalized)\n",
            "INFO:root:Loading weights from /root/.cache/omegafold_ckpt/model.pt\n",
            "INFO:root:Constructing OmegaFold\n",
            "INFO:root:Reading test_a2395.fasta\n",
            "INFO:root:Predicting 1th chain in test_a2395.fasta\n",
            "INFO:root:32 residues in this chain.\n",
            "INFO:root:Finished prediction in 3.10 seconds.\n",
            "INFO:root:Saving prediction to ./test_a2395.pdb\n",
            "INFO:root:Saved\n",
            "INFO:root:Done!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown ##Display\n",
        "import py3Dmol\n",
        "\n",
        "\n",
        "pymol_color_list = [\"#33ff33\",\"#00ffff\",\"#ff33cc\",\"#ffff00\",\"#ff9999\",\"#e5e5e5\",\"#7f7fff\",\"#ff7f00\",\n",
        "                    \"#7fff7f\",\"#199999\",\"#ff007f\",\"#ffdd5e\",\"#8c3f99\",\"#b2b2b2\",\"#007fff\",\"#c4b200\",\n",
        "                    \"#8cb266\",\"#00bfbf\",\"#b27f7f\",\"#fcd1a5\",\"#ff7f7f\",\"#ffbfdd\",\"#7fffff\",\"#ffff7f\",\n",
        "                    \"#00ff7f\",\"#337fcc\",\"#d8337f\",\"#bfff3f\",\"#ff7fff\",\"#d8d8ff\",\"#3fffbf\",\"#b78c4c\",\n",
        "                    \"#339933\",\"#66b2b2\",\"#ba8c84\",\"#84bf00\",\"#b24c66\",\"#7f7f7f\",\"#3f3fa5\",\"#a5512b\"]\n",
        "\n",
        "def show_pdb(pdb_str, show_sidechains=False, show_mainchains=False,\n",
        "             color=\"pLDDT\", chains=None, vmin=50, vmax=90,\n",
        "             size=(800,480), hbondCutoff=4.0,\n",
        "             Ls=None,\n",
        "             animate=False):\n",
        "\n",
        "  if chains is None:\n",
        "    chains = 1 if Ls is None else len(Ls)\n",
        "  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js', width=size[0], height=size[1])\n",
        "  if animate:\n",
        "    view.addModelsAsFrames(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "  else:\n",
        "    view.addModel(pdb_str,'pdb',{'hbondCutoff':hbondCutoff})\n",
        "  if color == \"pLDDT\":\n",
        "    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':vmin,'max':vmax}}})\n",
        "  elif color == \"rainbow\":\n",
        "    view.setStyle({'cartoon': {'color':'spectrum'}})\n",
        "  elif color == \"chain\":\n",
        "    for n,chain,color in zip(range(chains),alphabet_list,pymol_color_list):\n",
        "       view.setStyle({'chain':chain},{'cartoon': {'color':color}})\n",
        "  if show_sidechains:\n",
        "    BB = ['C','O','N']\n",
        "    HP = [\"ALA\",\"GLY\",\"VAL\",\"ILE\",\"LEU\",\"PHE\",\"MET\",\"PRO\",\"TRP\",\"CYS\",\"TYR\"]\n",
        "    view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n",
        "                  {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "    view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]},\n",
        "                  {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "    view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n",
        "                  {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "  if show_mainchains:\n",
        "    BB = ['C','O','N','CA']\n",
        "    view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n",
        "  view.zoomTo()\n",
        "  if animate: view.animate()\n",
        "  return view\n",
        "\n",
        "color = \"confidence\" #@param [\"confidence\", \"rainbow\", \"chain\"]\n",
        "if color == \"confidence\": color = \"pLDDT\"\n",
        "show_sidechains = False #@param {type:\"boolean\"}\n",
        "show_mainchains = False #@param {type:\"boolean\"}\n",
        "show_pdb(pdb_str, color=color, show_sidechains=show_sidechains, show_mainchains=show_mainchains,\n",
        "         Ls=lengths).show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "dHl9aS-HHFNQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython"
      ],
      "metadata": {
        "id": "KxboCaL1ev9v",
        "outputId": "bb8e8eee-5fc0-429b-eaad-5bdbdb02e1b4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.12/dist-packages (1.81)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Mutate DNA and inspect translation\n",
        "from IPython.display import display, HTML\n",
        "from textwrap import wrap\n",
        "import Bio\n",
        "from Bio.Seq import Seq\n",
        "\n",
        "# === Inputs ===\n",
        "original_seq = \"ATG AAA GTG ATT TTT CTG AAA AAA GGC GAA ATT AAA AAC GTG GCG GAT GGC TAT GCG AAC AAC TTT CTG TTT AAA CAG GGC CTG GCG ATT GAA GCG\" #@param {type:\"string\"}\n",
        "mutated_seq  = \"ATG AAA GTG ATT TTT CTG AAA ATA GGC GAA ATT AAA AAC GTG GCG GAT GGC TAT GCG AAC AAC TTT CTG TTT AAA CAG GGC CTG GCG ATT GAA GCG\" #@param {type:\"string\"}\n",
        "\n",
        "# Clean input\n",
        "original_seq = original_seq.replace(\" \", \"\").upper()\n",
        "mutated_seq  = mutated_seq.replace(\" \", \"\").upper()\n",
        "\n",
        "# === Checks ===\n",
        "if len(original_seq) != len(mutated_seq):\n",
        "    raise ValueError(f\"Sequences must have the same length (got {len(original_seq)} vs {len(mutated_seq)}).\")\n",
        "if len(original_seq) % 3 != 0:\n",
        "    raise ValueError(f\"Sequence length must be a multiple of 3 (got {len(original_seq)}).\")\n",
        "\n",
        "# === Genetic code (DNA codons) ===\n",
        "codon_table = {\n",
        "    \"TTT\":\"F\",\"TTC\":\"F\",\"TTA\":\"L\",\"TTG\":\"L\",\n",
        "    \"CTT\":\"L\",\"CTC\":\"L\",\"CTA\":\"L\",\"CTG\":\"L\",\n",
        "    \"ATT\":\"I\",\"ATC\":\"I\",\"ATA\":\"I\",\"ATG\":\"M\",\n",
        "    \"GTT\":\"V\",\"GTC\":\"V\",\"GTA\":\"V\",\"GTG\":\"V\",\n",
        "    \"TCT\":\"S\",\"TCC\":\"S\",\"TCA\":\"S\",\"TCG\":\"S\",\n",
        "    \"CCT\":\"P\",\"CCC\":\"P\",\"CCA\":\"P\",\"CCG\":\"P\",\n",
        "    \"ACT\":\"T\",\"ACC\":\"T\",\"ACA\":\"T\",\"ACG\":\"T\",\n",
        "    \"GCT\":\"A\",\"GCC\":\"A\",\"GCA\":\"A\",\"GCG\":\"A\",\n",
        "    \"TAT\":\"Y\",\"TAC\":\"Y\",\"TAA\":\"*\",\"TAG\":\"*\",\n",
        "    \"CAT\":\"H\",\"CAC\":\"H\",\"CAA\":\"Q\",\"CAG\":\"Q\",\n",
        "    \"AAT\":\"N\",\"AAC\":\"N\",\"AAA\":\"K\",\"AAG\":\"K\",\n",
        "    \"GAT\":\"D\",\"GAC\":\"D\",\"GAA\":\"E\",\"GAG\":\"E\",\n",
        "    \"TGT\":\"C\",\"TGC\":\"C\",\"TGA\":\"*\",\"TGG\":\"W\",\n",
        "    \"CGT\":\"R\",\"CGC\":\"R\",\"CGA\":\"R\",\"CGG\":\"R\",\n",
        "    \"AGT\":\"S\",\"AGC\":\"S\",\"AGA\":\"R\",\"AGG\":\"R\",\n",
        "    \"GGT\":\"G\",\"GGC\":\"G\",\"GGA\":\"G\",\"GGG\":\"G\"\n",
        "}\n",
        "\n",
        "def translate_dna(seq):\n",
        "    codons = wrap(seq, 3)\n",
        "    aa = [codon_table.get(c,\"?\") for c in codons]\n",
        "    return codons, aa\n",
        "\n",
        "from IPython.display import display, HTML\n",
        "\n",
        "def compare_sequences(seq1, seq2, space_every=10, spacer=\" \", no_number=False):\n",
        "    \"\"\"\n",
        "    Display two sequences in monospace, highlight differences in bold red,\n",
        "    insert spaces every n characters, and add numbering above each block.\n",
        "    \"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        raise ValueError(\"Sequences must be the same length\")\n",
        "\n",
        "    def highlight_block(s1, s2):\n",
        "        \"\"\"Highlight differences in one block of equal length\"\"\"\n",
        "        out1, out2 = [], []\n",
        "        for a, b in zip(s1, s2):\n",
        "            out1.append(a)\n",
        "            if a == b:\n",
        "                out2.append(b)\n",
        "            else:\n",
        "                out2.append(f\"<span style='color:red;font-weight:bold'>{b}</span>\")\n",
        "        return \"\".join(out1), \"\".join(out2)\n",
        "\n",
        "    # Split into blocks of size space_every\n",
        "    blocks1 = [seq1[i:i+space_every] for i in range(0, len(seq1), space_every)]\n",
        "    blocks2 = [seq2[i:i+space_every] for i in range(0, len(seq2), space_every)]\n",
        "\n",
        "    seq1_blocks, seq2_blocks, num_blocks = [], [], []\n",
        "    pos = 0\n",
        "    for b1, b2 in zip(blocks1, blocks2):\n",
        "        o1, o2 = highlight_block(b1, b2)\n",
        "        seq1_blocks.append(o1)\n",
        "        seq2_blocks.append(o2)\n",
        "        pos += len(b1)\n",
        "        num_blocks.append(str(pos).rjust(len(b1)))\n",
        "\n",
        "\n",
        "    number_line = spacer.join(num_blocks)\n",
        "    seq1_line   = spacer.join(seq1_blocks)\n",
        "    seq2_line   = spacer.join(seq2_blocks)\n",
        "\n",
        "    html = []\n",
        "    if no_number:\n",
        "        html = seq1_line + \"<br>\" + seq2_line\n",
        "    else:\n",
        "        html = number_line + \"<br>\" + seq1_line + \"<br>\" + seq2_line\n",
        "    display(HTML(f\"<pre style='font-family:monospace'>{html}</pre>\"))\n",
        "\n",
        "def dna_identity(seq1: str, seq2: str) -> float:\n",
        "    \"\"\"Return % identity at DNA level\"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        raise ValueError(\"Sequences must have the same length\")\n",
        "    matches = sum(a == b for a, b in zip(seq1.upper(), seq2.upper()))\n",
        "    return 100 * matches / len(seq1)\n",
        "\n",
        "def protein_identity(seq1: str, seq2: str) -> float:\n",
        "    \"\"\"Translate DNA → protein and return % identity\"\"\"\n",
        "    if len(seq1) % 3 != 0 or len(seq2) % 3 != 0:\n",
        "        raise ValueError(\"DNA sequences must be divisible by 3\")\n",
        "    prot1 = str(Seq(seq1).translate(to_stop=False))\n",
        "    prot2 = str(Seq(seq2).translate(to_stop=False))\n",
        "    if len(prot1) != len(prot2):\n",
        "        raise ValueError(\"Translated proteins have different lengths\")\n",
        "    matches = sum(a == b for a, b in zip(prot1, prot2))\n",
        "    return 100 * matches / len(prot1)\n",
        "\n",
        "def mutation_score(seq1: str, seq2: str) -> float:\n",
        "    \"\"\"Return mutation score\"\"\"\n",
        "    if len(seq1) != len(seq2):\n",
        "        raise ValueError(\"Sequences must have the same length\")\n",
        "\n",
        "    protein_id = protein_identity(seq1, seq2)\n",
        "    dna_id = dna_identity(seq1, seq2)\n",
        "\n",
        "    return ((100.-protein_id)+(100.-dna_id))/2\n",
        "\n",
        "# === Translation ===\n",
        "codons_o, aa_o = translate_dna(original_seq)\n",
        "codons_m, aa_m = translate_dna(mutated_seq)\n",
        "\n",
        "# === HTML display ===\n",
        "\n",
        "compare_sequences(original_seq,mutated_seq,space_every=3)\n",
        "compare_sequences(aa_o,aa_m,space_every=1, no_number=True, spacer=\"   \")\n",
        "\n",
        "#display(HTML(f\"<pre style='font-family:monospace'>{\"   \".join(aa_o)}</pre>\"))\n",
        "#compare_sequences(\"\".join(codons_o),\"\".join(codons_m),space_every=30)\n",
        "\n",
        "print(\"\\n\\nMutaded sequence without spaces:\")\n",
        "display(HTML(f\"<pre style='font-family:monospace'>{\"\".join(aa_m)}</pre>\"))\n",
        "\n",
        "dna_id = dna_identity(original_seq, mutated_seq)\n",
        "#print(f\"\\nDNA % identity: {dna_id:.1f}%\")\n",
        "protein_id = protein_identity(original_seq, mutated_seq)\n",
        "#print(f\"Protein % identity: {protein_id:.1f}%\")\n",
        "print(\"\\n\")\n",
        "mut_score = mutation_score(original_seq,mutated_seq)\n",
        "print(f\"Mutation score is {mut_score:.2f}\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "z3HeD2V8epVW",
        "outputId": "aef5b986-8e48-47c9-f0dc-31957f2fc4ce",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 229
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<pre style='font-family:monospace'>  3   6   9  12  15  18  21  24  27  30  33  36  39  42  45  48  51  54  57  60  63  66  69  72  75  78  81  84  87  90  93  96<br>ATG AAA GTG ATT TTT CTG AAA AAA GGC GAA ATT AAA AAC GTG GCG GAT GGC TAT GCG AAC AAC TTT CTG TTT AAA CAG GGC CTG GCG ATT GAA GCG<br>ATG AAA GTG ATT TTT CTG AAA A<span style='color:red;font-weight:bold'>T</span>A GGC GAA ATT AAA AAC GTG GCG GAT GGC TAT GCG AAC AAC TTT CTG TTT AAA CAG GGC CTG GCG ATT GAA GCG</pre>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<pre style='font-family:monospace'>M   K   V   I   F   L   K   K   G   E   I   K   N   V   A   D   G   Y   A   N   N   F   L   F   K   Q   G   L   A   I   E   A<br>M   K   V   I   F   L   K   <span style='color:red;font-weight:bold'>I</span>   G   E   I   K   N   V   A   D   G   Y   A   N   N   F   L   F   K   Q   G   L   A   I   E   A</pre>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Mutaded sequence without spaces:\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<pre style='font-family:monospace'>MKVIFLKIGEIKNVADGYANNFLFKQGLAIEA</pre>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Mutation score is 2.08\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Calculate RMSD-based final score\n",
        "import os, re, requests\n",
        "import torch\n",
        "from jax.tree_util import tree_map\n",
        "from Bio.Seq import Seq\n",
        "from Bio.PDB import PDBParser, Superimposer, PDBIO\n",
        "import numpy as np\n",
        "import tempfile\n",
        "\n",
        "# --- Superpose prediction onto reference ---\n",
        "def align_pred_to_ref(pred_pdb_str, ref_pdb_file, out_file=\"pred_aligned.pdb\"):\n",
        "    parser = PDBParser(QUIET=True)\n",
        "\n",
        "    # Save predicted PDB string to temp file\n",
        "    with tempfile.NamedTemporaryFile(delete=False, suffix=\".pdb\") as tmp_pred:\n",
        "        tmp_pred.write(pred_pdb_str.encode())\n",
        "        pred_file = tmp_pred.name\n",
        "\n",
        "    pred = parser.get_structure(\"pred\", pred_file)\n",
        "    ref  = parser.get_structure(\"ref\", ref_pdb_file)\n",
        "\n",
        "    # Use CA atoms for alignment\n",
        "    pred_atoms = [a for a in pred.get_atoms() if a.get_id() == \"CA\"]\n",
        "    ref_atoms  = [a for a in ref.get_atoms()  if a.get_id() == \"CA\"]\n",
        "\n",
        "    n = min(len(pred_atoms), len(ref_atoms))\n",
        "    si = Superimposer()\n",
        "    si.set_atoms(ref_atoms[:n], pred_atoms[:n])  # align prediction to reference\n",
        "    si.apply(pred.get_atoms())                   # apply transform to prediction\n",
        "\n",
        "    # Write aligned prediction\n",
        "    io = PDBIO()\n",
        "    io.set_structure(pred)\n",
        "    io.save(out_file)\n",
        "\n",
        "    return out_file, si.rms\n",
        "\n",
        "def rmsd_to_similarity(rmsd, sigma=2.0):\n",
        "    \"\"\"\n",
        "    Convert RMSD (Å) to a similarity score in [0,1].\n",
        "    Smaller RMSD -> score close to 1, larger RMSD -> score close to 0.\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    rmsd : float or np.array\n",
        "        Root-mean-square deviation in angstroms.\n",
        "    sigma : float\n",
        "        Scaling factor controlling decay. Typical RMSDs of 1-2 Å → high similarity.\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "    similarity : float or np.array\n",
        "        Score in [0,1].\n",
        "    \"\"\"\n",
        "    return np.exp(-rmsd / sigma)\n",
        "\n",
        "# --- Download reference PDB ---\n",
        "ref_file=\"reference.pdb\"\n",
        "if not os.path.isfile(ref_file):\n",
        "  !wget -q -O reference.pdb https://raw.githubusercontent.com/bforsbe/SK2534/main/ref_for_mutation.pdb\n",
        "\n",
        "aligned_file, rmsd_val = align_pred_to_ref(pdb_str, ref_file)\n",
        "\n",
        "#for rmsd in [0.5, 1.0, 2.0, 5.0]:\n",
        "#    print(rmsd, rmsd_to_similarity(rmsd))\n",
        "pred_sim_score = rmsd_to_similarity(rmsd_val)\n",
        "mut_score=mutation_score(original_seq,mutated_seq)\n",
        "\n",
        "# --- Compare ---\n",
        "print(f\"Input file is {o_pdb_file}\")\n",
        "#print(f\"Protein sequence: {protein_seq}\")\n",
        "print(f\"mutation score: \\t{mut_score:.2f})\")\n",
        "print(f\"RMSD vs reference: \\t{rmsd_val:.3f} Å (score: {pred_sim_score:.2f})\")\n",
        "print(\"\\n\\n\")\n",
        "print(f\"Total score: \\t{mut_score*pred_sim_score:.3f}\")\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "c51BDdsXe68l",
        "outputId": "ee565c57-7c27-4a6a-b7b7-f580d05102c6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Input file is test_a2395.pdb\n",
            "mutation score: \t2.08)\n",
            "RMSD vs reference: \t0.350 Å (score: 0.84)\n",
            "\n",
            "\n",
            "\n",
            "Total score: \t1.749\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Display predicted (superposed) + reference {run: \"auto\"}\n",
        "import py3Dmol\n",
        "from Bio.PDB import PDBParser, Superimposer, PDBIO\n",
        "import tempfile, os\n",
        "\n",
        "# --- Visualize both ---\n",
        "def show_pred_and_ref_aligned(pred_pdb_str, ref_pdb_file,\n",
        "                              color=\"pLDDT\", show_sidechains=False, show_mainchains=False,\n",
        "                              chains=None, vmin=50, vmax=90, size=(800,480), Ls=None, hide_pred=False):\n",
        "\n",
        "    aligned_file, rmsd = align_pred_to_ref(pred_pdb_str, ref_pdb_file)\n",
        "\n",
        "    view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',\n",
        "                        width=size[0], height=size[1])\n",
        "\n",
        "    # Reference (fixed, green)\n",
        "    with open(ref_pdb_file) as f:\n",
        "        ref_pdb_str = f.read()\n",
        "    view.addModel(ref_pdb_str,'pdb')\n",
        "    view.setStyle({'model':0},{'cartoon':{'color':'green','opacity':0.7}})\n",
        "\n",
        "    # Prediction (aligned, colored)\n",
        "    with open(aligned_file) as f:\n",
        "        pred_aligned_str = f.read()\n",
        "    if not hide_pred:\n",
        "        view.addModel(pred_aligned_str,'pdb')\n",
        "    if color == \"pLDDT\":\n",
        "        view.setStyle({'model':1,'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':vmin,'max':vmax}}})\n",
        "    elif color == \"rainbow\":\n",
        "        view.setStyle({'model':1,'cartoon': {'color':'spectrum'}})\n",
        "    elif color == \"chain\":\n",
        "        from string import ascii_uppercase\n",
        "        alphabet_list = list(ascii_uppercase)\n",
        "        for n,chain,c in zip(range(chains),alphabet_list,pymol_color_list):\n",
        "            view.setStyle({'model':1,'chain':chain},{'cartoon': {'color':c}})\n",
        "\n",
        "    if show_sidechains:\n",
        "        view.addStyle({'model':1,'stick':{}})\n",
        "\n",
        "    view.zoomTo()\n",
        "    print(f\"Structural alignment RMSD: {rmsd:.3f} Å\")\n",
        "    return view\n",
        "\n",
        "\n",
        "# --- Params ---\n",
        "color = \"confidence\" #@param [\"confidence\", \"rainbow\", \"chain\"]\n",
        "if color == \"confidence\": color = \"pLDDT\"\n",
        "show_sidechains = True #@param {type:\"boolean\"}\n",
        "show_mainchains = True #@param {type:\"boolean\"}\n",
        "hide_predicted = False #@param {type:\"boolean\"}\n",
        "\n",
        "# --- Run ---\n",
        "\n",
        "show_pred_and_ref_aligned(pdb_str, \"reference.pdb\",\n",
        "                          color=color,\n",
        "                          show_sidechains=show_sidechains,\n",
        "                          show_mainchains=show_mainchains,\n",
        "                          Ls=lengths,\n",
        "                          hide_pred=hide_predicted).show()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "ssk3BxdUgdyS",
        "outputId": "ae975f1d-a7b1-4bf6-fa62-01a91b506376",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 515
        }
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Structural alignment RMSD: 0.350 Å\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_17580091717215044\"  style=\"position: relative; width: 800px; height: 480px;\">\n        <p id=\"3dmolwarning_17580091717215044\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_17580091717215044 = null;\nvar warn = document.getElementById(\"3dmolwarning_17580091717215044\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_17580091717215044 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17580091717215044\"),{backgroundColor:\"white\"});\nviewer_17580091717215044.zoomTo();\n\tviewer_17580091717215044.addModel(\"PARENT N/A\\nATOM      1  N   MET A   1      -6.170  -7.770   1.344  1.00 93.76           N  \\nATOM      2  CA  MET A   1      -5.898  -7.356   2.717  1.00 94.04           C  \\nATOM      3  C   MET A   1      -6.200  -5.874   2.909  1.00 94.69           C  \\nATOM      4  CB  MET A   1      -4.442  -7.646   3.088  1.00 91.60           C  \\nATOM      5  O   MET A   1      -6.172  -5.101   1.950  1.00 93.73           O  \\nATOM      6  CG  MET A   1      -3.436  -6.783   2.343  1.00 80.00           C  \\nATOM      7  SD  MET A   1      -1.700  -7.222   2.743  1.00 70.55           S  \\nATOM      8  CE  MET A   1      -1.661  -8.939   2.156  1.00 69.74           C  \\nATOM      9  N   LYS A   2      -6.599  -5.431   3.986  1.00 95.80           N  \\nATOM     10  CA  LYS A   2      -6.886  -4.054   4.376  1.00 96.00           C  \\nATOM     11  C   LYS A   2      -5.621  -3.340   4.844  1.00 95.94           C  \\nATOM     12  CB  LYS A   2      -7.947  -4.017   5.478  1.00 94.97           C  \\nATOM     13  O   LYS A   2      -4.883  -3.861   5.683  1.00 95.43           O  \\nATOM     14  CG  LYS A   2      -8.489  -2.625   5.769  1.00 86.65           C  \\nATOM     15  CD  LYS A   2      -9.599  -2.665   6.812  1.00 82.11           C  \\nATOM     16  CE  LYS A   2     -10.196  -1.284   7.046  1.00 80.93           C  \\nATOM     17  NZ  LYS A   2     -11.327  -1.328   8.020  1.00 70.78           N  \\nATOM     18  N   VAL A   3      -5.353  -2.202   4.207  1.00 95.03           N  \\nATOM     19  CA  VAL A   3      -4.110  -1.508   4.526  1.00 95.38           C  \\nATOM     20  C   VAL A   3      -4.400  -0.037   4.819  1.00 94.98           C  \\nATOM     21  CB  VAL A   3      -3.080  -1.630   3.381  1.00 94.41           C  \\nATOM     22  O   VAL A   3      -5.432   0.493   4.403  1.00 94.15           O  \\nATOM     23  CG1 VAL A   3      -2.753  -3.096   3.104  1.00 85.54           C  \\nATOM     24  CG2 VAL A   3      -3.604  -0.949   2.118  1.00 84.45           C  \\nATOM     25  N   ILE A   4      -3.530   0.635   5.581  1.00 95.05           N  \\nATOM     26  CA  ILE A   4      -3.494   2.078   5.794  1.00 95.30           C  \\nATOM     27  C   ILE A   4      -2.356   2.692   4.981  1.00 95.22           C  \\nATOM     28  CB  ILE A   4      -3.330   2.424   7.291  1.00 94.33           C  \\nATOM     29  O   ILE A   4      -1.182   2.416   5.238  1.00 94.62           O  \\nATOM     30  CG1 ILE A   4      -4.429   1.748   8.119  1.00 88.67           C  \\nATOM     31  CG2 ILE A   4      -3.341   3.941   7.500  1.00 88.04           C  \\nATOM     32  CD1 ILE A   4      -4.165   1.752   9.619  1.00 82.29           C  \\nATOM     33  N   PHE A   5      -2.663   3.500   3.947  1.00 93.06           N  \\nATOM     34  CA  PHE A   5      -1.646   4.228   3.197  1.00 94.01           C  \\nATOM     35  C   PHE A   5      -1.079   5.374   4.027  1.00 93.25           C  \\nATOM     36  CB  PHE A   5      -2.227   4.768   1.886  1.00 92.21           C  \\nATOM     37  O   PHE A   5      -1.829   6.207   4.541  1.00 92.16           O  \\nATOM     38  CG  PHE A   5      -2.460   3.707   0.844  1.00 85.50           C  \\nATOM     39  CD1 PHE A   5      -1.509   3.457  -0.137  1.00 78.06           C  \\nATOM     40  CD2 PHE A   5      -3.630   2.960   0.846  1.00 79.69           C  \\nATOM     41  CE1 PHE A   5      -1.721   2.476  -1.103  1.00 84.65           C  \\nATOM     42  CE2 PHE A   5      -3.850   1.978  -0.115  1.00 83.91           C  \\nATOM     43  CZ  PHE A   5      -2.895   1.738  -1.090  1.00 84.90           C  \\nATOM     44  N   LEU A   6       0.127   5.418   4.151  1.00 94.20           N  \\nATOM     45  CA  LEU A   6       0.788   6.292   5.114  1.00 93.99           C  \\nATOM     46  C   LEU A   6       0.783   7.737   4.628  1.00 93.50           C  \\nATOM     47  CB  LEU A   6       2.227   5.828   5.359  1.00 93.21           C  \\nATOM     48  O   LEU A   6       0.779   8.668   5.437  1.00 92.13           O  \\nATOM     49  CG  LEU A   6       2.395   4.473   6.047  1.00 90.40           C  \\nATOM     50  CD1 LEU A   6       3.865   4.067   6.068  1.00 86.76           C  \\nATOM     51  CD2 LEU A   6       1.827   4.518   7.462  1.00 86.68           C  \\nATOM     52  N   LYS A   7       0.842   7.971   3.352  1.00 93.29           N  \\nATOM     53  CA  LYS A   7       0.926   9.328   2.821  1.00 92.89           C  \\nATOM     54  C   LYS A   7      -0.349  10.113   3.116  1.00 92.19           C  \\nATOM     55  CB  LYS A   7       1.188   9.300   1.315  1.00 91.34           C  \\nATOM     56  O   LYS A   7      -0.291  11.231   3.631  1.00 90.08           O  \\nATOM     57  CG  LYS A   7       1.450  10.670   0.707  1.00 83.48           C  \\nATOM     58  CD  LYS A   7       1.786  10.569  -0.776  1.00 78.24           C  \\nATOM     59  CE  LYS A   7       2.017  11.942  -1.392  1.00 74.52           C  \\nATOM     60  NZ  LYS A   7       2.387  11.846  -2.836  1.00 67.81           N  \\nATOM     61  N   LYS A   8      -1.548   9.629   2.889  1.00 89.59           N  \\nATOM     62  CA  LYS A   8      -2.804  10.359   3.038  1.00 88.28           C  \\nATOM     63  C   LYS A   8      -3.622   9.815   4.205  1.00 86.79           C  \\nATOM     64  CB  LYS A   8      -3.622  10.289   1.747  1.00 84.90           C  \\nATOM     65  O   LYS A   8      -4.647  10.391   4.574  1.00 83.20           O  \\nATOM     66  CG  LYS A   8      -3.053  11.121   0.607  1.00 75.06           C  \\nATOM     67  CD  LYS A   8      -4.032  11.220  -0.555  1.00 71.84           C  \\nATOM     68  CE  LYS A   8      -3.510  12.140  -1.650  1.00 65.02           C  \\nATOM     69  NZ  LYS A   8      -4.486  12.277  -2.772  1.00 59.33           N  \\nATOM     70  N   GLY A   9      -3.098   8.783   4.783  1.00 89.03           N  \\nATOM     71  CA  GLY A   9      -3.855   8.185   5.871  1.00 90.06           C  \\nATOM     72  C   GLY A   9      -5.116   7.482   5.405  1.00 86.45           C  \\nATOM     73  O   GLY A   9      -6.059   7.309   6.180  1.00 80.62           O  \\nATOM     74  N   GLU A  10      -5.183   7.049   4.141  1.00 93.36           N  \\nATOM     75  CA  GLU A  10      -6.365   6.401   3.581  1.00 93.22           C  \\nATOM     76  C   GLU A  10      -6.293   4.886   3.745  1.00 92.93           C  \\nATOM     77  CB  GLU A  10      -6.527   6.763   2.103  1.00 89.96           C  \\nATOM     78  O   GLU A  10      -5.228   4.288   3.577  1.00 91.53           O  \\nATOM     79  CG  GLU A  10      -6.805   8.239   1.858  1.00 75.06           C  \\nATOM     80  CD  GLU A  10      -7.052   8.569   0.395  1.00 72.15           C  \\nATOM     81  OE1 GLU A  10      -6.768   7.715  -0.476  1.00 69.12           O  \\nATOM     82  OE2 GLU A  10      -7.534   9.690   0.116  1.00 66.84           O  \\nATOM     83  N   ILE A  11      -7.474   4.318   4.108  1.00 94.14           N  \\nATOM     84  CA  ILE A  11      -7.596   2.869   4.227  1.00 94.55           C  \\nATOM     85  C   ILE A  11      -8.186   2.294   2.941  1.00 94.13           C  \\nATOM     86  CB  ILE A  11      -8.466   2.473   5.440  1.00 92.91           C  \\nATOM     87  O   ILE A  11      -9.228   2.754   2.469  1.00 93.11           O  \\nATOM     88  CG1 ILE A  11      -7.864   3.032   6.735  1.00 78.19           C  \\nATOM     89  CG2 ILE A  11      -8.620   0.951   5.520  1.00 77.84           C  \\nATOM     90  CD1 ILE A  11      -8.784   2.922   7.943  1.00 73.86           C  \\nATOM     91  N   LYS A  12      -7.548   1.210   2.482  1.00 94.52           N  \\nATOM     92  CA  LYS A  12      -8.046   0.587   1.259  1.00 94.67           C  \\nATOM     93  C   LYS A  12      -7.769  -0.914   1.255  1.00 94.68           C  \\nATOM     94  CB  LYS A  12      -7.415   1.240   0.029  1.00 92.98           C  \\nATOM     95  O   LYS A  12      -6.842  -1.381   1.919  1.00 94.11           O  \\nATOM     96  CG  LYS A  12      -7.680   2.734  -0.084  1.00 79.69           C  \\nATOM     97  CD  LYS A  12      -7.272   3.273  -1.449  1.00 76.19           C  \\nATOM     98  CE  LYS A  12      -7.593   4.755  -1.585  1.00 71.33           C  \\nATOM     99  NZ  LYS A  12      -7.230   5.279  -2.935  1.00 61.53           N  \\nATOM    100  N   ASN A  13      -8.592  -1.621   0.627  1.00 95.04           N  \\nATOM    101  CA  ASN A  13      -8.325  -3.019   0.310  1.00 95.29           C  \\nATOM    102  C   ASN A  13      -7.495  -3.157  -0.963  1.00 95.21           C  \\nATOM    103  CB  ASN A  13      -9.635  -3.799   0.177  1.00 94.19           C  \\nATOM    104  O   ASN A  13      -7.791  -2.517  -1.974  1.00 94.61           O  \\nATOM    105  CG  ASN A  13     -10.408  -3.869   1.479  1.00 88.42           C  \\nATOM    106  ND2 ASN A  13     -11.716  -3.648   1.405  1.00 85.06           N  \\nATOM    107  OD1 ASN A  13      -9.835  -4.117   2.543  1.00 82.96           O  \\nATOM    108  N   VAL A  14      -6.519  -3.906  -0.866  1.00 94.83           N  \\nATOM    109  CA  VAL A  14      -5.639  -4.122  -2.011  1.00 94.96           C  \\nATOM    110  C   VAL A  14      -5.317  -5.609  -2.142  1.00 94.97           C  \\nATOM    111  CB  VAL A  14      -4.335  -3.303  -1.886  1.00 94.40           C  \\nATOM    112  O   VAL A  14      -5.607  -6.396  -1.237  1.00 94.51           O  \\nATOM    113  CG1 VAL A  14      -4.639  -1.807  -1.844  1.00 92.07           C  \\nATOM    114  CG2 VAL A  14      -3.555  -3.729  -0.644  1.00 91.72           C  \\nATOM    115  N   ALA A  15      -4.807  -5.984  -3.266  1.00 95.07           N  \\nATOM    116  CA  ALA A  15      -4.357  -7.359  -3.469  1.00 95.25           C  \\nATOM    117  C   ALA A  15      -3.209  -7.705  -2.525  1.00 95.11           C  \\nATOM    118  CB  ALA A  15      -3.931  -7.571  -4.920  1.00 94.13           C  \\nATOM    119  O   ALA A  15      -2.340  -6.870  -2.261  1.00 94.45           O  \\nATOM    120  N   ASP A  16      -3.143  -8.994  -1.982  1.00 93.71           N  \\nATOM    121  CA  ASP A  16      -2.131  -9.455  -1.037  1.00 94.38           C  \\nATOM    122  C   ASP A  16      -0.724  -9.237  -1.590  1.00 93.62           C  \\nATOM    123  CB  ASP A  16      -2.341 -10.934  -0.704  1.00 91.59           C  \\nATOM    124  O   ASP A  16       0.164  -8.769  -0.873  1.00 92.13           O  \\nATOM    125  CG  ASP A  16      -3.557 -11.177   0.173  1.00 77.66           C  \\nATOM    126  OD1 ASP A  16      -4.268 -10.206   0.512  1.00 70.32           O  \\nATOM    127  OD2 ASP A  16      -3.804 -12.348   0.531  1.00 81.09           O  \\nATOM    128  N   GLY A  17      -0.522  -9.632  -2.901  1.00 93.59           N  \\nATOM    129  CA  GLY A  17       0.794  -9.478  -3.502  1.00 93.79           C  \\nATOM    130  C   GLY A  17       1.245  -8.032  -3.587  1.00 93.43           C  \\nATOM    131  O   GLY A  17       2.428  -7.734  -3.413  1.00 92.06           O  \\nATOM    132  N   TYR A  18       0.293  -7.173  -3.926  1.00 94.45           N  \\nATOM    133  CA  TYR A  18       0.585  -5.746  -4.004  1.00 94.85           C  \\nATOM    134  C   TYR A  18       0.945  -5.187  -2.632  1.00 94.25           C  \\nATOM    135  CB  TYR A  18      -0.612  -4.983  -4.580  1.00 93.82           C  \\nATOM    136  O   TYR A  18       1.924  -4.450  -2.493  1.00 93.28           O  \\nATOM    137  CG  TYR A  18      -0.383  -3.496  -4.701  1.00 89.63           C  \\nATOM    138  CD1 TYR A  18       0.679  -2.993  -5.450  1.00 84.12           C  \\nATOM    139  CD2 TYR A  18      -1.226  -2.591  -4.065  1.00 84.96           C  \\nATOM    140  CE1 TYR A  18       0.895  -1.624  -5.562  1.00 87.45           C  \\nATOM    141  CE2 TYR A  18      -1.019  -1.219  -4.171  1.00 87.89           C  \\nATOM    142  OH  TYR A  18       0.251   0.611  -5.028  1.00 72.19           O  \\nATOM    143  CZ  TYR A  18       0.042  -0.746  -4.920  1.00 87.01           C  \\nATOM    144  N   ALA A  19       0.340  -5.497  -1.670  1.00 94.99           N  \\nATOM    145  CA  ALA A  19       0.635  -5.070  -0.305  1.00 95.31           C  \\nATOM    146  C   ALA A  19       2.002  -5.576   0.145  1.00 95.03           C  \\nATOM    147  CB  ALA A  19      -0.451  -5.559   0.651  1.00 94.57           C  \\nATOM    148  O   ALA A  19       2.845  -4.793   0.590  1.00 94.36           O  \\nATOM    149  N   ASN A  20       2.263  -6.919  -0.012  1.00 95.00           N  \\nATOM    150  CA  ASN A  20       3.420  -7.570   0.593  1.00 94.99           C  \\nATOM    151  C   ASN A  20       4.692  -7.312  -0.211  1.00 94.82           C  \\nATOM    152  CB  ASN A  20       3.177  -9.074   0.734  1.00 94.06           C  \\nATOM    153  O   ASN A  20       5.782  -7.222   0.357  1.00 94.01           O  \\nATOM    154  CG  ASN A  20       2.135  -9.401   1.786  1.00 89.82           C  \\nATOM    155  ND2 ASN A  20       1.319 -10.413   1.517  1.00 88.19           N  \\nATOM    156  OD1 ASN A  20       2.064  -8.749   2.831  1.00 86.47           O  \\nATOM    157  N   ASN A  21       4.569  -7.240  -1.539  1.00 94.62           N  \\nATOM    158  CA  ASN A  21       5.761  -7.163  -2.377  1.00 94.54           C  \\nATOM    159  C   ASN A  21       6.103  -5.720  -2.736  1.00 94.35           C  \\nATOM    160  CB  ASN A  21       5.579  -7.997  -3.647  1.00 93.38           C  \\nATOM    161  O   ASN A  21       7.183  -5.446  -3.262  1.00 93.10           O  \\nATOM    162  CG  ASN A  21       5.445  -9.479  -3.359  1.00 88.98           C  \\nATOM    163  ND2 ASN A  21       4.502 -10.131  -4.029  1.00 84.70           N  \\nATOM    164  OD1 ASN A  21       6.183 -10.034  -2.541  1.00 83.89           O  \\nATOM    165  N   PHE A  22       5.284  -4.782  -2.438  1.00 95.28           N  \\nATOM    166  CA  PHE A  22       5.556  -3.400  -2.816  1.00 95.20           C  \\nATOM    167  C   PHE A  22       5.266  -2.455  -1.656  1.00 95.22           C  \\nATOM    168  CB  PHE A  22       4.722  -3.002  -4.038  1.00 94.42           C  \\nATOM    169  O   PHE A  22       6.180  -1.830  -1.115  1.00 94.28           O  \\nATOM    170  CG  PHE A  22       5.048  -1.635  -4.575  1.00 91.44           C  \\nATOM    171  CD1 PHE A  22       6.175  -1.432  -5.362  1.00 86.97           C  \\nATOM    172  CD2 PHE A  22       4.225  -0.551  -4.294  1.00 87.04           C  \\nATOM    173  CE1 PHE A  22       6.479  -0.167  -5.860  1.00 84.67           C  \\nATOM    174  CE2 PHE A  22       4.523   0.715  -4.789  1.00 84.37           C  \\nATOM    175  CZ  PHE A  22       5.650   0.905  -5.571  1.00 83.94           C  \\nATOM    176  N   LEU A  23       4.075  -2.403  -1.270  1.00 94.36           N  \\nATOM    177  CA  LEU A  23       3.685  -1.347  -0.342  1.00 94.82           C  \\nATOM    178  C   LEU A  23       4.481  -1.441   0.955  1.00 94.51           C  \\nATOM    179  CB  LEU A  23       2.186  -1.423  -0.042  1.00 94.06           C  \\nATOM    180  O   LEU A  23       5.087  -0.459   1.389  1.00 93.71           O  \\nATOM    181  CG  LEU A  23       1.244  -1.080  -1.197  1.00 91.49           C  \\nATOM    182  CD1 LEU A  23      -0.202  -1.363  -0.804  1.00 87.26           C  \\nATOM    183  CD2 LEU A  23       1.416   0.377  -1.613  1.00 87.61           C  \\nATOM    184  N   PHE A  24       4.495  -2.571   1.600  1.00 94.70           N  \\nATOM    185  CA  PHE A  24       5.171  -2.717   2.883  1.00 95.21           C  \\nATOM    186  C   PHE A  24       6.683  -2.621   2.713  1.00 94.88           C  \\nATOM    187  CB  PHE A  24       4.800  -4.050   3.540  1.00 94.39           C  \\nATOM    188  O   PHE A  24       7.368  -2.012   3.537  1.00 94.23           O  \\nATOM    189  CG  PHE A  24       3.350  -4.151   3.932  1.00 90.66           C  \\nATOM    190  CD1 PHE A  24       2.599  -3.006   4.172  1.00 85.07           C  \\nATOM    191  CD2 PHE A  24       2.738  -5.391   4.060  1.00 85.54           C  \\nATOM    192  CE1 PHE A  24       1.258  -3.097   4.535  1.00 88.82           C  \\nATOM    193  CE2 PHE A  24       1.398  -5.489   4.423  1.00 88.82           C  \\nATOM    194  CZ  PHE A  24       0.660  -4.341   4.660  1.00 88.24           C  \\nATOM    195  N   LYS A  25       7.235  -3.188   1.605  1.00 95.64           N  \\nATOM    196  CA  LYS A  25       8.674  -3.163   1.362  1.00 95.45           C  \\nATOM    197  C   LYS A  25       9.172  -1.736   1.152  1.00 95.33           C  \\nATOM    198  CB  LYS A  25       9.028  -4.025   0.149  1.00 94.37           C  \\nATOM    199  O   LYS A  25      10.273  -1.388   1.583  1.00 94.52           O  \\nATOM    200  CG  LYS A  25       8.931  -5.522   0.403  1.00 85.04           C  \\nATOM    201  CD  LYS A  25       9.508  -6.325  -0.756  1.00 79.39           C  \\nATOM    202  CE  LYS A  25       9.465  -7.821  -0.477  1.00 74.28           C  \\nATOM    203  NZ  LYS A  25      10.062  -8.610  -1.596  1.00 67.90           N  \\nATOM    204  N   GLN A  26       8.329  -0.912   0.584  1.00 94.91           N  \\nATOM    205  CA  GLN A  26       8.710   0.465   0.288  1.00 94.51           C  \\nATOM    206  C   GLN A  26       8.321   1.401   1.428  1.00 93.93           C  \\nATOM    207  CB  GLN A  26       8.064   0.932  -1.018  1.00 92.90           C  \\nATOM    208  O   GLN A  26       8.567   2.607   1.360  1.00 92.08           O  \\nATOM    209  CG  GLN A  26       8.520   0.152  -2.243  1.00 82.40           C  \\nATOM    210  CD  GLN A  26       9.949   0.470  -2.643  1.00 75.27           C  \\nATOM    211  NE2 GLN A  26      10.596  -0.468  -3.327  1.00 67.52           N  \\nATOM    212  OE1 GLN A  26      10.467   1.550  -2.340  1.00 75.38           O  \\nATOM    213  N   GLY A  27       7.721   0.768   2.494  1.00 94.56           N  \\nATOM    214  CA  GLY A  27       7.306   1.596   3.615  1.00 94.85           C  \\nATOM    215  C   GLY A  27       6.184   2.555   3.266  1.00 93.98           C  \\nATOM    216  O   GLY A  27       6.104   3.654   3.818  1.00 91.78           O  \\nATOM    217  N   LEU A  28       5.281   2.199   2.389  1.00 94.57           N  \\nATOM    218  CA  LEU A  28       4.263   3.104   1.865  1.00 94.67           C  \\nATOM    219  C   LEU A  28       2.913   2.843   2.524  1.00 94.11           C  \\nATOM    220  CB  LEU A  28       4.141   2.952   0.347  1.00 93.74           C  \\nATOM    221  O   LEU A  28       2.001   3.668   2.429  1.00 92.77           O  \\nATOM    222  CG  LEU A  28       5.340   3.418  -0.480  1.00 89.38           C  \\nATOM    223  CD1 LEU A  28       5.146   3.053  -1.948  1.00 81.99           C  \\nATOM    224  CD2 LEU A  28       5.550   4.919  -0.320  1.00 82.69           C  \\nATOM    225  N   ALA A  29       2.838   1.830   3.201  1.00 93.57           N  \\nATOM    226  CA  ALA A  29       1.608   1.418   3.873  1.00 94.46           C  \\nATOM    227  C   ALA A  29       1.902   0.426   4.995  1.00 93.52           C  \\nATOM    228  CB  ALA A  29       0.632   0.809   2.870  1.00 92.74           C  \\nATOM    229  O   ALA A  29       2.991  -0.150   5.053  1.00 92.09           O  \\nATOM    230  N   ILE A  30       0.899   0.249   6.011  1.00 95.49           N  \\nATOM    231  CA  ILE A  30       0.916  -0.803   7.022  1.00 95.59           C  \\nATOM    232  C   ILE A  30      -0.419  -1.545   7.018  1.00 95.47           C  \\nATOM    233  CB  ILE A  30       1.206  -0.230   8.428  1.00 94.89           C  \\nATOM    234  O   ILE A  30      -1.404  -1.059   6.457  1.00 94.99           O  \\nATOM    235  CG1 ILE A  30       0.095   0.738   8.850  1.00 91.87           C  \\nATOM    236  CG2 ILE A  30       2.573   0.458   8.459  1.00 91.59           C  \\nATOM    237  CD1 ILE A  30       0.166   1.164  10.310  1.00 88.23           C  \\nATOM    238  N   GLU A  31      -0.393  -2.723   7.484  1.00 94.94           N  \\nATOM    239  CA  GLU A  31      -1.638  -3.477   7.599  1.00 95.11           C  \\nATOM    240  C   GLU A  31      -2.583  -2.832   8.608  1.00 94.81           C  \\nATOM    241  CB  GLU A  31      -1.353  -4.928   7.996  1.00 93.80           C  \\nATOM    242  O   GLU A  31      -2.158  -2.424   9.691  1.00 93.93           O  \\nATOM    243  CG  GLU A  31      -2.584  -5.823   7.977  1.00 86.31           C  \\nATOM    244  CD  GLU A  31      -2.276  -7.272   8.317  1.00 83.42           C  \\nATOM    245  OE1 GLU A  31      -1.104  -7.589   8.621  1.00 84.59           O  \\nATOM    246  OE2 GLU A  31      -3.215  -8.098   8.279  1.00 84.56           O  \\nATOM    247  N   ALA A  32      -3.860  -2.647   8.224  1.00 94.97           N  \\nATOM    248  CA  ALA A  32      -4.867  -2.206   9.185  1.00 94.60           C  \\nATOM    249  C   ALA A  32      -5.276  -3.347  10.113  1.00 93.71           C  \\nATOM    250  CB  ALA A  32      -6.090  -1.651   8.459  1.00 93.35           C  \\nATOM    251  O   ALA A  32      -5.239  -4.516   9.723  1.00 91.28           O  \\nTER     252      ALA A  32\\nEND\\n\",\"pdb\");\n\tviewer_17580091717215044.setStyle({\"model\": 0},{\"cartoon\": {\"color\": \"green\", \"opacity\": 0.7}});\n\tviewer_17580091717215044.addModel(\"ATOM      1  N   MET A   1      -6.473  -7.251   1.654  1.00 91.96           N  \\nATOM      2  CA  MET A   1      -5.833  -6.977   2.937  1.00 91.96           C  \\nATOM      3  C   MET A   1      -5.912  -5.493   3.280  1.00 91.96           C  \\nATOM      4  O   MET A   1      -5.650  -4.641   2.429  1.00 91.96           O  \\nATOM      5  CB  MET A   1      -4.372  -7.432   2.919  1.00 91.96           C  \\nATOM      6  CG  MET A   1      -3.612  -7.107   4.193  1.00 91.96           C  \\nATOM      7  SD  MET A   1      -1.843  -7.584   4.099  1.00 91.96           S  \\nATOM      8  CE  MET A   1      -1.960  -9.337   4.552  1.00 91.96           C  \\nATOM      9  N   LYS A   2      -6.632  -5.346   4.185  1.00 93.29           N  \\nATOM     10  CA  LYS A   2      -6.861  -3.982   4.650  1.00 93.29           C  \\nATOM     11  C   LYS A   2      -5.555  -3.325   5.088  1.00 93.29           C  \\nATOM     12  O   LYS A   2      -4.845  -3.853   5.946  1.00 93.29           O  \\nATOM     13  CB  LYS A   2      -7.866  -3.967   5.802  1.00 93.29           C  \\nATOM     14  CG  LYS A   2      -8.503  -2.608   6.052  1.00 93.29           C  \\nATOM     15  CD  LYS A   2      -9.632  -2.698   7.069  1.00 93.29           C  \\nATOM     16  CE  LYS A   2     -10.255  -1.334   7.337  1.00 93.29           C  \\nATOM     17  NZ  LYS A   2     -11.368  -1.419   8.330  1.00 93.29           N  \\nATOM     18  N   VAL A   3      -5.215  -2.256   4.311  1.00 94.76           N  \\nATOM     19  CA  VAL A   3      -3.983  -1.559   4.666  1.00 94.76           C  \\nATOM     20  C   VAL A   3      -4.261  -0.067   4.828  1.00 94.76           C  \\nATOM     21  O   VAL A   3      -5.256   0.445   4.310  1.00 94.76           O  \\nATOM     22  CB  VAL A   3      -2.880  -1.784   3.608  1.00 94.76           C  \\nATOM     23  CG1 VAL A   3      -2.540  -3.269   3.488  1.00 94.76           C  \\nATOM     24  CG2 VAL A   3      -3.314  -1.221   2.256  1.00 94.76           C  \\nATOM     25  N   ILE A   4      -3.392   0.660   5.555  1.00 95.11           N  \\nATOM     26  CA  ILE A   4      -3.429   2.108   5.722  1.00 95.11           C  \\nATOM     27  C   ILE A   4      -2.257   2.743   4.976  1.00 95.11           C  \\nATOM     28  O   ILE A   4      -1.096   2.430   5.250  1.00 95.11           O  \\nATOM     29  CB  ILE A   4      -3.394   2.506   7.214  1.00 95.11           C  \\nATOM     30  CG1 ILE A   4      -4.562   1.860   7.968  1.00 95.11           C  \\nATOM     31  CG2 ILE A   4      -3.419   4.030   7.367  1.00 95.11           C  \\nATOM     32  CD1 ILE A   4      -4.490   2.024   9.479  1.00 95.11           C  \\nATOM     33  N   PHE A   5      -2.509   3.598   4.050  1.00 93.85           N  \\nATOM     34  CA  PHE A   5      -1.483   4.360   3.349  1.00 93.85           C  \\nATOM     35  C   PHE A   5      -0.955   5.490   4.224  1.00 93.85           C  \\nATOM     36  O   PHE A   5      -1.714   6.370   4.635  1.00 93.85           O  \\nATOM     37  CB  PHE A   5      -2.036   4.927   2.037  1.00 93.85           C  \\nATOM     38  CG  PHE A   5      -2.328   3.879   0.997  1.00 93.85           C  \\nATOM     39  CD1 PHE A   5      -1.395   3.576   0.014  1.00 93.85           C  \\nATOM     40  CD2 PHE A   5      -3.537   3.195   1.004  1.00 93.85           C  \\nATOM     41  CE1 PHE A   5      -1.663   2.606  -0.949  1.00 93.85           C  \\nATOM     42  CE2 PHE A   5      -3.812   2.225   0.046  1.00 93.85           C  \\nATOM     43  CZ  PHE A   5      -2.874   1.933  -0.931  1.00 93.85           C  \\nATOM     44  N   LEU A   6       0.309   5.442   4.468  1.00 92.63           N  \\nATOM     45  CA  LEU A   6       0.906   6.315   5.472  1.00 92.63           C  \\nATOM     46  C   LEU A   6       0.898   7.767   5.006  1.00 92.63           C  \\nATOM     47  O   LEU A   6       0.618   8.675   5.791  1.00 92.63           O  \\nATOM     48  CB  LEU A   6       2.340   5.875   5.782  1.00 92.63           C  \\nATOM     49  CG  LEU A   6       2.499   4.515   6.463  1.00 92.63           C  \\nATOM     50  CD1 LEU A   6       3.975   4.152   6.583  1.00 92.63           C  \\nATOM     51  CD2 LEU A   6       1.831   4.521   7.834  1.00 92.63           C  \\nATOM     52  N   LYS A   7       1.006   7.997   3.747  1.00 90.46           N  \\nATOM     53  CA  LYS A   7       1.064   9.366   3.243  1.00 90.46           C  \\nATOM     54  C   LYS A   7      -0.336   9.950   3.075  1.00 90.46           C  \\nATOM     55  O   LYS A   7      -0.558  11.132   3.345  1.00 90.46           O  \\nATOM     56  CB  LYS A   7       1.815   9.416   1.911  1.00 90.46           C  \\nATOM     57  CG  LYS A   7       3.318   9.217   2.040  1.00 90.46           C  \\nATOM     58  CD  LYS A   7       4.024   9.400   0.703  1.00 90.46           C  \\nATOM     59  CE  LYS A   7       5.525   9.179   0.827  1.00 90.46           C  \\nATOM     60  NZ  LYS A   7       6.210   9.272  -0.498  1.00 90.46           N  \\nATOM     61  N   LYS A   8      -1.239   9.124   2.770  1.00 89.16           N  \\nATOM     62  CA  LYS A   8      -2.593   9.599   2.506  1.00 89.16           C  \\nATOM     63  C   LYS A   8      -3.502   9.370   3.710  1.00 89.16           C  \\nATOM     64  O   LYS A   8      -4.460  10.116   3.922  1.00 89.16           O  \\nATOM     65  CB  LYS A   8      -3.174   8.906   1.272  1.00 89.16           C  \\nATOM     66  CG  LYS A   8      -2.599   9.405  -0.046  1.00 89.16           C  \\nATOM     67  CD  LYS A   8      -3.314   8.782  -1.238  1.00 89.16           C  \\nATOM     68  CE  LYS A   8      -2.698   9.231  -2.557  1.00 89.16           C  \\nATOM     69  NZ  LYS A   8      -3.372   8.593  -3.728  1.00 89.16           N  \\nATOM     70  N   GLY A   9      -3.090   8.432   4.583  1.00 92.79           N  \\nATOM     71  CA  GLY A   9      -3.923   8.147   5.740  1.00 92.79           C  \\nATOM     72  C   GLY A   9      -5.217   7.439   5.385  1.00 92.79           C  \\nATOM     73  O   GLY A   9      -6.249   7.665   6.019  1.00 92.79           O  \\nATOM     74  N   GLU A  10      -5.272   6.818   4.294  1.00 92.57           N  \\nATOM     75  CA  GLU A  10      -6.472   6.137   3.817  1.00 92.57           C  \\nATOM     76  C   GLU A  10      -6.337   4.623   3.952  1.00 92.57           C  \\nATOM     77  O   GLU A  10      -5.232   4.083   3.876  1.00 92.57           O  \\nATOM     78  CB  GLU A  10      -6.759   6.510   2.361  1.00 92.57           C  \\nATOM     79  CG  GLU A  10      -7.166   7.965   2.167  1.00 92.57           C  \\nATOM     80  CD  GLU A  10      -7.526   8.301   0.729  1.00 92.57           C  \\nATOM     81  OE1 GLU A  10      -8.011   9.426   0.472  1.00 92.57           O  \\nATOM     82  OE2 GLU A  10      -7.324   7.431  -0.148  1.00 92.57           O  \\nATOM     83  N   ILE A  11      -7.531   4.069   4.140  1.00 95.62           N  \\nATOM     84  CA  ILE A  11      -7.592   2.613   4.208  1.00 95.62           C  \\nATOM     85  C   ILE A  11      -8.180   2.061   2.911  1.00 95.62           C  \\nATOM     86  O   ILE A  11      -9.250   2.492   2.474  1.00 95.62           O  \\nATOM     87  CB  ILE A  11      -8.426   2.138   5.419  1.00 95.62           C  \\nATOM     88  CG1 ILE A  11      -7.850   2.710   6.720  1.00 95.62           C  \\nATOM     89  CG2 ILE A  11      -8.479   0.608   5.470  1.00 95.62           C  \\nATOM     90  CD1 ILE A  11      -8.704   2.430   7.949  1.00 95.62           C  \\nATOM     91  N   LYS A  12      -7.476   1.182   2.319  1.00 94.49           N  \\nATOM     92  CA  LYS A  12      -7.927   0.569   1.072  1.00 94.49           C  \\nATOM     93  C   LYS A  12      -7.683  -0.937   1.079  1.00 94.49           C  \\nATOM     94  O   LYS A  12      -6.798  -1.424   1.785  1.00 94.49           O  \\nATOM     95  CB  LYS A  12      -7.224   1.209  -0.126  1.00 94.49           C  \\nATOM     96  CG  LYS A  12      -7.676   2.632  -0.422  1.00 94.49           C  \\nATOM     97  CD  LYS A  12      -7.128   3.126  -1.754  1.00 94.49           C  \\nATOM     98  CE  LYS A  12      -7.656   4.512  -2.097  1.00 94.49           C  \\nATOM     99  NZ  LYS A  12      -7.107   5.009  -3.395  1.00 94.49           N  \\nATOM    100  N   ASN A  13      -8.560  -1.585   0.329  1.00 96.14           N  \\nATOM    101  CA  ASN A  13      -8.325  -3.004   0.081  1.00 96.14           C  \\nATOM    102  C   ASN A  13      -7.477  -3.222  -1.168  1.00 96.14           C  \\nATOM    103  O   ASN A  13      -7.810  -2.722  -2.244  1.00 96.14           O  \\nATOM    104  CB  ASN A  13      -9.652  -3.755  -0.041  1.00 96.14           C  \\nATOM    105  CG  ASN A  13     -10.336  -3.953   1.298  1.00 96.14           C  \\nATOM    106  OD1 ASN A  13      -9.744  -4.492   2.238  1.00 96.14           O  \\nATOM    107  ND2 ASN A  13     -11.587  -3.519   1.396  1.00 96.14           N  \\nATOM    108  N   VAL A  14      -6.463  -3.833  -0.970  1.00 95.82           N  \\nATOM    109  CA  VAL A  14      -5.603  -4.119  -2.114  1.00 95.82           C  \\nATOM    110  C   VAL A  14      -5.335  -5.620  -2.198  1.00 95.82           C  \\nATOM    111  O   VAL A  14      -5.637  -6.363  -1.262  1.00 95.82           O  \\nATOM    112  CB  VAL A  14      -4.269  -3.342  -2.029  1.00 95.82           C  \\nATOM    113  CG1 VAL A  14      -4.514  -1.837  -2.109  1.00 95.82           C  \\nATOM    114  CG2 VAL A  14      -3.527  -3.699  -0.743  1.00 95.82           C  \\nATOM    115  N   ALA A  15      -4.709  -6.094  -3.310  1.00 96.50           N  \\nATOM    116  CA  ALA A  15      -4.286  -7.484  -3.455  1.00 96.50           C  \\nATOM    117  C   ALA A  15      -3.159  -7.820  -2.484  1.00 96.50           C  \\nATOM    118  O   ALA A  15      -2.325  -6.967  -2.173  1.00 96.50           O  \\nATOM    119  CB  ALA A  15      -3.848  -7.759  -4.892  1.00 96.50           C  \\nATOM    120  N   ASP A  16      -3.085  -9.029  -2.019  1.00 95.07           N  \\nATOM    121  CA  ASP A  16      -2.104  -9.476  -1.033  1.00 95.07           C  \\nATOM    122  C   ASP A  16      -0.681  -9.299  -1.557  1.00 95.07           C  \\nATOM    123  O   ASP A  16       0.191  -8.798  -0.845  1.00 95.07           O  \\nATOM    124  CB  ASP A  16      -2.349 -10.938  -0.657  1.00 95.07           C  \\nATOM    125  CG  ASP A  16      -3.589 -11.129   0.199  1.00 95.07           C  \\nATOM    126  OD1 ASP A  16      -4.039 -10.160   0.847  1.00 95.07           O  \\nATOM    127  OD2 ASP A  16      -4.119 -12.261   0.228  1.00 95.07           O  \\nATOM    128  N   GLY A  17      -0.520  -9.702  -2.879  1.00 96.41           N  \\nATOM    129  CA  GLY A  17       0.803  -9.617  -3.477  1.00 96.41           C  \\nATOM    130  C   GLY A  17       1.304  -8.192  -3.618  1.00 96.41           C  \\nATOM    131  O   GLY A  17       2.471  -7.909  -3.340  1.00 96.41           O  \\nATOM    132  N   TYR A  18       0.390  -7.276  -4.001  1.00 96.10           N  \\nATOM    133  CA  TYR A  18       0.714  -5.861  -4.135  1.00 96.10           C  \\nATOM    134  C   TYR A  18       1.035  -5.246  -2.778  1.00 96.10           C  \\nATOM    135  O   TYR A  18       2.029  -4.531  -2.633  1.00 96.10           O  \\nATOM    136  CB  TYR A  18      -0.445  -5.103  -4.790  1.00 96.10           C  \\nATOM    137  CG  TYR A  18      -0.172  -3.632  -4.990  1.00 96.10           C  \\nATOM    138  CD1 TYR A  18      -0.937  -2.668  -4.337  1.00 96.10           C  \\nATOM    139  CD2 TYR A  18       0.848  -3.203  -5.832  1.00 96.10           C  \\nATOM    140  CE1 TYR A  18      -0.692  -1.311  -4.518  1.00 96.10           C  \\nATOM    141  CE2 TYR A  18       1.102  -1.849  -6.022  1.00 96.10           C  \\nATOM    142  CZ  TYR A  18       0.329  -0.912  -5.360  1.00 96.10           C  \\nATOM    143  OH  TYR A  18       0.577   0.430  -5.545  1.00 96.10           O  \\nATOM    144  N   ALA A  19       0.312  -5.600  -1.756  1.00 96.34           N  \\nATOM    145  CA  ALA A  19       0.541  -5.085  -0.408  1.00 96.34           C  \\nATOM    146  C   ALA A  19       1.898  -5.534   0.126  1.00 96.34           C  \\nATOM    147  O   ALA A  19       2.697  -4.711   0.580  1.00 96.34           O  \\nATOM    148  CB  ALA A  19      -0.573  -5.538   0.531  1.00 96.34           C  \\nATOM    149  N   ASN A  20       2.264  -6.774  -0.059  1.00 95.93           N  \\nATOM    150  CA  ASN A  20       3.451  -7.360   0.556  1.00 95.93           C  \\nATOM    151  C   ASN A  20       4.718  -6.994  -0.212  1.00 95.93           C  \\nATOM    152  O   ASN A  20       5.747  -6.686   0.392  1.00 95.93           O  \\nATOM    153  CB  ASN A  20       3.309  -8.880   0.654  1.00 95.93           C  \\nATOM    154  CG  ASN A  20       2.290  -9.306   1.692  1.00 95.93           C  \\nATOM    155  OD1 ASN A  20       2.196  -8.712   2.769  1.00 95.93           O  \\nATOM    156  ND2 ASN A  20       1.516 -10.338   1.375  1.00 95.93           N  \\nATOM    157  N   ASN A  21       4.546  -6.914  -1.585  1.00 96.23           N  \\nATOM    158  CA  ASN A  21       5.752  -6.789  -2.397  1.00 96.23           C  \\nATOM    159  C   ASN A  21       6.016  -5.337  -2.789  1.00 96.23           C  \\nATOM    160  O   ASN A  21       7.120  -4.997  -3.217  1.00 96.23           O  \\nATOM    161  CB  ASN A  21       5.653  -7.667  -3.645  1.00 96.23           C  \\nATOM    162  CG  ASN A  21       5.723  -9.148  -3.325  1.00 96.23           C  \\nATOM    163  OD1 ASN A  21       6.552  -9.582  -2.521  1.00 96.23           O  \\nATOM    164  ND2 ASN A  21       4.854  -9.932  -3.951  1.00 96.23           N  \\nATOM    165  N   PHE A  22       5.015  -4.516  -2.622  1.00 96.02           N  \\nATOM    166  CA  PHE A  22       5.190  -3.134  -3.052  1.00 96.02           C  \\nATOM    167  C   PHE A  22       4.943  -2.172  -1.894  1.00 96.02           C  \\nATOM    168  O   PHE A  22       5.832  -1.406  -1.518  1.00 96.02           O  \\nATOM    169  CB  PHE A  22       4.250  -2.807  -4.216  1.00 96.02           C  \\nATOM    170  CG  PHE A  22       4.456  -1.435  -4.797  1.00 96.02           C  \\nATOM    171  CD1 PHE A  22       3.566  -0.405  -4.515  1.00 96.02           C  \\nATOM    172  CD2 PHE A  22       5.540  -1.174  -5.624  1.00 96.02           C  \\nATOM    173  CE1 PHE A  22       3.755   0.867  -5.052  1.00 96.02           C  \\nATOM    174  CE2 PHE A  22       5.734   0.094  -6.164  1.00 96.02           C  \\nATOM    175  CZ  PHE A  22       4.841   1.113  -5.877  1.00 96.02           C  \\nATOM    176  N   LEU A  23       3.902  -2.264  -1.205  1.00 95.97           N  \\nATOM    177  CA  LEU A  23       3.506  -1.248  -0.236  1.00 95.97           C  \\nATOM    178  C   LEU A  23       4.270  -1.417   1.072  1.00 95.97           C  \\nATOM    179  O   LEU A  23       4.861  -0.459   1.579  1.00 95.97           O  \\nATOM    180  CB  LEU A  23       1.999  -1.318   0.027  1.00 95.97           C  \\nATOM    181  CG  LEU A  23       1.087  -0.968  -1.150  1.00 95.97           C  \\nATOM    182  CD1 LEU A  23      -0.374  -1.184  -0.771  1.00 95.97           C  \\nATOM    183  CD2 LEU A  23       1.321   0.471  -1.599  1.00 95.97           C  \\nATOM    184  N   PHE A  24       4.313  -2.597   1.609  1.00 95.80           N  \\nATOM    185  CA  PHE A  24       4.995  -2.797   2.883  1.00 95.80           C  \\nATOM    186  C   PHE A  24       6.507  -2.733   2.703  1.00 95.80           C  \\nATOM    187  O   PHE A  24       7.213  -2.166   3.539  1.00 95.80           O  \\nATOM    188  CB  PHE A  24       4.597  -4.141   3.502  1.00 95.80           C  \\nATOM    189  CG  PHE A  24       3.167  -4.194   3.967  1.00 95.80           C  \\nATOM    190  CD1 PHE A  24       2.540  -3.059   4.467  1.00 95.80           C  \\nATOM    191  CD2 PHE A  24       2.448  -5.381   3.905  1.00 95.80           C  \\nATOM    192  CE1 PHE A  24       1.216  -3.106   4.898  1.00 95.80           C  \\nATOM    193  CE2 PHE A  24       1.125  -5.435   4.333  1.00 95.80           C  \\nATOM    194  CZ  PHE A  24       0.511  -4.298   4.831  1.00 95.80           C  \\nATOM    195  N   LYS A  25       6.957  -3.265   1.625  1.00 95.47           N  \\nATOM    196  CA  LYS A  25       8.395  -3.295   1.377  1.00 95.47           C  \\nATOM    197  C   LYS A  25       8.957  -1.885   1.225  1.00 95.47           C  \\nATOM    198  O   LYS A  25      10.075  -1.605   1.663  1.00 95.47           O  \\nATOM    199  CB  LYS A  25       8.708  -4.120   0.128  1.00 95.47           C  \\nATOM    200  CG  LYS A  25      10.184  -4.443  -0.048  1.00 95.47           C  \\nATOM    201  CD  LYS A  25      10.402  -5.499  -1.123  1.00 95.47           C  \\nATOM    202  CE  LYS A  25      11.878  -5.841  -1.283  1.00 95.47           C  \\nATOM    203  NZ  LYS A  25      12.084  -6.978  -2.230  1.00 95.47           N  \\nATOM    204  N   GLN A  26       8.123  -1.009   0.673  1.00 94.66           N  \\nATOM    205  CA  GLN A  26       8.573   0.362   0.455  1.00 94.66           C  \\nATOM    206  C   GLN A  26       8.179   1.262   1.623  1.00 94.66           C  \\nATOM    207  O   GLN A  26       8.524   2.445   1.645  1.00 94.66           O  \\nATOM    208  CB  GLN A  26       7.998   0.915  -0.849  1.00 94.66           C  \\nATOM    209  CG  GLN A  26       8.512   0.208  -2.097  1.00 94.66           C  \\nATOM    210  CD  GLN A  26      10.015   0.328  -2.262  1.00 94.66           C  \\nATOM    211  OE1 GLN A  26      10.605   1.369  -1.958  1.00 94.66           O  \\nATOM    212  NE2 GLN A  26      10.646  -0.738  -2.744  1.00 94.66           N  \\nATOM    213  N   GLY A  27       7.449   0.698   2.507  1.00 94.71           N  \\nATOM    214  CA  GLY A  27       7.056   1.497   3.658  1.00 94.71           C  \\nATOM    215  C   GLY A  27       5.935   2.473   3.352  1.00 94.71           C  \\nATOM    216  O   GLY A  27       5.877   3.559   3.930  1.00 94.71           O  \\nATOM    217  N   LEU A  28       5.171   2.191   2.316  1.00 94.13           N  \\nATOM    218  CA  LEU A  28       4.115   3.091   1.864  1.00 94.13           C  \\nATOM    219  C   LEU A  28       2.813   2.820   2.610  1.00 94.13           C  \\nATOM    220  O   LEU A  28       1.934   3.682   2.669  1.00 94.13           O  \\nATOM    221  CB  LEU A  28       3.893   2.944   0.356  1.00 94.13           C  \\nATOM    222  CG  LEU A  28       5.053   3.371  -0.545  1.00 94.13           C  \\nATOM    223  CD1 LEU A  28       4.770   2.984  -1.992  1.00 94.13           C  \\nATOM    224  CD2 LEU A  28       5.300   4.871  -0.424  1.00 94.13           C  \\nATOM    225  N   ALA A  29       2.770   1.593   3.176  1.00 95.99           N  \\nATOM    226  CA  ALA A  29       1.528   1.238   3.857  1.00 95.99           C  \\nATOM    227  C   ALA A  29       1.793   0.293   5.024  1.00 95.99           C  \\nATOM    228  O   ALA A  29       2.859  -0.320   5.106  1.00 95.99           O  \\nATOM    229  CB  ALA A  29       0.546   0.604   2.875  1.00 95.99           C  \\nATOM    230  N   ILE A  30       0.948   0.243   6.007  1.00 95.27           N  \\nATOM    231  CA  ILE A  30       0.979  -0.690   7.127  1.00 95.27           C  \\nATOM    232  C   ILE A  30      -0.315  -1.501   7.159  1.00 95.27           C  \\nATOM    233  O   ILE A  30      -1.296  -1.142   6.503  1.00 95.27           O  \\nATOM    234  CB  ILE A  30       1.182   0.046   8.471  1.00 95.27           C  \\nATOM    235  CG1 ILE A  30       0.075   1.084   8.684  1.00 95.27           C  \\nATOM    236  CG2 ILE A  30       2.565   0.701   8.523  1.00 95.27           C  \\nATOM    237  CD1 ILE A  30       0.082   1.721  10.067  1.00 95.27           C  \\nATOM    238  N   GLU A  31      -0.191  -2.534   7.711  1.00 92.92           N  \\nATOM    239  CA  GLU A  31      -1.386  -3.361   7.844  1.00 92.92           C  \\nATOM    240  C   GLU A  31      -2.475  -2.637   8.631  1.00 92.92           C  \\nATOM    241  O   GLU A  31      -2.210  -2.078   9.696  1.00 92.92           O  \\nATOM    242  CB  GLU A  31      -1.046  -4.692   8.520  1.00 92.92           C  \\nATOM    243  CG  GLU A  31      -0.223  -5.632   7.650  1.00 92.92           C  \\nATOM    244  CD  GLU A  31       0.093  -6.956   8.328  1.00 92.92           C  \\nATOM    245  OE1 GLU A  31       0.650  -7.859   7.665  1.00 92.92           O  \\nATOM    246  OE2 GLU A  31      -0.223  -7.090   9.532  1.00 92.92           O  \\nATOM    247  N   ALA A  32      -3.479  -2.209   7.753  1.00 89.09           N  \\nATOM    248  CA  ALA A  32      -4.626  -1.671   8.477  1.00 89.09           C  \\nATOM    249  C   ALA A  32      -5.173  -2.689   9.474  1.00 89.09           C  \\nATOM    250  O   ALA A  32      -4.987  -3.896   9.303  1.00 89.09           O  \\nATOM    251  CB  ALA A  32      -5.721  -1.247   7.501  1.00 89.09           C  \\nTER     252      ALA A  32                                                       \\nEND   \\n\",\"pdb\");\n\tviewer_17580091717215044.setStyle({\"model\": 1, \"cartoon\": {\"colorscheme\": {\"prop\": \"b\", \"gradient\": \"roygb\", \"min\": 50, \"max\": 90}}});\n\tviewer_17580091717215044.addStyle({\"model\": 1, \"stick\": {}});\n\tviewer_17580091717215044.zoomTo();\nviewer_17580091717215044.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_17580091717215044\"  style=\"position: relative; width: 800px; height: 480px;\">\n",
              "        <p id=\"3dmolwarning_17580091717215044\" style=\"background-color:#ffcccc;color:black\">3Dmol.js failed to load for some reason.  Please check your browser console for error messages.<br></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    //this is to ignore the existence of requirejs amd\n",
              "    var savedexports, savedmodule;\n",
              "    if (typeof exports !== 'undefined') savedexports = exports;\n",
              "    else exports = {}\n",
              "    if (typeof module !== 'undefined') savedmodule = module;\n",
              "    else module = {}\n",
              "\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "        exports = savedexports;\n",
              "        module = savedmodule;\n",
              "        resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_17580091717215044 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_17580091717215044\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_17580091717215044 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_17580091717215044\"),{backgroundColor:\"white\"});\n",
              "viewer_17580091717215044.zoomTo();\n",
              "\tviewer_17580091717215044.addModel(\"PARENT N/A\\nATOM      1  N   MET A   1      -6.170  -7.770   1.344  1.00 93.76           N  \\nATOM      2  CA  MET A   1      -5.898  -7.356   2.717  1.00 94.04           C  \\nATOM      3  C   MET A   1      -6.200  -5.874   2.909  1.00 94.69           C  \\nATOM      4  CB  MET A   1      -4.442  -7.646   3.088  1.00 91.60           C  \\nATOM      5  O   MET A   1      -6.172  -5.101   1.950  1.00 93.73           O  \\nATOM      6  CG  MET A   1      -3.436  -6.783   2.343  1.00 80.00           C  \\nATOM      7  SD  MET A   1      -1.700  -7.222   2.743  1.00 70.55           S  \\nATOM      8  CE  MET A   1      -1.661  -8.939   2.156  1.00 69.74           C  \\nATOM      9  N   LYS A   2      -6.599  -5.431   3.986  1.00 95.80           N  \\nATOM     10  CA  LYS A   2      -6.886  -4.054   4.376  1.00 96.00           C  \\nATOM     11  C   LYS A   2      -5.621  -3.340   4.844  1.00 95.94           C  \\nATOM     12  CB  LYS A   2      -7.947  -4.017   5.478  1.00 94.97           C  \\nATOM     13  O   LYS A   2      -4.883  -3.861   5.683  1.00 95.43           O  \\nATOM     14  CG  LYS A   2      -8.489  -2.625   5.769  1.00 86.65           C  \\nATOM     15  CD  LYS A   2      -9.599  -2.665   6.812  1.00 82.11           C  \\nATOM     16  CE  LYS A   2     -10.196  -1.284   7.046  1.00 80.93           C  \\nATOM     17  NZ  LYS A   2     -11.327  -1.328   8.020  1.00 70.78           N  \\nATOM     18  N   VAL A   3      -5.353  -2.202   4.207  1.00 95.03           N  \\nATOM     19  CA  VAL A   3      -4.110  -1.508   4.526  1.00 95.38           C  \\nATOM     20  C   VAL A   3      -4.400  -0.037   4.819  1.00 94.98           C  \\nATOM     21  CB  VAL A   3      -3.080  -1.630   3.381  1.00 94.41           C  \\nATOM     22  O   VAL A   3      -5.432   0.493   4.403  1.00 94.15           O  \\nATOM     23  CG1 VAL A   3      -2.753  -3.096   3.104  1.00 85.54           C  \\nATOM     24  CG2 VAL A   3      -3.604  -0.949   2.118  1.00 84.45           C  \\nATOM     25  N   ILE A   4      -3.530   0.635   5.581  1.00 95.05           N  \\nATOM     26  CA  ILE A   4      -3.494   2.078   5.794  1.00 95.30           C  \\nATOM     27  C   ILE A   4      -2.356   2.692   4.981  1.00 95.22           C  \\nATOM     28  CB  ILE A   4      -3.330   2.424   7.291  1.00 94.33           C  \\nATOM     29  O   ILE A   4      -1.182   2.416   5.238  1.00 94.62           O  \\nATOM     30  CG1 ILE A   4      -4.429   1.748   8.119  1.00 88.67           C  \\nATOM     31  CG2 ILE A   4      -3.341   3.941   7.500  1.00 88.04           C  \\nATOM     32  CD1 ILE A   4      -4.165   1.752   9.619  1.00 82.29           C  \\nATOM     33  N   PHE A   5      -2.663   3.500   3.947  1.00 93.06           N  \\nATOM     34  CA  PHE A   5      -1.646   4.228   3.197  1.00 94.01           C  \\nATOM     35  C   PHE A   5      -1.079   5.374   4.027  1.00 93.25           C  \\nATOM     36  CB  PHE A   5      -2.227   4.768   1.886  1.00 92.21           C  \\nATOM     37  O   PHE A   5      -1.829   6.207   4.541  1.00 92.16           O  \\nATOM     38  CG  PHE A   5      -2.460   3.707   0.844  1.00 85.50           C  \\nATOM     39  CD1 PHE A   5      -1.509   3.457  -0.137  1.00 78.06           C  \\nATOM     40  CD2 PHE A   5      -3.630   2.960   0.846  1.00 79.69           C  \\nATOM     41  CE1 PHE A   5      -1.721   2.476  -1.103  1.00 84.65           C  \\nATOM     42  CE2 PHE A   5      -3.850   1.978  -0.115  1.00 83.91           C  \\nATOM     43  CZ  PHE A   5      -2.895   1.738  -1.090  1.00 84.90           C  \\nATOM     44  N   LEU A   6       0.127   5.418   4.151  1.00 94.20           N  \\nATOM     45  CA  LEU A   6       0.788   6.292   5.114  1.00 93.99           C  \\nATOM     46  C   LEU A   6       0.783   7.737   4.628  1.00 93.50           C  \\nATOM     47  CB  LEU A   6       2.227   5.828   5.359  1.00 93.21           C  \\nATOM     48  O   LEU A   6       0.779   8.668   5.437  1.00 92.13           O  \\nATOM     49  CG  LEU A   6       2.395   4.473   6.047  1.00 90.40           C  \\nATOM     50  CD1 LEU A   6       3.865   4.067   6.068  1.00 86.76           C  \\nATOM     51  CD2 LEU A   6       1.827   4.518   7.462  1.00 86.68           C  \\nATOM     52  N   LYS A   7       0.842   7.971   3.352  1.00 93.29           N  \\nATOM     53  CA  LYS A   7       0.926   9.328   2.821  1.00 92.89           C  \\nATOM     54  C   LYS A   7      -0.349  10.113   3.116  1.00 92.19           C  \\nATOM     55  CB  LYS A   7       1.188   9.300   1.315  1.00 91.34           C  \\nATOM     56  O   LYS A   7      -0.291  11.231   3.631  1.00 90.08           O  \\nATOM     57  CG  LYS A   7       1.450  10.670   0.707  1.00 83.48           C  \\nATOM     58  CD  LYS A   7       1.786  10.569  -0.776  1.00 78.24           C  \\nATOM     59  CE  LYS A   7       2.017  11.942  -1.392  1.00 74.52           C  \\nATOM     60  NZ  LYS A   7       2.387  11.846  -2.836  1.00 67.81           N  \\nATOM     61  N   LYS A   8      -1.548   9.629   2.889  1.00 89.59           N  \\nATOM     62  CA  LYS A   8      -2.804  10.359   3.038  1.00 88.28           C  \\nATOM     63  C   LYS A   8      -3.622   9.815   4.205  1.00 86.79           C  \\nATOM     64  CB  LYS A   8      -3.622  10.289   1.747  1.00 84.90           C  \\nATOM     65  O   LYS A   8      -4.647  10.391   4.574  1.00 83.20           O  \\nATOM     66  CG  LYS A   8      -3.053  11.121   0.607  1.00 75.06           C  \\nATOM     67  CD  LYS A   8      -4.032  11.220  -0.555  1.00 71.84           C  \\nATOM     68  CE  LYS A   8      -3.510  12.140  -1.650  1.00 65.02           C  \\nATOM     69  NZ  LYS A   8      -4.486  12.277  -2.772  1.00 59.33           N  \\nATOM     70  N   GLY A   9      -3.098   8.783   4.783  1.00 89.03           N  \\nATOM     71  CA  GLY A   9      -3.855   8.185   5.871  1.00 90.06           C  \\nATOM     72  C   GLY A   9      -5.116   7.482   5.405  1.00 86.45           C  \\nATOM     73  O   GLY A   9      -6.059   7.309   6.180  1.00 80.62           O  \\nATOM     74  N   GLU A  10      -5.183   7.049   4.141  1.00 93.36           N  \\nATOM     75  CA  GLU A  10      -6.365   6.401   3.581  1.00 93.22           C  \\nATOM     76  C   GLU A  10      -6.293   4.886   3.745  1.00 92.93           C  \\nATOM     77  CB  GLU A  10      -6.527   6.763   2.103  1.00 89.96           C  \\nATOM     78  O   GLU A  10      -5.228   4.288   3.577  1.00 91.53           O  \\nATOM     79  CG  GLU A  10      -6.805   8.239   1.858  1.00 75.06           C  \\nATOM     80  CD  GLU A  10      -7.052   8.569   0.395  1.00 72.15           C  \\nATOM     81  OE1 GLU A  10      -6.768   7.715  -0.476  1.00 69.12           O  \\nATOM     82  OE2 GLU A  10      -7.534   9.690   0.116  1.00 66.84           O  \\nATOM     83  N   ILE A  11      -7.474   4.318   4.108  1.00 94.14           N  \\nATOM     84  CA  ILE A  11      -7.596   2.869   4.227  1.00 94.55           C  \\nATOM     85  C   ILE A  11      -8.186   2.294   2.941  1.00 94.13           C  \\nATOM     86  CB  ILE A  11      -8.466   2.473   5.440  1.00 92.91           C  \\nATOM     87  O   ILE A  11      -9.228   2.754   2.469  1.00 93.11           O  \\nATOM     88  CG1 ILE A  11      -7.864   3.032   6.735  1.00 78.19           C  \\nATOM     89  CG2 ILE A  11      -8.620   0.951   5.520  1.00 77.84           C  \\nATOM     90  CD1 ILE A  11      -8.784   2.922   7.943  1.00 73.86           C  \\nATOM     91  N   LYS A  12      -7.548   1.210   2.482  1.00 94.52           N  \\nATOM     92  CA  LYS A  12      -8.046   0.587   1.259  1.00 94.67           C  \\nATOM     93  C   LYS A  12      -7.769  -0.914   1.255  1.00 94.68           C  \\nATOM     94  CB  LYS A  12      -7.415   1.240   0.029  1.00 92.98           C  \\nATOM     95  O   LYS A  12      -6.842  -1.381   1.919  1.00 94.11           O  \\nATOM     96  CG  LYS A  12      -7.680   2.734  -0.084  1.00 79.69           C  \\nATOM     97  CD  LYS A  12      -7.272   3.273  -1.449  1.00 76.19           C  \\nATOM     98  CE  LYS A  12      -7.593   4.755  -1.585  1.00 71.33           C  \\nATOM     99  NZ  LYS A  12      -7.230   5.279  -2.935  1.00 61.53           N  \\nATOM    100  N   ASN A  13      -8.592  -1.621   0.627  1.00 95.04           N  \\nATOM    101  CA  ASN A  13      -8.325  -3.019   0.310  1.00 95.29           C  \\nATOM    102  C   ASN A  13      -7.495  -3.157  -0.963  1.00 95.21           C  \\nATOM    103  CB  ASN A  13      -9.635  -3.799   0.177  1.00 94.19           C  \\nATOM    104  O   ASN A  13      -7.791  -2.517  -1.974  1.00 94.61           O  \\nATOM    105  CG  ASN A  13     -10.408  -3.869   1.479  1.00 88.42           C  \\nATOM    106  ND2 ASN A  13     -11.716  -3.648   1.405  1.00 85.06           N  \\nATOM    107  OD1 ASN A  13      -9.835  -4.117   2.543  1.00 82.96           O  \\nATOM    108  N   VAL A  14      -6.519  -3.906  -0.866  1.00 94.83           N  \\nATOM    109  CA  VAL A  14      -5.639  -4.122  -2.011  1.00 94.96           C  \\nATOM    110  C   VAL A  14      -5.317  -5.609  -2.142  1.00 94.97           C  \\nATOM    111  CB  VAL A  14      -4.335  -3.303  -1.886  1.00 94.40           C  \\nATOM    112  O   VAL A  14      -5.607  -6.396  -1.237  1.00 94.51           O  \\nATOM    113  CG1 VAL A  14      -4.639  -1.807  -1.844  1.00 92.07           C  \\nATOM    114  CG2 VAL A  14      -3.555  -3.729  -0.644  1.00 91.72           C  \\nATOM    115  N   ALA A  15      -4.807  -5.984  -3.266  1.00 95.07           N  \\nATOM    116  CA  ALA A  15      -4.357  -7.359  -3.469  1.00 95.25           C  \\nATOM    117  C   ALA A  15      -3.209  -7.705  -2.525  1.00 95.11           C  \\nATOM    118  CB  ALA A  15      -3.931  -7.571  -4.920  1.00 94.13           C  \\nATOM    119  O   ALA A  15      -2.340  -6.870  -2.261  1.00 94.45           O  \\nATOM    120  N   ASP A  16      -3.143  -8.994  -1.982  1.00 93.71           N  \\nATOM    121  CA  ASP A  16      -2.131  -9.455  -1.037  1.00 94.38           C  \\nATOM    122  C   ASP A  16      -0.724  -9.237  -1.590  1.00 93.62           C  \\nATOM    123  CB  ASP A  16      -2.341 -10.934  -0.704  1.00 91.59           C  \\nATOM    124  O   ASP A  16       0.164  -8.769  -0.873  1.00 92.13           O  \\nATOM    125  CG  ASP A  16      -3.557 -11.177   0.173  1.00 77.66           C  \\nATOM    126  OD1 ASP A  16      -4.268 -10.206   0.512  1.00 70.32           O  \\nATOM    127  OD2 ASP A  16      -3.804 -12.348   0.531  1.00 81.09           O  \\nATOM    128  N   GLY A  17      -0.522  -9.632  -2.901  1.00 93.59           N  \\nATOM    129  CA  GLY A  17       0.794  -9.478  -3.502  1.00 93.79           C  \\nATOM    130  C   GLY A  17       1.245  -8.032  -3.587  1.00 93.43           C  \\nATOM    131  O   GLY A  17       2.428  -7.734  -3.413  1.00 92.06           O  \\nATOM    132  N   TYR A  18       0.293  -7.173  -3.926  1.00 94.45           N  \\nATOM    133  CA  TYR A  18       0.585  -5.746  -4.004  1.00 94.85           C  \\nATOM    134  C   TYR A  18       0.945  -5.187  -2.632  1.00 94.25           C  \\nATOM    135  CB  TYR A  18      -0.612  -4.983  -4.580  1.00 93.82           C  \\nATOM    136  O   TYR A  18       1.924  -4.450  -2.493  1.00 93.28           O  \\nATOM    137  CG  TYR A  18      -0.383  -3.496  -4.701  1.00 89.63           C  \\nATOM    138  CD1 TYR A  18       0.679  -2.993  -5.450  1.00 84.12           C  \\nATOM    139  CD2 TYR A  18      -1.226  -2.591  -4.065  1.00 84.96           C  \\nATOM    140  CE1 TYR A  18       0.895  -1.624  -5.562  1.00 87.45           C  \\nATOM    141  CE2 TYR A  18      -1.019  -1.219  -4.171  1.00 87.89           C  \\nATOM    142  OH  TYR A  18       0.251   0.611  -5.028  1.00 72.19           O  \\nATOM    143  CZ  TYR A  18       0.042  -0.746  -4.920  1.00 87.01           C  \\nATOM    144  N   ALA A  19       0.340  -5.497  -1.670  1.00 94.99           N  \\nATOM    145  CA  ALA A  19       0.635  -5.070  -0.305  1.00 95.31           C  \\nATOM    146  C   ALA A  19       2.002  -5.576   0.145  1.00 95.03           C  \\nATOM    147  CB  ALA A  19      -0.451  -5.559   0.651  1.00 94.57           C  \\nATOM    148  O   ALA A  19       2.845  -4.793   0.590  1.00 94.36           O  \\nATOM    149  N   ASN A  20       2.263  -6.919  -0.012  1.00 95.00           N  \\nATOM    150  CA  ASN A  20       3.420  -7.570   0.593  1.00 94.99           C  \\nATOM    151  C   ASN A  20       4.692  -7.312  -0.211  1.00 94.82           C  \\nATOM    152  CB  ASN A  20       3.177  -9.074   0.734  1.00 94.06           C  \\nATOM    153  O   ASN A  20       5.782  -7.222   0.357  1.00 94.01           O  \\nATOM    154  CG  ASN A  20       2.135  -9.401   1.786  1.00 89.82           C  \\nATOM    155  ND2 ASN A  20       1.319 -10.413   1.517  1.00 88.19           N  \\nATOM    156  OD1 ASN A  20       2.064  -8.749   2.831  1.00 86.47           O  \\nATOM    157  N   ASN A  21       4.569  -7.240  -1.539  1.00 94.62           N  \\nATOM    158  CA  ASN A  21       5.761  -7.163  -2.377  1.00 94.54           C  \\nATOM    159  C   ASN A  21       6.103  -5.720  -2.736  1.00 94.35           C  \\nATOM    160  CB  ASN A  21       5.579  -7.997  -3.647  1.00 93.38           C  \\nATOM    161  O   ASN A  21       7.183  -5.446  -3.262  1.00 93.10           O  \\nATOM    162  CG  ASN A  21       5.445  -9.479  -3.359  1.00 88.98           C  \\nATOM    163  ND2 ASN A  21       4.502 -10.131  -4.029  1.00 84.70           N  \\nATOM    164  OD1 ASN A  21       6.183 -10.034  -2.541  1.00 83.89           O  \\nATOM    165  N   PHE A  22       5.284  -4.782  -2.438  1.00 95.28           N  \\nATOM    166  CA  PHE A  22       5.556  -3.400  -2.816  1.00 95.20           C  \\nATOM    167  C   PHE A  22       5.266  -2.455  -1.656  1.00 95.22           C  \\nATOM    168  CB  PHE A  22       4.722  -3.002  -4.038  1.00 94.42           C  \\nATOM    169  O   PHE A  22       6.180  -1.830  -1.115  1.00 94.28           O  \\nATOM    170  CG  PHE A  22       5.048  -1.635  -4.575  1.00 91.44           C  \\nATOM    171  CD1 PHE A  22       6.175  -1.432  -5.362  1.00 86.97           C  \\nATOM    172  CD2 PHE A  22       4.225  -0.551  -4.294  1.00 87.04           C  \\nATOM    173  CE1 PHE A  22       6.479  -0.167  -5.860  1.00 84.67           C  \\nATOM    174  CE2 PHE A  22       4.523   0.715  -4.789  1.00 84.37           C  \\nATOM    175  CZ  PHE A  22       5.650   0.905  -5.571  1.00 83.94           C  \\nATOM    176  N   LEU A  23       4.075  -2.403  -1.270  1.00 94.36           N  \\nATOM    177  CA  LEU A  23       3.685  -1.347  -0.342  1.00 94.82           C  \\nATOM    178  C   LEU A  23       4.481  -1.441   0.955  1.00 94.51           C  \\nATOM    179  CB  LEU A  23       2.186  -1.423  -0.042  1.00 94.06           C  \\nATOM    180  O   LEU A  23       5.087  -0.459   1.389  1.00 93.71           O  \\nATOM    181  CG  LEU A  23       1.244  -1.080  -1.197  1.00 91.49           C  \\nATOM    182  CD1 LEU A  23      -0.202  -1.363  -0.804  1.00 87.26           C  \\nATOM    183  CD2 LEU A  23       1.416   0.377  -1.613  1.00 87.61           C  \\nATOM    184  N   PHE A  24       4.495  -2.571   1.600  1.00 94.70           N  \\nATOM    185  CA  PHE A  24       5.171  -2.717   2.883  1.00 95.21           C  \\nATOM    186  C   PHE A  24       6.683  -2.621   2.713  1.00 94.88           C  \\nATOM    187  CB  PHE A  24       4.800  -4.050   3.540  1.00 94.39           C  \\nATOM    188  O   PHE A  24       7.368  -2.012   3.537  1.00 94.23           O  \\nATOM    189  CG  PHE A  24       3.350  -4.151   3.932  1.00 90.66           C  \\nATOM    190  CD1 PHE A  24       2.599  -3.006   4.172  1.00 85.07           C  \\nATOM    191  CD2 PHE A  24       2.738  -5.391   4.060  1.00 85.54           C  \\nATOM    192  CE1 PHE A  24       1.258  -3.097   4.535  1.00 88.82           C  \\nATOM    193  CE2 PHE A  24       1.398  -5.489   4.423  1.00 88.82           C  \\nATOM    194  CZ  PHE A  24       0.660  -4.341   4.660  1.00 88.24           C  \\nATOM    195  N   LYS A  25       7.235  -3.188   1.605  1.00 95.64           N  \\nATOM    196  CA  LYS A  25       8.674  -3.163   1.362  1.00 95.45           C  \\nATOM    197  C   LYS A  25       9.172  -1.736   1.152  1.00 95.33           C  \\nATOM    198  CB  LYS A  25       9.028  -4.025   0.149  1.00 94.37           C  \\nATOM    199  O   LYS A  25      10.273  -1.388   1.583  1.00 94.52           O  \\nATOM    200  CG  LYS A  25       8.931  -5.522   0.403  1.00 85.04           C  \\nATOM    201  CD  LYS A  25       9.508  -6.325  -0.756  1.00 79.39           C  \\nATOM    202  CE  LYS A  25       9.465  -7.821  -0.477  1.00 74.28           C  \\nATOM    203  NZ  LYS A  25      10.062  -8.610  -1.596  1.00 67.90           N  \\nATOM    204  N   GLN A  26       8.329  -0.912   0.584  1.00 94.91           N  \\nATOM    205  CA  GLN A  26       8.710   0.465   0.288  1.00 94.51           C  \\nATOM    206  C   GLN A  26       8.321   1.401   1.428  1.00 93.93           C  \\nATOM    207  CB  GLN A  26       8.064   0.932  -1.018  1.00 92.90           C  \\nATOM    208  O   GLN A  26       8.567   2.607   1.360  1.00 92.08           O  \\nATOM    209  CG  GLN A  26       8.520   0.152  -2.243  1.00 82.40           C  \\nATOM    210  CD  GLN A  26       9.949   0.470  -2.643  1.00 75.27           C  \\nATOM    211  NE2 GLN A  26      10.596  -0.468  -3.327  1.00 67.52           N  \\nATOM    212  OE1 GLN A  26      10.467   1.550  -2.340  1.00 75.38           O  \\nATOM    213  N   GLY A  27       7.721   0.768   2.494  1.00 94.56           N  \\nATOM    214  CA  GLY A  27       7.306   1.596   3.615  1.00 94.85           C  \\nATOM    215  C   GLY A  27       6.184   2.555   3.266  1.00 93.98           C  \\nATOM    216  O   GLY A  27       6.104   3.654   3.818  1.00 91.78           O  \\nATOM    217  N   LEU A  28       5.281   2.199   2.389  1.00 94.57           N  \\nATOM    218  CA  LEU A  28       4.263   3.104   1.865  1.00 94.67           C  \\nATOM    219  C   LEU A  28       2.913   2.843   2.524  1.00 94.11           C  \\nATOM    220  CB  LEU A  28       4.141   2.952   0.347  1.00 93.74           C  \\nATOM    221  O   LEU A  28       2.001   3.668   2.429  1.00 92.77           O  \\nATOM    222  CG  LEU A  28       5.340   3.418  -0.480  1.00 89.38           C  \\nATOM    223  CD1 LEU A  28       5.146   3.053  -1.948  1.00 81.99           C  \\nATOM    224  CD2 LEU A  28       5.550   4.919  -0.320  1.00 82.69           C  \\nATOM    225  N   ALA A  29       2.838   1.830   3.201  1.00 93.57           N  \\nATOM    226  CA  ALA A  29       1.608   1.418   3.873  1.00 94.46           C  \\nATOM    227  C   ALA A  29       1.902   0.426   4.995  1.00 93.52           C  \\nATOM    228  CB  ALA A  29       0.632   0.809   2.870  1.00 92.74           C  \\nATOM    229  O   ALA A  29       2.991  -0.150   5.053  1.00 92.09           O  \\nATOM    230  N   ILE A  30       0.899   0.249   6.011  1.00 95.49           N  \\nATOM    231  CA  ILE A  30       0.916  -0.803   7.022  1.00 95.59           C  \\nATOM    232  C   ILE A  30      -0.419  -1.545   7.018  1.00 95.47           C  \\nATOM    233  CB  ILE A  30       1.206  -0.230   8.428  1.00 94.89           C  \\nATOM    234  O   ILE A  30      -1.404  -1.059   6.457  1.00 94.99           O  \\nATOM    235  CG1 ILE A  30       0.095   0.738   8.850  1.00 91.87           C  \\nATOM    236  CG2 ILE A  30       2.573   0.458   8.459  1.00 91.59           C  \\nATOM    237  CD1 ILE A  30       0.166   1.164  10.310  1.00 88.23           C  \\nATOM    238  N   GLU A  31      -0.393  -2.723   7.484  1.00 94.94           N  \\nATOM    239  CA  GLU A  31      -1.638  -3.477   7.599  1.00 95.11           C  \\nATOM    240  C   GLU A  31      -2.583  -2.832   8.608  1.00 94.81           C  \\nATOM    241  CB  GLU A  31      -1.353  -4.928   7.996  1.00 93.80           C  \\nATOM    242  O   GLU A  31      -2.158  -2.424   9.691  1.00 93.93           O  \\nATOM    243  CG  GLU A  31      -2.584  -5.823   7.977  1.00 86.31           C  \\nATOM    244  CD  GLU A  31      -2.276  -7.272   8.317  1.00 83.42           C  \\nATOM    245  OE1 GLU A  31      -1.104  -7.589   8.621  1.00 84.59           O  \\nATOM    246  OE2 GLU A  31      -3.215  -8.098   8.279  1.00 84.56           O  \\nATOM    247  N   ALA A  32      -3.860  -2.647   8.224  1.00 94.97           N  \\nATOM    248  CA  ALA A  32      -4.867  -2.206   9.185  1.00 94.60           C  \\nATOM    249  C   ALA A  32      -5.276  -3.347  10.113  1.00 93.71           C  \\nATOM    250  CB  ALA A  32      -6.090  -1.651   8.459  1.00 93.35           C  \\nATOM    251  O   ALA A  32      -5.239  -4.516   9.723  1.00 91.28           O  \\nTER     252      ALA A  32\\nEND\\n\",\"pdb\");\n",
              "\tviewer_17580091717215044.setStyle({\"model\": 0},{\"cartoon\": {\"color\": \"green\", \"opacity\": 0.7}});\n",
              "\tviewer_17580091717215044.addModel(\"ATOM      1  N   MET A   1      -6.473  -7.251   1.654  1.00 91.96           N  \\nATOM      2  CA  MET A   1      -5.833  -6.977   2.937  1.00 91.96           C  \\nATOM      3  C   MET A   1      -5.912  -5.493   3.280  1.00 91.96           C  \\nATOM      4  O   MET A   1      -5.650  -4.641   2.429  1.00 91.96           O  \\nATOM      5  CB  MET A   1      -4.372  -7.432   2.919  1.00 91.96           C  \\nATOM      6  CG  MET A   1      -3.612  -7.107   4.193  1.00 91.96           C  \\nATOM      7  SD  MET A   1      -1.843  -7.584   4.099  1.00 91.96           S  \\nATOM      8  CE  MET A   1      -1.960  -9.337   4.552  1.00 91.96           C  \\nATOM      9  N   LYS A   2      -6.632  -5.346   4.185  1.00 93.29           N  \\nATOM     10  CA  LYS A   2      -6.861  -3.982   4.650  1.00 93.29           C  \\nATOM     11  C   LYS A   2      -5.555  -3.325   5.088  1.00 93.29           C  \\nATOM     12  O   LYS A   2      -4.845  -3.853   5.946  1.00 93.29           O  \\nATOM     13  CB  LYS A   2      -7.866  -3.967   5.802  1.00 93.29           C  \\nATOM     14  CG  LYS A   2      -8.503  -2.608   6.052  1.00 93.29           C  \\nATOM     15  CD  LYS A   2      -9.632  -2.698   7.069  1.00 93.29           C  \\nATOM     16  CE  LYS A   2     -10.255  -1.334   7.337  1.00 93.29           C  \\nATOM     17  NZ  LYS A   2     -11.368  -1.419   8.330  1.00 93.29           N  \\nATOM     18  N   VAL A   3      -5.215  -2.256   4.311  1.00 94.76           N  \\nATOM     19  CA  VAL A   3      -3.983  -1.559   4.666  1.00 94.76           C  \\nATOM     20  C   VAL A   3      -4.261  -0.067   4.828  1.00 94.76           C  \\nATOM     21  O   VAL A   3      -5.256   0.445   4.310  1.00 94.76           O  \\nATOM     22  CB  VAL A   3      -2.880  -1.784   3.608  1.00 94.76           C  \\nATOM     23  CG1 VAL A   3      -2.540  -3.269   3.488  1.00 94.76           C  \\nATOM     24  CG2 VAL A   3      -3.314  -1.221   2.256  1.00 94.76           C  \\nATOM     25  N   ILE A   4      -3.392   0.660   5.555  1.00 95.11           N  \\nATOM     26  CA  ILE A   4      -3.429   2.108   5.722  1.00 95.11           C  \\nATOM     27  C   ILE A   4      -2.257   2.743   4.976  1.00 95.11           C  \\nATOM     28  O   ILE A   4      -1.096   2.430   5.250  1.00 95.11           O  \\nATOM     29  CB  ILE A   4      -3.394   2.506   7.214  1.00 95.11           C  \\nATOM     30  CG1 ILE A   4      -4.562   1.860   7.968  1.00 95.11           C  \\nATOM     31  CG2 ILE A   4      -3.419   4.030   7.367  1.00 95.11           C  \\nATOM     32  CD1 ILE A   4      -4.490   2.024   9.479  1.00 95.11           C  \\nATOM     33  N   PHE A   5      -2.509   3.598   4.050  1.00 93.85           N  \\nATOM     34  CA  PHE A   5      -1.483   4.360   3.349  1.00 93.85           C  \\nATOM     35  C   PHE A   5      -0.955   5.490   4.224  1.00 93.85           C  \\nATOM     36  O   PHE A   5      -1.714   6.370   4.635  1.00 93.85           O  \\nATOM     37  CB  PHE A   5      -2.036   4.927   2.037  1.00 93.85           C  \\nATOM     38  CG  PHE A   5      -2.328   3.879   0.997  1.00 93.85           C  \\nATOM     39  CD1 PHE A   5      -1.395   3.576   0.014  1.00 93.85           C  \\nATOM     40  CD2 PHE A   5      -3.537   3.195   1.004  1.00 93.85           C  \\nATOM     41  CE1 PHE A   5      -1.663   2.606  -0.949  1.00 93.85           C  \\nATOM     42  CE2 PHE A   5      -3.812   2.225   0.046  1.00 93.85           C  \\nATOM     43  CZ  PHE A   5      -2.874   1.933  -0.931  1.00 93.85           C  \\nATOM     44  N   LEU A   6       0.309   5.442   4.468  1.00 92.63           N  \\nATOM     45  CA  LEU A   6       0.906   6.315   5.472  1.00 92.63           C  \\nATOM     46  C   LEU A   6       0.898   7.767   5.006  1.00 92.63           C  \\nATOM     47  O   LEU A   6       0.618   8.675   5.791  1.00 92.63           O  \\nATOM     48  CB  LEU A   6       2.340   5.875   5.782  1.00 92.63           C  \\nATOM     49  CG  LEU A   6       2.499   4.515   6.463  1.00 92.63           C  \\nATOM     50  CD1 LEU A   6       3.975   4.152   6.583  1.00 92.63           C  \\nATOM     51  CD2 LEU A   6       1.831   4.521   7.834  1.00 92.63           C  \\nATOM     52  N   LYS A   7       1.006   7.997   3.747  1.00 90.46           N  \\nATOM     53  CA  LYS A   7       1.064   9.366   3.243  1.00 90.46           C  \\nATOM     54  C   LYS A   7      -0.336   9.950   3.075  1.00 90.46           C  \\nATOM     55  O   LYS A   7      -0.558  11.132   3.345  1.00 90.46           O  \\nATOM     56  CB  LYS A   7       1.815   9.416   1.911  1.00 90.46           C  \\nATOM     57  CG  LYS A   7       3.318   9.217   2.040  1.00 90.46           C  \\nATOM     58  CD  LYS A   7       4.024   9.400   0.703  1.00 90.46           C  \\nATOM     59  CE  LYS A   7       5.525   9.179   0.827  1.00 90.46           C  \\nATOM     60  NZ  LYS A   7       6.210   9.272  -0.498  1.00 90.46           N  \\nATOM     61  N   LYS A   8      -1.239   9.124   2.770  1.00 89.16           N  \\nATOM     62  CA  LYS A   8      -2.593   9.599   2.506  1.00 89.16           C  \\nATOM     63  C   LYS A   8      -3.502   9.370   3.710  1.00 89.16           C  \\nATOM     64  O   LYS A   8      -4.460  10.116   3.922  1.00 89.16           O  \\nATOM     65  CB  LYS A   8      -3.174   8.906   1.272  1.00 89.16           C  \\nATOM     66  CG  LYS A   8      -2.599   9.405  -0.046  1.00 89.16           C  \\nATOM     67  CD  LYS A   8      -3.314   8.782  -1.238  1.00 89.16           C  \\nATOM     68  CE  LYS A   8      -2.698   9.231  -2.557  1.00 89.16           C  \\nATOM     69  NZ  LYS A   8      -3.372   8.593  -3.728  1.00 89.16           N  \\nATOM     70  N   GLY A   9      -3.090   8.432   4.583  1.00 92.79           N  \\nATOM     71  CA  GLY A   9      -3.923   8.147   5.740  1.00 92.79           C  \\nATOM     72  C   GLY A   9      -5.217   7.439   5.385  1.00 92.79           C  \\nATOM     73  O   GLY A   9      -6.249   7.665   6.019  1.00 92.79           O  \\nATOM     74  N   GLU A  10      -5.272   6.818   4.294  1.00 92.57           N  \\nATOM     75  CA  GLU A  10      -6.472   6.137   3.817  1.00 92.57           C  \\nATOM     76  C   GLU A  10      -6.337   4.623   3.952  1.00 92.57           C  \\nATOM     77  O   GLU A  10      -5.232   4.083   3.876  1.00 92.57           O  \\nATOM     78  CB  GLU A  10      -6.759   6.510   2.361  1.00 92.57           C  \\nATOM     79  CG  GLU A  10      -7.166   7.965   2.167  1.00 92.57           C  \\nATOM     80  CD  GLU A  10      -7.526   8.301   0.729  1.00 92.57           C  \\nATOM     81  OE1 GLU A  10      -8.011   9.426   0.472  1.00 92.57           O  \\nATOM     82  OE2 GLU A  10      -7.324   7.431  -0.148  1.00 92.57           O  \\nATOM     83  N   ILE A  11      -7.531   4.069   4.140  1.00 95.62           N  \\nATOM     84  CA  ILE A  11      -7.592   2.613   4.208  1.00 95.62           C  \\nATOM     85  C   ILE A  11      -8.180   2.061   2.911  1.00 95.62           C  \\nATOM     86  O   ILE A  11      -9.250   2.492   2.474  1.00 95.62           O  \\nATOM     87  CB  ILE A  11      -8.426   2.138   5.419  1.00 95.62           C  \\nATOM     88  CG1 ILE A  11      -7.850   2.710   6.720  1.00 95.62           C  \\nATOM     89  CG2 ILE A  11      -8.479   0.608   5.470  1.00 95.62           C  \\nATOM     90  CD1 ILE A  11      -8.704   2.430   7.949  1.00 95.62           C  \\nATOM     91  N   LYS A  12      -7.476   1.182   2.319  1.00 94.49           N  \\nATOM     92  CA  LYS A  12      -7.927   0.569   1.072  1.00 94.49           C  \\nATOM     93  C   LYS A  12      -7.683  -0.937   1.079  1.00 94.49           C  \\nATOM     94  O   LYS A  12      -6.798  -1.424   1.785  1.00 94.49           O  \\nATOM     95  CB  LYS A  12      -7.224   1.209  -0.126  1.00 94.49           C  \\nATOM     96  CG  LYS A  12      -7.676   2.632  -0.422  1.00 94.49           C  \\nATOM     97  CD  LYS A  12      -7.128   3.126  -1.754  1.00 94.49           C  \\nATOM     98  CE  LYS A  12      -7.656   4.512  -2.097  1.00 94.49           C  \\nATOM     99  NZ  LYS A  12      -7.107   5.009  -3.395  1.00 94.49           N  \\nATOM    100  N   ASN A  13      -8.560  -1.585   0.329  1.00 96.14           N  \\nATOM    101  CA  ASN A  13      -8.325  -3.004   0.081  1.00 96.14           C  \\nATOM    102  C   ASN A  13      -7.477  -3.222  -1.168  1.00 96.14           C  \\nATOM    103  O   ASN A  13      -7.810  -2.722  -2.244  1.00 96.14           O  \\nATOM    104  CB  ASN A  13      -9.652  -3.755  -0.041  1.00 96.14           C  \\nATOM    105  CG  ASN A  13     -10.336  -3.953   1.298  1.00 96.14           C  \\nATOM    106  OD1 ASN A  13      -9.744  -4.492   2.238  1.00 96.14           O  \\nATOM    107  ND2 ASN A  13     -11.587  -3.519   1.396  1.00 96.14           N  \\nATOM    108  N   VAL A  14      -6.463  -3.833  -0.970  1.00 95.82           N  \\nATOM    109  CA  VAL A  14      -5.603  -4.119  -2.114  1.00 95.82           C  \\nATOM    110  C   VAL A  14      -5.335  -5.620  -2.198  1.00 95.82           C  \\nATOM    111  O   VAL A  14      -5.637  -6.363  -1.262  1.00 95.82           O  \\nATOM    112  CB  VAL A  14      -4.269  -3.342  -2.029  1.00 95.82           C  \\nATOM    113  CG1 VAL A  14      -4.514  -1.837  -2.109  1.00 95.82           C  \\nATOM    114  CG2 VAL A  14      -3.527  -3.699  -0.743  1.00 95.82           C  \\nATOM    115  N   ALA A  15      -4.709  -6.094  -3.310  1.00 96.50           N  \\nATOM    116  CA  ALA A  15      -4.286  -7.484  -3.455  1.00 96.50           C  \\nATOM    117  C   ALA A  15      -3.159  -7.820  -2.484  1.00 96.50           C  \\nATOM    118  O   ALA A  15      -2.325  -6.967  -2.173  1.00 96.50           O  \\nATOM    119  CB  ALA A  15      -3.848  -7.759  -4.892  1.00 96.50           C  \\nATOM    120  N   ASP A  16      -3.085  -9.029  -2.019  1.00 95.07           N  \\nATOM    121  CA  ASP A  16      -2.104  -9.476  -1.033  1.00 95.07           C  \\nATOM    122  C   ASP A  16      -0.681  -9.299  -1.557  1.00 95.07           C  \\nATOM    123  O   ASP A  16       0.191  -8.798  -0.845  1.00 95.07           O  \\nATOM    124  CB  ASP A  16      -2.349 -10.938  -0.657  1.00 95.07           C  \\nATOM    125  CG  ASP A  16      -3.589 -11.129   0.199  1.00 95.07           C  \\nATOM    126  OD1 ASP A  16      -4.039 -10.160   0.847  1.00 95.07           O  \\nATOM    127  OD2 ASP A  16      -4.119 -12.261   0.228  1.00 95.07           O  \\nATOM    128  N   GLY A  17      -0.520  -9.702  -2.879  1.00 96.41           N  \\nATOM    129  CA  GLY A  17       0.803  -9.617  -3.477  1.00 96.41           C  \\nATOM    130  C   GLY A  17       1.304  -8.192  -3.618  1.00 96.41           C  \\nATOM    131  O   GLY A  17       2.471  -7.909  -3.340  1.00 96.41           O  \\nATOM    132  N   TYR A  18       0.390  -7.276  -4.001  1.00 96.10           N  \\nATOM    133  CA  TYR A  18       0.714  -5.861  -4.135  1.00 96.10           C  \\nATOM    134  C   TYR A  18       1.035  -5.246  -2.778  1.00 96.10           C  \\nATOM    135  O   TYR A  18       2.029  -4.531  -2.633  1.00 96.10           O  \\nATOM    136  CB  TYR A  18      -0.445  -5.103  -4.790  1.00 96.10           C  \\nATOM    137  CG  TYR A  18      -0.172  -3.632  -4.990  1.00 96.10           C  \\nATOM    138  CD1 TYR A  18      -0.937  -2.668  -4.337  1.00 96.10           C  \\nATOM    139  CD2 TYR A  18       0.848  -3.203  -5.832  1.00 96.10           C  \\nATOM    140  CE1 TYR A  18      -0.692  -1.311  -4.518  1.00 96.10           C  \\nATOM    141  CE2 TYR A  18       1.102  -1.849  -6.022  1.00 96.10           C  \\nATOM    142  CZ  TYR A  18       0.329  -0.912  -5.360  1.00 96.10           C  \\nATOM    143  OH  TYR A  18       0.577   0.430  -5.545  1.00 96.10           O  \\nATOM    144  N   ALA A  19       0.312  -5.600  -1.756  1.00 96.34           N  \\nATOM    145  CA  ALA A  19       0.541  -5.085  -0.408  1.00 96.34           C  \\nATOM    146  C   ALA A  19       1.898  -5.534   0.126  1.00 96.34           C  \\nATOM    147  O   ALA A  19       2.697  -4.711   0.580  1.00 96.34           O  \\nATOM    148  CB  ALA A  19      -0.573  -5.538   0.531  1.00 96.34           C  \\nATOM    149  N   ASN A  20       2.264  -6.774  -0.059  1.00 95.93           N  \\nATOM    150  CA  ASN A  20       3.451  -7.360   0.556  1.00 95.93           C  \\nATOM    151  C   ASN A  20       4.718  -6.994  -0.212  1.00 95.93           C  \\nATOM    152  O   ASN A  20       5.747  -6.686   0.392  1.00 95.93           O  \\nATOM    153  CB  ASN A  20       3.309  -8.880   0.654  1.00 95.93           C  \\nATOM    154  CG  ASN A  20       2.290  -9.306   1.692  1.00 95.93           C  \\nATOM    155  OD1 ASN A  20       2.196  -8.712   2.769  1.00 95.93           O  \\nATOM    156  ND2 ASN A  20       1.516 -10.338   1.375  1.00 95.93           N  \\nATOM    157  N   ASN A  21       4.546  -6.914  -1.585  1.00 96.23           N  \\nATOM    158  CA  ASN A  21       5.752  -6.789  -2.397  1.00 96.23           C  \\nATOM    159  C   ASN A  21       6.016  -5.337  -2.789  1.00 96.23           C  \\nATOM    160  O   ASN A  21       7.120  -4.997  -3.217  1.00 96.23           O  \\nATOM    161  CB  ASN A  21       5.653  -7.667  -3.645  1.00 96.23           C  \\nATOM    162  CG  ASN A  21       5.723  -9.148  -3.325  1.00 96.23           C  \\nATOM    163  OD1 ASN A  21       6.552  -9.582  -2.521  1.00 96.23           O  \\nATOM    164  ND2 ASN A  21       4.854  -9.932  -3.951  1.00 96.23           N  \\nATOM    165  N   PHE A  22       5.015  -4.516  -2.622  1.00 96.02           N  \\nATOM    166  CA  PHE A  22       5.190  -3.134  -3.052  1.00 96.02           C  \\nATOM    167  C   PHE A  22       4.943  -2.172  -1.894  1.00 96.02           C  \\nATOM    168  O   PHE A  22       5.832  -1.406  -1.518  1.00 96.02           O  \\nATOM    169  CB  PHE A  22       4.250  -2.807  -4.216  1.00 96.02           C  \\nATOM    170  CG  PHE A  22       4.456  -1.435  -4.797  1.00 96.02           C  \\nATOM    171  CD1 PHE A  22       3.566  -0.405  -4.515  1.00 96.02           C  \\nATOM    172  CD2 PHE A  22       5.540  -1.174  -5.624  1.00 96.02           C  \\nATOM    173  CE1 PHE A  22       3.755   0.867  -5.052  1.00 96.02           C  \\nATOM    174  CE2 PHE A  22       5.734   0.094  -6.164  1.00 96.02           C  \\nATOM    175  CZ  PHE A  22       4.841   1.113  -5.877  1.00 96.02           C  \\nATOM    176  N   LEU A  23       3.902  -2.264  -1.205  1.00 95.97           N  \\nATOM    177  CA  LEU A  23       3.506  -1.248  -0.236  1.00 95.97           C  \\nATOM    178  C   LEU A  23       4.270  -1.417   1.072  1.00 95.97           C  \\nATOM    179  O   LEU A  23       4.861  -0.459   1.579  1.00 95.97           O  \\nATOM    180  CB  LEU A  23       1.999  -1.318   0.027  1.00 95.97           C  \\nATOM    181  CG  LEU A  23       1.087  -0.968  -1.150  1.00 95.97           C  \\nATOM    182  CD1 LEU A  23      -0.374  -1.184  -0.771  1.00 95.97           C  \\nATOM    183  CD2 LEU A  23       1.321   0.471  -1.599  1.00 95.97           C  \\nATOM    184  N   PHE A  24       4.313  -2.597   1.609  1.00 95.80           N  \\nATOM    185  CA  PHE A  24       4.995  -2.797   2.883  1.00 95.80           C  \\nATOM    186  C   PHE A  24       6.507  -2.733   2.703  1.00 95.80           C  \\nATOM    187  O   PHE A  24       7.213  -2.166   3.539  1.00 95.80           O  \\nATOM    188  CB  PHE A  24       4.597  -4.141   3.502  1.00 95.80           C  \\nATOM    189  CG  PHE A  24       3.167  -4.194   3.967  1.00 95.80           C  \\nATOM    190  CD1 PHE A  24       2.540  -3.059   4.467  1.00 95.80           C  \\nATOM    191  CD2 PHE A  24       2.448  -5.381   3.905  1.00 95.80           C  \\nATOM    192  CE1 PHE A  24       1.216  -3.106   4.898  1.00 95.80           C  \\nATOM    193  CE2 PHE A  24       1.125  -5.435   4.333  1.00 95.80           C  \\nATOM    194  CZ  PHE A  24       0.511  -4.298   4.831  1.00 95.80           C  \\nATOM    195  N   LYS A  25       6.957  -3.265   1.625  1.00 95.47           N  \\nATOM    196  CA  LYS A  25       8.395  -3.295   1.377  1.00 95.47           C  \\nATOM    197  C   LYS A  25       8.957  -1.885   1.225  1.00 95.47           C  \\nATOM    198  O   LYS A  25      10.075  -1.605   1.663  1.00 95.47           O  \\nATOM    199  CB  LYS A  25       8.708  -4.120   0.128  1.00 95.47           C  \\nATOM    200  CG  LYS A  25      10.184  -4.443  -0.048  1.00 95.47           C  \\nATOM    201  CD  LYS A  25      10.402  -5.499  -1.123  1.00 95.47           C  \\nATOM    202  CE  LYS A  25      11.878  -5.841  -1.283  1.00 95.47           C  \\nATOM    203  NZ  LYS A  25      12.084  -6.978  -2.230  1.00 95.47           N  \\nATOM    204  N   GLN A  26       8.123  -1.009   0.673  1.00 94.66           N  \\nATOM    205  CA  GLN A  26       8.573   0.362   0.455  1.00 94.66           C  \\nATOM    206  C   GLN A  26       8.179   1.262   1.623  1.00 94.66           C  \\nATOM    207  O   GLN A  26       8.524   2.445   1.645  1.00 94.66           O  \\nATOM    208  CB  GLN A  26       7.998   0.915  -0.849  1.00 94.66           C  \\nATOM    209  CG  GLN A  26       8.512   0.208  -2.097  1.00 94.66           C  \\nATOM    210  CD  GLN A  26      10.015   0.328  -2.262  1.00 94.66           C  \\nATOM    211  OE1 GLN A  26      10.605   1.369  -1.958  1.00 94.66           O  \\nATOM    212  NE2 GLN A  26      10.646  -0.738  -2.744  1.00 94.66           N  \\nATOM    213  N   GLY A  27       7.449   0.698   2.507  1.00 94.71           N  \\nATOM    214  CA  GLY A  27       7.056   1.497   3.658  1.00 94.71           C  \\nATOM    215  C   GLY A  27       5.935   2.473   3.352  1.00 94.71           C  \\nATOM    216  O   GLY A  27       5.877   3.559   3.930  1.00 94.71           O  \\nATOM    217  N   LEU A  28       5.171   2.191   2.316  1.00 94.13           N  \\nATOM    218  CA  LEU A  28       4.115   3.091   1.864  1.00 94.13           C  \\nATOM    219  C   LEU A  28       2.813   2.820   2.610  1.00 94.13           C  \\nATOM    220  O   LEU A  28       1.934   3.682   2.669  1.00 94.13           O  \\nATOM    221  CB  LEU A  28       3.893   2.944   0.356  1.00 94.13           C  \\nATOM    222  CG  LEU A  28       5.053   3.371  -0.545  1.00 94.13           C  \\nATOM    223  CD1 LEU A  28       4.770   2.984  -1.992  1.00 94.13           C  \\nATOM    224  CD2 LEU A  28       5.300   4.871  -0.424  1.00 94.13           C  \\nATOM    225  N   ALA A  29       2.770   1.593   3.176  1.00 95.99           N  \\nATOM    226  CA  ALA A  29       1.528   1.238   3.857  1.00 95.99           C  \\nATOM    227  C   ALA A  29       1.793   0.293   5.024  1.00 95.99           C  \\nATOM    228  O   ALA A  29       2.859  -0.320   5.106  1.00 95.99           O  \\nATOM    229  CB  ALA A  29       0.546   0.604   2.875  1.00 95.99           C  \\nATOM    230  N   ILE A  30       0.948   0.243   6.007  1.00 95.27           N  \\nATOM    231  CA  ILE A  30       0.979  -0.690   7.127  1.00 95.27           C  \\nATOM    232  C   ILE A  30      -0.315  -1.501   7.159  1.00 95.27           C  \\nATOM    233  O   ILE A  30      -1.296  -1.142   6.503  1.00 95.27           O  \\nATOM    234  CB  ILE A  30       1.182   0.046   8.471  1.00 95.27           C  \\nATOM    235  CG1 ILE A  30       0.075   1.084   8.684  1.00 95.27           C  \\nATOM    236  CG2 ILE A  30       2.565   0.701   8.523  1.00 95.27           C  \\nATOM    237  CD1 ILE A  30       0.082   1.721  10.067  1.00 95.27           C  \\nATOM    238  N   GLU A  31      -0.191  -2.534   7.711  1.00 92.92           N  \\nATOM    239  CA  GLU A  31      -1.386  -3.361   7.844  1.00 92.92           C  \\nATOM    240  C   GLU A  31      -2.475  -2.637   8.631  1.00 92.92           C  \\nATOM    241  O   GLU A  31      -2.210  -2.078   9.696  1.00 92.92           O  \\nATOM    242  CB  GLU A  31      -1.046  -4.692   8.520  1.00 92.92           C  \\nATOM    243  CG  GLU A  31      -0.223  -5.632   7.650  1.00 92.92           C  \\nATOM    244  CD  GLU A  31       0.093  -6.956   8.328  1.00 92.92           C  \\nATOM    245  OE1 GLU A  31       0.650  -7.859   7.665  1.00 92.92           O  \\nATOM    246  OE2 GLU A  31      -0.223  -7.090   9.532  1.00 92.92           O  \\nATOM    247  N   ALA A  32      -3.479  -2.209   7.753  1.00 89.09           N  \\nATOM    248  CA  ALA A  32      -4.626  -1.671   8.477  1.00 89.09           C  \\nATOM    249  C   ALA A  32      -5.173  -2.689   9.474  1.00 89.09           C  \\nATOM    250  O   ALA A  32      -4.987  -3.896   9.303  1.00 89.09           O  \\nATOM    251  CB  ALA A  32      -5.721  -1.247   7.501  1.00 89.09           C  \\nTER     252      ALA A  32                                                       \\nEND   \\n\",\"pdb\");\n",
              "\tviewer_17580091717215044.setStyle({\"model\": 1, \"cartoon\": {\"colorscheme\": {\"prop\": \"b\", \"gradient\": \"roygb\", \"min\": 50, \"max\": 90}}});\n",
              "\tviewer_17580091717215044.addStyle({\"model\": 1, \"stick\": {}});\n",
              "\tviewer_17580091717215044.zoomTo();\n",
              "viewer_17580091717215044.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "conservative_substitutions = {\n",
        "    'A': ['G', 'S', 'T'], # Alanine (small, nonpolar)\n",
        "    'R': ['K', 'H'],     # Arginine (basic, positive)\n",
        "    'N': ['Q', 'S', 'T'], # Asparagine (polar, uncharged)\n",
        "    'D': ['E'],          # Aspartic Acid (acidic, negative)\n",
        "    'C': ['S'],          # Cysteine (sulfur-containing)\n",
        "    'Q': ['N', 'H', 'K', 'R', 'E', 'D'], # Glutamine (polar, uncharged) - larger group due to flexibility\n",
        "    'E': ['D', 'Q', 'N', 'K', 'R', 'H'], # Glutamic Acid (acidic, negative) - larger group due to flexibility\n",
        "    'G': ['A', 'S'],     # Glycine (small, nonpolar)\n",
        "    'H': ['R', 'K', 'N', 'Q'], # Histidine (basic, positive, aromatic)\n",
        "    'I': ['L', 'V', 'M'], # Isoleucine (hydrophobic)\n",
        "    'L': ['I', 'V', 'M'], # Leucine (hydrophobic)\n",
        "    'K': ['R', 'H', 'Q', 'N'], # Lysine (basic, positive)\n",
        "    'M': ['L', 'I', 'V'], # Methionine (hydrophobic, sulfur-containing)\n",
        "    'F': ['Y', 'W'],     # Phenylalanine (aromatic, hydrophobic)\n",
        "    'P': [],             # Proline (cyclic, often disrupts structure - few conservative substitutions)\n",
        "    'S': ['T', 'N', 'Q', 'A', 'G', 'C'], # Serine (polar, uncharged, hydroxyl)\n",
        "    'T': ['S', 'N', 'Q', 'A', 'G'], # Threonine (polar, uncharged, hydroxyl)\n",
        "    'W': ['Y', 'F'],     # Tryptophan (aromatic, hydrophobic, large)\n",
        "    'Y': ['F', 'W'],     # Tyrosine (aromatic, hydrophobic, hydroxyl)\n",
        "    'V': ['I', 'L', 'M']  # Valine (hydrophobic)\n",
        "}"
      ],
      "metadata": {
        "id": "xvDihPkUhKFJ"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def generate_conservative_mutations(dna_sequence, codon_position):\n",
        "    \"\"\"\n",
        "    Generates a list of mutated DNA sequences with conservative amino acid\n",
        "    substitutions at a given codon position.\n",
        "\n",
        "    Args:\n",
        "        dna_sequence: The original DNA sequence string.\n",
        "        codon_position: The 0-indexed position of the codon to mutate.\n",
        "\n",
        "    Returns:\n",
        "        A list of mutated DNA sequences.\n",
        "    \"\"\"\n",
        "    start_index = codon_position * 3\n",
        "    end_index = start_index + 3\n",
        "    original_codon = dna_sequence[start_index:end_index]\n",
        "\n",
        "    if original_codon not in codon_table:\n",
        "        print(f\"Warning: Original codon '{original_codon}' not found in codon table.\")\n",
        "        return []\n",
        "\n",
        "    original_amino_acid = codon_table[original_codon]\n",
        "\n",
        "    if original_amino_acid not in conservative_substitutions:\n",
        "        # Handle stop codons or amino acids with no defined conservative substitutions\n",
        "        return []\n",
        "\n",
        "    conservative_aas = conservative_substitutions[original_amino_acid]\n",
        "\n",
        "    mutated_sequences = []\n",
        "    for sub_aa in conservative_aas:\n",
        "        for codon, aa in codon_table.items():\n",
        "            if aa == sub_aa:\n",
        "                mutated_seq = list(dna_sequence)\n",
        "                mutated_seq[start_index:end_index] = list(codon)\n",
        "                mutated_sequences.append(\"\".join(mutated_seq))\n",
        "\n",
        "    return mutated_sequences\n",
        "\n",
        "# Example usage (assuming original_seq is defined from previous cells)\n",
        "# For testing, let's use the first codon (position 0)\n",
        "# mutated_dna_list = generate_conservative_mutations(original_seq, 0)\n",
        "# print(f\"Original DNA: {original_seq}\")\n",
        "# print(f\"Mutated DNA sequences (conservative substitutions at position 0):\")\n",
        "# for mut_seq in mutated_dna_list:\n",
        "#     print(mut_seq)\n"
      ],
      "metadata": {
        "id": "pQjQhtNNhN3l"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# @title auto_seq\n",
        "\n",
        "import numpy as np\n",
        "import numpy.random as random\n",
        "\n",
        "num_iterations = 100 #@param\n",
        "\n",
        "best_dna_sequence = original_seq\n",
        "#best_dna_sequence = \"ATGCGTGTGTTATTTATTCGTCGTGGTGAATTACGTAACATTGGTGAAGCTTTTGCGCAACAATATCTGTTTCGTAATGGCTTAGCGATTGAAGCG\"\n",
        "best_protein_sequence = str(Bio.Seq.Seq(best_dna_sequence).translate(to_stop=False))\n",
        "best_total_score = -1\n",
        "\n",
        "accepted_mutation_scores = []\n",
        "accepted_sim_scores = []\n",
        "accepted_total_scores = []\n",
        "accepted_iterations = []\n",
        "all_mutation_scores = []\n",
        "all_sim_scores = []\n",
        "all_total_scores = []\n",
        "all_iterations = []\n",
        "\n",
        "all_tried_protein_sequences = []\n",
        "all_tried_protein_sequences.append(best_protein_sequence)\n",
        "\n",
        "completed_iters = 0\n",
        "num_codons = len(original_seq) // 3\n",
        "\n",
        "while completed_iters < num_iterations:\n",
        "    # Choose a random codon position to mutate\n",
        "    random_codon_pos = random.randint(0, num_codons - 1)\n",
        "\n",
        "    # Generate conservative mutations for this position\n",
        "    if completed_iters == 0:\n",
        "        mutated_dna_list_single = best_dna_sequence\n",
        "    else:\n",
        "        mutated_dna_list_single = generate_conservative_mutations(best_dna_sequence, random_codon_pos)\n",
        "\n",
        "    if not mutated_dna_list_single:\n",
        "        continue # Skip to the next iteration if no mutations are possible\n",
        "\n",
        "    # Select the first generated mutation\n",
        "    single_mutated_dna_seq = mutated_dna_list_single[0]\n",
        "    single_mutated_protein_seq = str(Bio.Seq.Seq(single_mutated_dna_seq).translate(to_stop=False))\n",
        "\n",
        "    # Check if protein sequence has been tried before\n",
        "    if single_mutated_protein_seq in all_tried_protein_sequences:\n",
        "        continue\n",
        "\n",
        "    all_tried_protein_sequences.append(single_mutated_protein_seq)\n",
        "\n",
        "    # Predict structure using ESMFold\n",
        "    try:\n",
        "        single_mutated_output = model.infer(single_mutated_protein_seq,\n",
        "                                     num_recycles=num_recycles,\n",
        "                                     chain_linker=\"X\"*chain_linker,\n",
        "                                     residue_index_offset=512)\n",
        "\n",
        "        # Extract PDB string\n",
        "        single_mutated_pdb_str = model.output_to_pdb(single_mutated_output)[0]\n",
        "\n",
        "        # Align to reference and calculate RMSD\n",
        "        aligned_file_single, single_mutated_rmsd_val = align_pred_to_ref(single_mutated_pdb_str, \"reference.pdb\")\n",
        "\n",
        "        # Convert RMSD to similarity score\n",
        "        single_mutated_pred_sim_score = rmsd_to_similarity(single_mutated_rmsd_val)\n",
        "\n",
        "        # Calculate mutation score\n",
        "        single_current_mutation_score = mutation_score(original_seq, single_mutated_dna_seq)\n",
        "\n",
        "        # Calculate total score\n",
        "        single_current_total_score = single_current_mutation_score * single_mutated_pred_sim_score\n",
        "\n",
        "        # Store scores for ALL tried sequences\n",
        "        all_iterations.append(iteration + 1)\n",
        "        all_mutation_scores.append(single_current_mutation_score)\n",
        "        all_sim_scores.append(single_mutated_pred_sim_score)\n",
        "        all_total_scores.append(single_current_total_score)\n",
        "\n",
        "        # Print trial sequence and total score\n",
        "        print(f\"Iteration {iteration + 1}: Trial Sequence: {single_mutated_protein_seq}, Total Score: {single_current_total_score:.3f}\")\n",
        "\n",
        "        # Acceptance criteria: accept if total score improves\n",
        "        if single_current_total_score > best_total_score or single_mutated_pred_sim_score > 0.9:\n",
        "            best_total_score = single_current_total_score\n",
        "            best_dna_sequence = single_mutated_dna_seq\n",
        "            best_protein_sequence = single_mutated_protein_seq\n",
        "\n",
        "            # Store scores for ACCEPTED sequences\n",
        "            accepted_iterations.append(iteration + 1)\n",
        "            accepted_mutation_scores.append(single_current_mutation_score)\n",
        "            accepted_sim_scores.append(single_mutated_pred_sim_score)\n",
        "            accepted_total_scores.append(single_current_total_score)\n",
        "\n",
        "            #print(f\"  Accepted mutation. New best total score: {best_total_score:.3f}\")\n",
        "\n",
        "        completed_iters += 1\n",
        "\n",
        "    except Exception as e:\n",
        "        print(f\"Iteration {iteration + 1}: Error processing mutation: {e}. Skipping.\")\n",
        "        continue # Continue to the next iteration if an error occurs\n",
        "\n",
        "print(\"\\nIterative mutation process complete.\")\n",
        "print(f\"Best Mutated DNA Sequence: {best_dna_sequence}\")\n",
        "print(f\"Best Mutated Protein Sequence: {best_protein_sequence}\")\n",
        "# Recalculate final scores for the best sequence\n",
        "best_mutation_score = mutation_score(original_seq, best_dna_sequence)\n",
        "# To get the final RMSD and prediction similarity score for the best sequence,\n",
        "# you would need to re-run the prediction and alignment for best_protein_sequence\n",
        "# or store these values when the best sequence is updated.\n",
        "# For simplicity, we'll just report the best total score found.\n",
        "print(f\"Best Total Score Found: {best_total_score:.3f}\")\n",
        "print(f\"Total number of tried protein sequences: {len(all_tried_protein_sequences)}\")\n",
        "print(f\"Total number of accepted mutations: {len(accepted_total_scores)}\")"
      ],
      "metadata": {
        "id": "kHO4EmWdi9CC",
        "outputId": "01f2706a-7136-4398-dc28-5a7590c7cf53",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 451
        }
      },
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.12/dist-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "name 'iteration' is not defined",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-3645738639.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     52\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m         single_mutated_output = model.infer(single_mutated_protein_seq,\n\u001b[0m\u001b[1;32m     54\u001b[0m                                      \u001b[0mnum_recycles\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnum_recycles\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'model' is not defined",
            "\nDuring handling of the above exception, another exception occurred:\n",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-3645738639.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     97\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 99\u001b[0;31m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Iteration {iteration + 1}: Error processing mutation: {e}. Skipping.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    100\u001b[0m         \u001b[0;32mcontinue\u001b[0m \u001b[0;31m# Continue to the next iteration if an error occurs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'iteration' is not defined"
          ]
        }
      ]
    }
  ]
}